{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### import library"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3/dist-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.26.4\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import mannwhitneyu\n",
    "from scipy.stats import fisher_exact\n",
    "from sklearn.metrics import confusion_matrix\n",
    "# import utility\n",
    "import importlib\n",
    "pd.set_option('mode.chained_assignment', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### function to generate mutation file combined"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parquet_to_csv_converter(input_folder_path, output_folder_path, flank_size, genome, complete_data):\n",
    "\n",
    "    import os\n",
    "    import pandas as pd\n",
    "\n",
    "    counter = 0\n",
    "\n",
    "    for filename in os.listdir(input_folder_path):\n",
    "\n",
    "        try:\n",
    "            if ((os.path.isfile(os.path.join(input_folder_path, filename))==True) & (os.path.join(input_folder_path, filename).endswith(\".tsv\")==True)):\n",
    "\n",
    "                # df = pd.read_parquet(os.path.join(input_folder_path, filename))\n",
    "                counter  = counter + 1\n",
    "                dfi = pd.read_csv(os.path.join(input_folder_path, filename), sep = \"\\t\", encoding='utf-8')\n",
    "                \n",
    "                # print(\"Converted file is :\" + str(counter) + \" with file name \" + filename)\n",
    "                # print(dfi.columns)\n",
    "                df = dfi[[\"CHROM\", \"POS\", \"REF\", \"ALT\", \"ANN\", 'G70_ref_counts', 'G1410_ref_counts', 'G2640_ref_counts',\n",
    "       'G5150_ref_counts', 'G7530_ref_counts', 'G10150_ref_counts',\n",
    "       'G70_alt_counts', 'G1410_alt_counts', 'G2640_alt_counts',\n",
    "       'G5150_alt_counts', 'G7530_alt_counts', 'G10150_alt_counts']]\n",
    "\n",
    "                for i in range(df.shape[0]):\n",
    "\n",
    "                  if ((len(df.loc[i, \"REF\"])==1) & (len(df.loc[i, \"ALT\"])==1)):\n",
    "\n",
    "                    # print(\"Hi\")\n",
    "\n",
    "                    df.loc[i, \"up stream\"] = genome.fetch(df.loc[i, \"CHROM\"], df.loc[i, \"POS\"]-flank_size-1, df.loc[i, \"POS\"]-1)\n",
    "                    df.loc[i, \"down stram\"] = genome.fetch(df.loc[i, \"CHROM\"], df.loc[i, \"POS\"], df.loc[i, \"POS\"] + flank_size)\n",
    "\n",
    "                df[\"ref_flank\"] = df[\"up stream\"]+ df[\"REF\"] + df[\"down stram\"]\n",
    "                df[\"alt_flank\"] = df[\"up stream\"]+ df[\"ALT\"] + df[\"down stram\"]\n",
    "                # df.to_excel(output_folder_path + os.path.splitext(os.path.basename(os.path.join(input_folder_path, filename)))[0] + \".xlsx\", index = False)\n",
    "\n",
    "                df[\"file_name\"] = filename\n",
    "\n",
    "                df['population_name'] = os.path.splitext(os.path.basename(os.path.join(input_folder_path, filename)))[0][0:os.path.splitext(os.path.basename(os.path.join(input_folder_path, filename)))[0].find(\"_\")]\n",
    "\n",
    "\n",
    "                # print(\"File completed: \", counter)\n",
    "                complete_data = pd.concat([complete_data, df])\n",
    "\n",
    "\n",
    "\n",
    "        except:\n",
    "            continue\n",
    "\n",
    "        # break\n",
    "    return complete_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_folder_path = \"../data/well_output/\"\n",
    "output_folder_path = \"../Results/VCF/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Generate flank data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "file `../data/w303_vlte.fasta` not found",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_82601/229410136.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mkmer_length\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mfile_path_yeast\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"../data/w303_vlte.fasta\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mgenome\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpysam\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mFastaFile\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile_path_yeast\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mcomplete_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparquet_to_csv_converter\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput_folder_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput_folder_path\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mflank_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgenome\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcomplete_data\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/pysam/libcfaidx.pyx\u001b[0m in \u001b[0;36mpysam.libcfaidx.FastaFile.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/pysam/libcfaidx.pyx\u001b[0m in \u001b[0;36mpysam.libcfaidx.FastaFile._open\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: file `../data/w303_vlte.fasta` not found"
     ]
    }
   ],
   "source": [
    "complete_data = pd.DataFrame(columns = [\"CHROM\", \"POS\", \"REF\", \"ALT\", \"ANN\", 'G70_ref_counts', 'G1410_ref_counts', 'G2640_ref_counts',\n",
    "       'G5150_ref_counts', 'G7530_ref_counts', 'G10150_ref_counts',\n",
    "       'G70_alt_counts', 'G1410_alt_counts', 'G2640_alt_counts',\n",
    "       'G5150_alt_counts', 'G7530_alt_counts', 'G10150_alt_counts', 'population_name'])\n",
    "\n",
    "flank_size = 10\n",
    "kmer_length =10\n",
    "file_path_yeast = \"../data/w303_vlte.fasta\"\n",
    "genome = pysam.FastaFile(file_path_yeast)\n",
    "\n",
    "complete_data = parquet_to_csv_converter(input_folder_path, output_folder_path, flank_size, genome, complete_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ANN</th>\n",
       "      <th>G70_ref_counts</th>\n",
       "      <th>G1410_ref_counts</th>\n",
       "      <th>G2640_ref_counts</th>\n",
       "      <th>G5150_ref_counts</th>\n",
       "      <th>G7530_ref_counts</th>\n",
       "      <th>G10150_ref_counts</th>\n",
       "      <th>G70_alt_counts</th>\n",
       "      <th>G1410_alt_counts</th>\n",
       "      <th>G2640_alt_counts</th>\n",
       "      <th>G5150_alt_counts</th>\n",
       "      <th>G7530_alt_counts</th>\n",
       "      <th>G10150_alt_counts</th>\n",
       "      <th>population_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [CHROM, POS, REF, ALT, ANN, G70_ref_counts, G1410_ref_counts, G2640_ref_counts, G5150_ref_counts, G7530_ref_counts, G10150_ref_counts, G70_alt_counts, G1410_alt_counts, G2640_alt_counts, G5150_alt_counts, G7530_alt_counts, G10150_alt_counts, population_name]\n",
       "Index: []"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "['up stream']",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_82601/3532591416.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mcomplete_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdropna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m\"up stream\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mcomplete_data\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.10/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36mdropna\u001b[0;34m(self, axis, how, thresh, subset, inplace, ignore_index)\u001b[0m\n\u001b[1;32m   6655\u001b[0m             \u001b[0mcheck\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mindices\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6656\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mcheck\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6657\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msubset\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcheck\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtolist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6658\u001b[0m             \u001b[0magg_obj\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindices\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0magg_axis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6659\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: ['up stream']"
     ]
    }
   ],
   "source": [
    "complete_data.dropna(subset = [\"up stream\"], inplace=True, axis = 0)\n",
    "complete_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>ANN</th>\n",
       "      <th>G70_ref_counts</th>\n",
       "      <th>G1410_ref_counts</th>\n",
       "      <th>G2640_ref_counts</th>\n",
       "      <th>G5150_ref_counts</th>\n",
       "      <th>G7530_ref_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>G2640_alt_counts</th>\n",
       "      <th>G5150_alt_counts</th>\n",
       "      <th>G7530_alt_counts</th>\n",
       "      <th>G10150_alt_counts</th>\n",
       "      <th>population_name</th>\n",
       "      <th>up stream</th>\n",
       "      <th>down stram</th>\n",
       "      <th>ref_flank</th>\n",
       "      <th>alt_flank</th>\n",
       "      <th>file_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>135</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YAL069W|YAL06...</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>CCATTACCCT</td>\n",
       "      <td>CCTCCACTCG</td>\n",
       "      <td>CCATTACCCTGCCTCCACTCG</td>\n",
       "      <td>CCATTACCCTACCTCCACTCG</td>\n",
       "      <td>P1C08_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>140</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|YAL069W|YAL06...</td>\n",
       "      <td>168.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>ACCCTGCCTC</td>\n",
       "      <td>ACTCGTTACC</td>\n",
       "      <td>ACCCTGCCTCCACTCGTTACC</td>\n",
       "      <td>ACCCTGCCTCTACTCGTTACC</td>\n",
       "      <td>P1C08_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>366</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL069W|YAL069W|trans...</td>\n",
       "      <td>88.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>TGCTTACCCT</td>\n",
       "      <td>CCACTTTATA</td>\n",
       "      <td>TGCTTACCCTACCACTTTATA</td>\n",
       "      <td>TGCTTACCCTGCCACTTTATA</td>\n",
       "      <td>P1C08_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>374</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G|missense_variant|MODERATE|YAL069W|YAL069W|tr...</td>\n",
       "      <td>94.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>111.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>CTACCACTTT</td>\n",
       "      <td>TACCACCACC</td>\n",
       "      <td>CTACCACTTTATACCACCACC</td>\n",
       "      <td>CTACCACTTTGTACCACCACC</td>\n",
       "      <td>P1C08_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>395</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>G|missense_variant|MODERATE|YAL069W|YAL069W|tr...</td>\n",
       "      <td>90.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>121.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>ACATGCCATA</td>\n",
       "      <td>TCACCCTCAC</td>\n",
       "      <td>ACATGCCATACTCACCCTCAC</td>\n",
       "      <td>ACATGCCATAGTCACCCTCAC</td>\n",
       "      <td>P1C08_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144704</th>\n",
       "      <td>chrMito</td>\n",
       "      <td>69308</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>T|intergenic_region|MODIFIER|||||||||||||</td>\n",
       "      <td>51.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>TATTTTCCGT</td>\n",
       "      <td>TATATCTTTA</td>\n",
       "      <td>TATTTTCCGTATATATCTTTA</td>\n",
       "      <td>TATTTTCCGTTTATATCTTTA</td>\n",
       "      <td>P2G09_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144705</th>\n",
       "      <td>chrMito</td>\n",
       "      <td>69315</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>C|intergenic_region|MODIFIER|||||||||||||</td>\n",
       "      <td>51.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>63.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>CGTATATATC</td>\n",
       "      <td>TTAATTTAAT</td>\n",
       "      <td>CGTATATATCTTTAATTTAAT</td>\n",
       "      <td>CGTATATATCCTTAATTTAAT</td>\n",
       "      <td>P2G09_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144706</th>\n",
       "      <td>chrMito</td>\n",
       "      <td>71519</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G|intergenic_region|MODIFIER|||||||||||||</td>\n",
       "      <td>90.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>68.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>AATTCCTATT</td>\n",
       "      <td>AGGACGATAA</td>\n",
       "      <td>AATTCCTATTAAGGACGATAA</td>\n",
       "      <td>AATTCCTATTGAGGACGATAA</td>\n",
       "      <td>P2G09_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144707</th>\n",
       "      <td>chrMito</td>\n",
       "      <td>71520</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>G|intergenic_region|MODIFIER|||||||||||||</td>\n",
       "      <td>90.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>ATTCCTATTA</td>\n",
       "      <td>GGACGATAAT</td>\n",
       "      <td>ATTCCTATTAAGGACGATAAT</td>\n",
       "      <td>ATTCCTATTAGGGACGATAAT</td>\n",
       "      <td>P2G09_full.tsv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144708</th>\n",
       "      <td>chrMito</td>\n",
       "      <td>71534</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>C|intergenic_region|MODIFIER|||||||||||||</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>CGATAATAAT</td>\n",
       "      <td>TATATATATT</td>\n",
       "      <td>CGATAATAATATATATATATT</td>\n",
       "      <td>CGATAATAATCTATATATATT</td>\n",
       "      <td>P2G09_full.tsv</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>144709 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          CHROM    POS REF ALT  \\\n",
       "0          chrI    135   G   A   \n",
       "1          chrI    140   C   T   \n",
       "2          chrI    366   A   G   \n",
       "3          chrI    374   A   G   \n",
       "4          chrI    395   C   G   \n",
       "...         ...    ...  ..  ..   \n",
       "144704  chrMito  69308   A   T   \n",
       "144705  chrMito  69315   T   C   \n",
       "144706  chrMito  71519   A   G   \n",
       "144707  chrMito  71520   A   G   \n",
       "144708  chrMito  71534   A   C   \n",
       "\n",
       "                                                      ANN  G70_ref_counts  \\\n",
       "0       A|upstream_gene_variant|MODIFIER|YAL069W|YAL06...           170.0   \n",
       "1       T|upstream_gene_variant|MODIFIER|YAL069W|YAL06...           168.0   \n",
       "2       G|synonymous_variant|LOW|YAL069W|YAL069W|trans...            88.0   \n",
       "3       G|missense_variant|MODERATE|YAL069W|YAL069W|tr...            94.0   \n",
       "4       G|missense_variant|MODERATE|YAL069W|YAL069W|tr...            90.0   \n",
       "...                                                   ...             ...   \n",
       "144704          T|intergenic_region|MODIFIER|||||||||||||            51.0   \n",
       "144705          C|intergenic_region|MODIFIER|||||||||||||            51.0   \n",
       "144706          G|intergenic_region|MODIFIER|||||||||||||            90.0   \n",
       "144707          G|intergenic_region|MODIFIER|||||||||||||            90.0   \n",
       "144708          C|intergenic_region|MODIFIER|||||||||||||            73.0   \n",
       "\n",
       "        G1410_ref_counts  G2640_ref_counts  G5150_ref_counts  \\\n",
       "0                   34.0             120.0              59.0   \n",
       "1                   38.0             123.0              60.0   \n",
       "2                   40.0              81.0              54.0   \n",
       "3                   40.0              81.0              55.0   \n",
       "4                   45.0              81.0              52.0   \n",
       "...                  ...               ...               ...   \n",
       "144704              44.0              62.0              54.0   \n",
       "144705              44.0              62.0              49.0   \n",
       "144706              32.0              60.0              39.0   \n",
       "144707              29.0              60.0              39.0   \n",
       "144708              19.0              44.0              14.0   \n",
       "\n",
       "        G7530_ref_counts  ...  G2640_alt_counts  G5150_alt_counts  \\\n",
       "0                  181.0  ...              14.0               9.0   \n",
       "1                  179.0  ...              14.0               7.0   \n",
       "2                  111.0  ...               0.0               0.0   \n",
       "3                  111.0  ...               0.0               4.0   \n",
       "4                  121.0  ...               0.0               6.0   \n",
       "...                  ...  ...               ...               ...   \n",
       "144704              63.0  ...               0.0               6.0   \n",
       "144705              63.0  ...               0.0               6.0   \n",
       "144706              68.0  ...               5.0               0.0   \n",
       "144707              67.0  ...               5.0               0.0   \n",
       "144708              50.0  ...               2.0               0.0   \n",
       "\n",
       "        G7530_alt_counts  G10150_alt_counts  population_name   up stream  \\\n",
       "0                    0.0                6.0            P1C08  CCATTACCCT   \n",
       "1                    0.0                6.0            P1C08  ACCCTGCCTC   \n",
       "2                    0.0                0.0            P1C08  TGCTTACCCT   \n",
       "3                    0.0                0.0            P1C08  CTACCACTTT   \n",
       "4                    0.0                0.0            P1C08  ACATGCCATA   \n",
       "...                  ...                ...              ...         ...   \n",
       "144704               0.0                0.0            P2G09  TATTTTCCGT   \n",
       "144705               0.0                0.0            P2G09  CGTATATATC   \n",
       "144706               6.0                0.0            P2G09  AATTCCTATT   \n",
       "144707               5.0                0.0            P2G09  ATTCCTATTA   \n",
       "144708               3.0                0.0            P2G09  CGATAATAAT   \n",
       "\n",
       "        down stram              ref_flank              alt_flank  \\\n",
       "0       CCTCCACTCG  CCATTACCCTGCCTCCACTCG  CCATTACCCTACCTCCACTCG   \n",
       "1       ACTCGTTACC  ACCCTGCCTCCACTCGTTACC  ACCCTGCCTCTACTCGTTACC   \n",
       "2       CCACTTTATA  TGCTTACCCTACCACTTTATA  TGCTTACCCTGCCACTTTATA   \n",
       "3       TACCACCACC  CTACCACTTTATACCACCACC  CTACCACTTTGTACCACCACC   \n",
       "4       TCACCCTCAC  ACATGCCATACTCACCCTCAC  ACATGCCATAGTCACCCTCAC   \n",
       "...            ...                    ...                    ...   \n",
       "144704  TATATCTTTA  TATTTTCCGTATATATCTTTA  TATTTTCCGTTTATATCTTTA   \n",
       "144705  TTAATTTAAT  CGTATATATCTTTAATTTAAT  CGTATATATCCTTAATTTAAT   \n",
       "144706  AGGACGATAA  AATTCCTATTAAGGACGATAA  AATTCCTATTGAGGACGATAA   \n",
       "144707  GGACGATAAT  ATTCCTATTAAGGACGATAAT  ATTCCTATTAGGGACGATAAT   \n",
       "144708  TATATATATT  CGATAATAATATATATATATT  CGATAATAATCTATATATATT   \n",
       "\n",
       "             file_name  \n",
       "0       P1C08_full.tsv  \n",
       "1       P1C08_full.tsv  \n",
       "2       P1C08_full.tsv  \n",
       "3       P1C08_full.tsv  \n",
       "4       P1C08_full.tsv  \n",
       "...                ...  \n",
       "144704  P2G09_full.tsv  \n",
       "144705  P2G09_full.tsv  \n",
       "144706  P2G09_full.tsv  \n",
       "144707  P2G09_full.tsv  \n",
       "144708  P2G09_full.tsv  \n",
       "\n",
       "[144709 rows x 23 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "complete_data.reset_index(inplace =True, drop =True)\n",
    "complete_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import kaos\n",
    "kmer_length =10\n",
    "file_path_yeast = \"../data/w303_vlte.fasta\"\n",
    "fcgr_sequence_yeast = kaos.kaos_fcgr().chaos_frequency_dictionary(fasta_string= kaos.Read().read_fasta_removing_non_atgc(file_path_yeast), kmer_length=kmer_length, pseudo_count=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "rows = []\n",
    "\n",
    "kmer_length = 10\n",
    "\n",
    "df_merge_int_snp = complete_data.copy()\n",
    "\n",
    "count = 0\n",
    "\n",
    "for i in range(df_merge_int_snp.shape[0]):\n",
    "\n",
    "    try:\n",
    "\n",
    "        ref_seq = df_merge_int_snp.loc[i, 'ref_flank']\n",
    "        alt_seq = df_merge_int_snp.loc[i, 'alt_flank']\n",
    "        center_pos = int(len(ref_seq) / 2)\n",
    "\n",
    "        # print(ref_seq, alt_seq)\n",
    "\n",
    "        # print(i)\n",
    "        \n",
    "        if not (set(list(set(list(ref_seq)))).issubset(['G', 'T', 'A', 'C'])):\n",
    "            continue\n",
    "        if not (set(list(set(list(alt_seq)))).issubset(['G', 'T', 'A', 'C'])):\n",
    "            continue     \n",
    "\n",
    "        for j in range(kmer_length):\n",
    "            altered_allele = alt_seq[center_pos - kmer_length + j + 1:center_pos + 1 + j]\n",
    "            reference_allele = ref_seq[center_pos - kmer_length + j + 1:center_pos + j + 1]\n",
    "\n",
    "            # print(altered_allele, reference_allele)\n",
    "            # if (fcgr_sequence_yeast[altered_allele]==0):\n",
    "            #     count = count +1\n",
    "            rows.append({\n",
    "                'altered allete': altered_allele,\n",
    "                'reference allele': reference_allele,\n",
    "                'gain': np.log(fcgr_sequence_yeast[altered_allele] / fcgr_sequence_yeast[reference_allele]),\n",
    "                'Position': df_merge_int_snp.loc[i, 'POS'],\n",
    "                'CHROM': df_merge_int_snp.loc[i, 'CHROM'],\n",
    "                'REF': df_merge_int_snp.loc[i, 'REF'],\n",
    "                'ALT': df_merge_int_snp.loc[i, 'ALT'],\n",
    "                # 'group_mutation_type' : df_merge_int_snp.loc[i, 'group_mutation_type'],\n",
    "                'population_name' : df_merge_int_snp.loc[i, 'population_name'],\n",
    "                'G70_ref_counts': df_merge_int_snp.loc[i, 'G70_ref_counts'], \n",
    "                'G1410_ref_counts': df_merge_int_snp.loc[i, 'G1410_ref_counts'], \n",
    "                'G2640_ref_counts': df_merge_int_snp.loc[i, 'G2640_ref_counts'], \n",
    "                'G5150_ref_counts': df_merge_int_snp.loc[i, 'G5150_ref_counts'], \n",
    "                'G7530_ref_counts': df_merge_int_snp.loc[i, 'G7530_ref_counts'],\n",
    "                'G10150_ref_counts': df_merge_int_snp.loc[i, 'G10150_ref_counts'], \n",
    "                'G70_alt_counts': df_merge_int_snp.loc[i, 'G70_alt_counts'],\n",
    "                'G1410_alt_counts': df_merge_int_snp.loc[i, 'G1410_alt_counts'],\n",
    "                'G2640_alt_counts': df_merge_int_snp.loc[i, 'G2640_alt_counts'],\n",
    "                'G5150_alt_counts': df_merge_int_snp.loc[i, 'G5150_alt_counts'],\n",
    "                'G7530_alt_counts': df_merge_int_snp.loc[i, 'G7530_alt_counts'],\n",
    "                'G10150_alt_counts': df_merge_int_snp.loc[i, 'G10150_alt_counts']\n",
    "        })\n",
    "    except:\n",
    "        continue\n",
    "            \n",
    "\n",
    "        \n",
    "    # del altered_allele, reference_allele, ref_seq, alt_seq, center_pos\n",
    "        \n",
    "#     if(i==240000):\n",
    "#         break\n",
    "\n",
    "\n",
    "df_complete = pd.DataFrame(rows)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# complete_data.to_"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_columns = [\"Position\", \"CHROM\", \"REF\", \"ALT\", 'population_name']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>altered allete</th>\n",
       "      <th>reference allele</th>\n",
       "      <th>gain</th>\n",
       "      <th>Position</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>population_name</th>\n",
       "      <th>G70_ref_counts</th>\n",
       "      <th>G1410_ref_counts</th>\n",
       "      <th>G2640_ref_counts</th>\n",
       "      <th>G5150_ref_counts</th>\n",
       "      <th>G7530_ref_counts</th>\n",
       "      <th>G10150_ref_counts</th>\n",
       "      <th>G70_alt_counts</th>\n",
       "      <th>G1410_alt_counts</th>\n",
       "      <th>G2640_alt_counts</th>\n",
       "      <th>G5150_alt_counts</th>\n",
       "      <th>G7530_alt_counts</th>\n",
       "      <th>G10150_alt_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATTACCCTA</td>\n",
       "      <td>CATTACCCTG</td>\n",
       "      <td>0.955511</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATTACCCTAC</td>\n",
       "      <td>ATTACCCTGC</td>\n",
       "      <td>1.131402</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTACCCTACC</td>\n",
       "      <td>TTACCCTGCC</td>\n",
       "      <td>0.716678</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACCCTACCT</td>\n",
       "      <td>TACCCTGCCT</td>\n",
       "      <td>0.635989</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACCCTACCTC</td>\n",
       "      <td>ACCCTGCCTC</td>\n",
       "      <td>1.178655</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335375</th>\n",
       "      <td>TAATCTATAT</td>\n",
       "      <td>TAATATATAT</td>\n",
       "      <td>-2.197225</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335376</th>\n",
       "      <td>AATCTATATA</td>\n",
       "      <td>AATATATATA</td>\n",
       "      <td>-2.297165</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335377</th>\n",
       "      <td>ATCTATATAT</td>\n",
       "      <td>ATATATATAT</td>\n",
       "      <td>-3.315454</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335378</th>\n",
       "      <td>TCTATATATA</td>\n",
       "      <td>TATATATATA</td>\n",
       "      <td>-3.143530</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335379</th>\n",
       "      <td>CTATATATAT</td>\n",
       "      <td>ATATATATAT</td>\n",
       "      <td>-3.135072</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1335380 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        altered allete reference allele      gain  Position    CHROM REF ALT  \\\n",
       "0           CATTACCCTA       CATTACCCTG  0.955511       135     chrI   G   A   \n",
       "1           ATTACCCTAC       ATTACCCTGC  1.131402       135     chrI   G   A   \n",
       "2           TTACCCTACC       TTACCCTGCC  0.716678       135     chrI   G   A   \n",
       "3           TACCCTACCT       TACCCTGCCT  0.635989       135     chrI   G   A   \n",
       "4           ACCCTACCTC       ACCCTGCCTC  1.178655       135     chrI   G   A   \n",
       "...                ...              ...       ...       ...      ...  ..  ..   \n",
       "1335375     TAATCTATAT       TAATATATAT -2.197225     71534  chrMito   A   C   \n",
       "1335376     AATCTATATA       AATATATATA -2.297165     71534  chrMito   A   C   \n",
       "1335377     ATCTATATAT       ATATATATAT -3.315454     71534  chrMito   A   C   \n",
       "1335378     TCTATATATA       TATATATATA -3.143530     71534  chrMito   A   C   \n",
       "1335379     CTATATATAT       ATATATATAT -3.135072     71534  chrMito   A   C   \n",
       "\n",
       "        population_name  G70_ref_counts  G1410_ref_counts  G2640_ref_counts  \\\n",
       "0                 P1C08           170.0              34.0             120.0   \n",
       "1                 P1C08           170.0              34.0             120.0   \n",
       "2                 P1C08           170.0              34.0             120.0   \n",
       "3                 P1C08           170.0              34.0             120.0   \n",
       "4                 P1C08           170.0              34.0             120.0   \n",
       "...                 ...             ...               ...               ...   \n",
       "1335375           P2G09            73.0              19.0              44.0   \n",
       "1335376           P2G09            73.0              19.0              44.0   \n",
       "1335377           P2G09            73.0              19.0              44.0   \n",
       "1335378           P2G09            73.0              19.0              44.0   \n",
       "1335379           P2G09            73.0              19.0              44.0   \n",
       "\n",
       "         G5150_ref_counts  G7530_ref_counts  G10150_ref_counts  \\\n",
       "0                    59.0             181.0               33.0   \n",
       "1                    59.0             181.0               33.0   \n",
       "2                    59.0             181.0               33.0   \n",
       "3                    59.0             181.0               33.0   \n",
       "4                    59.0             181.0               33.0   \n",
       "...                   ...               ...                ...   \n",
       "1335375              14.0              50.0               21.0   \n",
       "1335376              14.0              50.0               21.0   \n",
       "1335377              14.0              50.0               21.0   \n",
       "1335378              14.0              50.0               21.0   \n",
       "1335379              14.0              50.0               21.0   \n",
       "\n",
       "         G70_alt_counts  G1410_alt_counts  G2640_alt_counts  G5150_alt_counts  \\\n",
       "0                   0.0               5.0              14.0               9.0   \n",
       "1                   0.0               5.0              14.0               9.0   \n",
       "2                   0.0               5.0              14.0               9.0   \n",
       "3                   0.0               5.0              14.0               9.0   \n",
       "4                   0.0               5.0              14.0               9.0   \n",
       "...                 ...               ...               ...               ...   \n",
       "1335375             0.0               1.0               2.0               0.0   \n",
       "1335376             0.0               1.0               2.0               0.0   \n",
       "1335377             0.0               1.0               2.0               0.0   \n",
       "1335378             0.0               1.0               2.0               0.0   \n",
       "1335379             0.0               1.0               2.0               0.0   \n",
       "\n",
       "         G7530_alt_counts  G10150_alt_counts  \n",
       "0                     0.0                6.0  \n",
       "1                     0.0                6.0  \n",
       "2                     0.0                6.0  \n",
       "3                     0.0                6.0  \n",
       "4                     0.0                6.0  \n",
       "...                   ...                ...  \n",
       "1335375               3.0                0.0  \n",
       "1335376               3.0                0.0  \n",
       "1335377               3.0                0.0  \n",
       "1335378               3.0                0.0  \n",
       "1335379               3.0                0.0  \n",
       "\n",
       "[1335380 rows x 20 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>altered allete</th>\n",
       "      <th>reference allele</th>\n",
       "      <th>gain</th>\n",
       "      <th>Position</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>population_name</th>\n",
       "      <th>G70_ref_counts</th>\n",
       "      <th>G1410_ref_counts</th>\n",
       "      <th>G2640_ref_counts</th>\n",
       "      <th>G5150_ref_counts</th>\n",
       "      <th>G7530_ref_counts</th>\n",
       "      <th>G10150_ref_counts</th>\n",
       "      <th>G70_alt_counts</th>\n",
       "      <th>G1410_alt_counts</th>\n",
       "      <th>G2640_alt_counts</th>\n",
       "      <th>G5150_alt_counts</th>\n",
       "      <th>G7530_alt_counts</th>\n",
       "      <th>G10150_alt_counts</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATTACCCTAC</td>\n",
       "      <td>ATTACCCTGC</td>\n",
       "      <td>1.131402</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTACCCTACC</td>\n",
       "      <td>TTACCCTGCC</td>\n",
       "      <td>0.716678</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACCCTACCT</td>\n",
       "      <td>TACCCTGCCT</td>\n",
       "      <td>0.635989</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACCCTACCTC</td>\n",
       "      <td>ACCCTGCCTC</td>\n",
       "      <td>1.178655</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CCCTACCTCC</td>\n",
       "      <td>CCCTGCCTCC</td>\n",
       "      <td>0.251314</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335375</th>\n",
       "      <td>TAATCTATAT</td>\n",
       "      <td>TAATATATAT</td>\n",
       "      <td>-2.197225</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335376</th>\n",
       "      <td>AATCTATATA</td>\n",
       "      <td>AATATATATA</td>\n",
       "      <td>-2.297165</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335377</th>\n",
       "      <td>ATCTATATAT</td>\n",
       "      <td>ATATATATAT</td>\n",
       "      <td>-3.315454</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335378</th>\n",
       "      <td>TCTATATATA</td>\n",
       "      <td>TATATATATA</td>\n",
       "      <td>-3.143530</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335379</th>\n",
       "      <td>CTATATATAT</td>\n",
       "      <td>ATATATATAT</td>\n",
       "      <td>-3.135072</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1201842 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        altered allete reference allele      gain  Position    CHROM REF ALT  \\\n",
       "1           ATTACCCTAC       ATTACCCTGC  1.131402       135     chrI   G   A   \n",
       "2           TTACCCTACC       TTACCCTGCC  0.716678       135     chrI   G   A   \n",
       "3           TACCCTACCT       TACCCTGCCT  0.635989       135     chrI   G   A   \n",
       "4           ACCCTACCTC       ACCCTGCCTC  1.178655       135     chrI   G   A   \n",
       "5           CCCTACCTCC       CCCTGCCTCC  0.251314       135     chrI   G   A   \n",
       "...                ...              ...       ...       ...      ...  ..  ..   \n",
       "1335375     TAATCTATAT       TAATATATAT -2.197225     71534  chrMito   A   C   \n",
       "1335376     AATCTATATA       AATATATATA -2.297165     71534  chrMito   A   C   \n",
       "1335377     ATCTATATAT       ATATATATAT -3.315454     71534  chrMito   A   C   \n",
       "1335378     TCTATATATA       TATATATATA -3.143530     71534  chrMito   A   C   \n",
       "1335379     CTATATATAT       ATATATATAT -3.135072     71534  chrMito   A   C   \n",
       "\n",
       "        population_name  G70_ref_counts  G1410_ref_counts  G2640_ref_counts  \\\n",
       "1                 P1C08           170.0              34.0             120.0   \n",
       "2                 P1C08           170.0              34.0             120.0   \n",
       "3                 P1C08           170.0              34.0             120.0   \n",
       "4                 P1C08           170.0              34.0             120.0   \n",
       "5                 P1C08           170.0              34.0             120.0   \n",
       "...                 ...             ...               ...               ...   \n",
       "1335375           P2G09            73.0              19.0              44.0   \n",
       "1335376           P2G09            73.0              19.0              44.0   \n",
       "1335377           P2G09            73.0              19.0              44.0   \n",
       "1335378           P2G09            73.0              19.0              44.0   \n",
       "1335379           P2G09            73.0              19.0              44.0   \n",
       "\n",
       "         G5150_ref_counts  G7530_ref_counts  G10150_ref_counts  \\\n",
       "1                    59.0             181.0               33.0   \n",
       "2                    59.0             181.0               33.0   \n",
       "3                    59.0             181.0               33.0   \n",
       "4                    59.0             181.0               33.0   \n",
       "5                    59.0             181.0               33.0   \n",
       "...                   ...               ...                ...   \n",
       "1335375              14.0              50.0               21.0   \n",
       "1335376              14.0              50.0               21.0   \n",
       "1335377              14.0              50.0               21.0   \n",
       "1335378              14.0              50.0               21.0   \n",
       "1335379              14.0              50.0               21.0   \n",
       "\n",
       "         G70_alt_counts  G1410_alt_counts  G2640_alt_counts  G5150_alt_counts  \\\n",
       "1                   0.0               5.0              14.0               9.0   \n",
       "2                   0.0               5.0              14.0               9.0   \n",
       "3                   0.0               5.0              14.0               9.0   \n",
       "4                   0.0               5.0              14.0               9.0   \n",
       "5                   0.0               5.0              14.0               9.0   \n",
       "...                 ...               ...               ...               ...   \n",
       "1335375             0.0               1.0               2.0               0.0   \n",
       "1335376             0.0               1.0               2.0               0.0   \n",
       "1335377             0.0               1.0               2.0               0.0   \n",
       "1335378             0.0               1.0               2.0               0.0   \n",
       "1335379             0.0               1.0               2.0               0.0   \n",
       "\n",
       "         G7530_alt_counts  G10150_alt_counts  \n",
       "1                     0.0                6.0  \n",
       "2                     0.0                6.0  \n",
       "3                     0.0                6.0  \n",
       "4                     0.0                6.0  \n",
       "5                     0.0                6.0  \n",
       "...                   ...                ...  \n",
       "1335375               3.0                0.0  \n",
       "1335376               3.0                0.0  \n",
       "1335377               3.0                0.0  \n",
       "1335378               3.0                0.0  \n",
       "1335379               3.0                0.0  \n",
       "\n",
       "[1201842 rows x 20 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_complete[df_complete.duplicated(subset=key_columns)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>population_name</th>\n",
       "      <th>accumulate_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31</td>\n",
       "      <td>chrVIII</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P3G06</td>\n",
       "      <td>-3.577119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>34</td>\n",
       "      <td>2-micron</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>P1E04</td>\n",
       "      <td>0.392871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34</td>\n",
       "      <td>chrII</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P2C06</td>\n",
       "      <td>-4.563850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>36</td>\n",
       "      <td>chrIX</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P2F11</td>\n",
       "      <td>-11.012777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36</td>\n",
       "      <td>chrIX</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P3F11</td>\n",
       "      <td>-11.012777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133533</th>\n",
       "      <td>1525964</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>P3G10</td>\n",
       "      <td>8.835324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133534</th>\n",
       "      <td>1525964</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>P3G11</td>\n",
       "      <td>8.835324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133535</th>\n",
       "      <td>1525998</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P3B07</td>\n",
       "      <td>-14.496006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133536</th>\n",
       "      <td>1525998</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P3B10</td>\n",
       "      <td>-14.496006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133537</th>\n",
       "      <td>1526031</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>P3D05</td>\n",
       "      <td>-5.155277</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133538 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Position     CHROM REF ALT population_name  accumulate_gain\n",
       "0             31   chrVIII   C   A           P3G06        -3.577119\n",
       "1             34  2-micron   G   T           P1E04         0.392871\n",
       "2             34     chrII   G   A           P2C06        -4.563850\n",
       "3             36     chrIX   C   A           P2F11       -11.012777\n",
       "4             36     chrIX   C   A           P3F11       -11.012777\n",
       "...          ...       ...  ..  ..             ...              ...\n",
       "133533   1525964     chrIV   C   T           P3G10         8.835324\n",
       "133534   1525964     chrIV   C   T           P3G11         8.835324\n",
       "133535   1525998     chrIV   T   C           P3B07       -14.496006\n",
       "133536   1525998     chrIV   T   C           P3B10       -14.496006\n",
       "133537   1526031     chrIV   A   G           P3D05        -5.155277\n",
       "\n",
       "[133538 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped_df_with_accumulated = df_complete.groupby(key_columns).agg({'gain':'sum'}).reset_index()\n",
    "grouped_df_with_accumulated.rename(columns={'gain': 'accumulate_gain'}, inplace=True)\n",
    "grouped_df_with_accumulated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>altered allete</th>\n",
       "      <th>reference allele</th>\n",
       "      <th>gain</th>\n",
       "      <th>Position</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>population_name</th>\n",
       "      <th>G70_ref_counts</th>\n",
       "      <th>G1410_ref_counts</th>\n",
       "      <th>...</th>\n",
       "      <th>G5150_ref_counts</th>\n",
       "      <th>G7530_ref_counts</th>\n",
       "      <th>G10150_ref_counts</th>\n",
       "      <th>G70_alt_counts</th>\n",
       "      <th>G1410_alt_counts</th>\n",
       "      <th>G2640_alt_counts</th>\n",
       "      <th>G5150_alt_counts</th>\n",
       "      <th>G7530_alt_counts</th>\n",
       "      <th>G10150_alt_counts</th>\n",
       "      <th>accumulate_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CATTACCCTA</td>\n",
       "      <td>CATTACCCTG</td>\n",
       "      <td>0.955511</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.968131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ATTACCCTAC</td>\n",
       "      <td>ATTACCCTGC</td>\n",
       "      <td>1.131402</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.968131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TTACCCTACC</td>\n",
       "      <td>TTACCCTGCC</td>\n",
       "      <td>0.716678</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.968131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TACCCTACCT</td>\n",
       "      <td>TACCCTGCCT</td>\n",
       "      <td>0.635989</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.968131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ACCCTACCTC</td>\n",
       "      <td>ACCCTGCCTC</td>\n",
       "      <td>1.178655</td>\n",
       "      <td>135</td>\n",
       "      <td>chrI</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P1C08</td>\n",
       "      <td>170.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.968131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335375</th>\n",
       "      <td>TAATCTATAT</td>\n",
       "      <td>TAATATATAT</td>\n",
       "      <td>-2.197225</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-23.671064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335376</th>\n",
       "      <td>AATCTATATA</td>\n",
       "      <td>AATATATATA</td>\n",
       "      <td>-2.297165</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-23.671064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335377</th>\n",
       "      <td>ATCTATATAT</td>\n",
       "      <td>ATATATATAT</td>\n",
       "      <td>-3.315454</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-23.671064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335378</th>\n",
       "      <td>TCTATATATA</td>\n",
       "      <td>TATATATATA</td>\n",
       "      <td>-3.143530</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-23.671064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1335379</th>\n",
       "      <td>CTATATATAT</td>\n",
       "      <td>ATATATATAT</td>\n",
       "      <td>-3.135072</td>\n",
       "      <td>71534</td>\n",
       "      <td>chrMito</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>P2G09</td>\n",
       "      <td>73.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-23.671064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1335380 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        altered allete reference allele      gain  Position    CHROM REF ALT  \\\n",
       "0           CATTACCCTA       CATTACCCTG  0.955511       135     chrI   G   A   \n",
       "1           ATTACCCTAC       ATTACCCTGC  1.131402       135     chrI   G   A   \n",
       "2           TTACCCTACC       TTACCCTGCC  0.716678       135     chrI   G   A   \n",
       "3           TACCCTACCT       TACCCTGCCT  0.635989       135     chrI   G   A   \n",
       "4           ACCCTACCTC       ACCCTGCCTC  1.178655       135     chrI   G   A   \n",
       "...                ...              ...       ...       ...      ...  ..  ..   \n",
       "1335375     TAATCTATAT       TAATATATAT -2.197225     71534  chrMito   A   C   \n",
       "1335376     AATCTATATA       AATATATATA -2.297165     71534  chrMito   A   C   \n",
       "1335377     ATCTATATAT       ATATATATAT -3.315454     71534  chrMito   A   C   \n",
       "1335378     TCTATATATA       TATATATATA -3.143530     71534  chrMito   A   C   \n",
       "1335379     CTATATATAT       ATATATATAT -3.135072     71534  chrMito   A   C   \n",
       "\n",
       "        population_name  G70_ref_counts  G1410_ref_counts  ...  \\\n",
       "0                 P1C08           170.0              34.0  ...   \n",
       "1                 P1C08           170.0              34.0  ...   \n",
       "2                 P1C08           170.0              34.0  ...   \n",
       "3                 P1C08           170.0              34.0  ...   \n",
       "4                 P1C08           170.0              34.0  ...   \n",
       "...                 ...             ...               ...  ...   \n",
       "1335375           P2G09            73.0              19.0  ...   \n",
       "1335376           P2G09            73.0              19.0  ...   \n",
       "1335377           P2G09            73.0              19.0  ...   \n",
       "1335378           P2G09            73.0              19.0  ...   \n",
       "1335379           P2G09            73.0              19.0  ...   \n",
       "\n",
       "         G5150_ref_counts  G7530_ref_counts  G10150_ref_counts  \\\n",
       "0                    59.0             181.0               33.0   \n",
       "1                    59.0             181.0               33.0   \n",
       "2                    59.0             181.0               33.0   \n",
       "3                    59.0             181.0               33.0   \n",
       "4                    59.0             181.0               33.0   \n",
       "...                   ...               ...                ...   \n",
       "1335375              14.0              50.0               21.0   \n",
       "1335376              14.0              50.0               21.0   \n",
       "1335377              14.0              50.0               21.0   \n",
       "1335378              14.0              50.0               21.0   \n",
       "1335379              14.0              50.0               21.0   \n",
       "\n",
       "         G70_alt_counts  G1410_alt_counts  G2640_alt_counts  G5150_alt_counts  \\\n",
       "0                   0.0               5.0              14.0               9.0   \n",
       "1                   0.0               5.0              14.0               9.0   \n",
       "2                   0.0               5.0              14.0               9.0   \n",
       "3                   0.0               5.0              14.0               9.0   \n",
       "4                   0.0               5.0              14.0               9.0   \n",
       "...                 ...               ...               ...               ...   \n",
       "1335375             0.0               1.0               2.0               0.0   \n",
       "1335376             0.0               1.0               2.0               0.0   \n",
       "1335377             0.0               1.0               2.0               0.0   \n",
       "1335378             0.0               1.0               2.0               0.0   \n",
       "1335379             0.0               1.0               2.0               0.0   \n",
       "\n",
       "         G7530_alt_counts  G10150_alt_counts  accumulate_gain  \n",
       "0                     0.0                6.0         2.968131  \n",
       "1                     0.0                6.0         2.968131  \n",
       "2                     0.0                6.0         2.968131  \n",
       "3                     0.0                6.0         2.968131  \n",
       "4                     0.0                6.0         2.968131  \n",
       "...                   ...                ...              ...  \n",
       "1335375               3.0                0.0       -23.671064  \n",
       "1335376               3.0                0.0       -23.671064  \n",
       "1335377               3.0                0.0       -23.671064  \n",
       "1335378               3.0                0.0       -23.671064  \n",
       "1335379               3.0                0.0       -23.671064  \n",
       "\n",
       "[1335380 rows x 21 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_complete = pd.merge(df_complete, grouped_df_with_accumulated, on=key_columns, how='left')\n",
    "df_complete"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['altered allete', 'reference allele', 'gain', 'Position', 'CHROM',\n",
       "       'REF', 'ALT', 'population_name', 'G70_ref_counts', 'G1410_ref_counts',\n",
       "       'G2640_ref_counts', 'G5150_ref_counts', 'G7530_ref_counts',\n",
       "       'G10150_ref_counts', 'G70_alt_counts', 'G1410_alt_counts',\n",
       "       'G2640_alt_counts', 'G5150_alt_counts', 'G7530_alt_counts',\n",
       "       'G10150_alt_counts', 'accumulate_gain'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_complete.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>accumulate_gain</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Position</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>population_name</th>\n",
       "      <th>G70_ref_counts</th>\n",
       "      <th>G1410_ref_counts</th>\n",
       "      <th>G2640_ref_counts</th>\n",
       "      <th>G5150_ref_counts</th>\n",
       "      <th>G7530_ref_counts</th>\n",
       "      <th>G10150_ref_counts</th>\n",
       "      <th>G70_alt_counts</th>\n",
       "      <th>G1410_alt_counts</th>\n",
       "      <th>G2640_alt_counts</th>\n",
       "      <th>G5150_alt_counts</th>\n",
       "      <th>G7530_alt_counts</th>\n",
       "      <th>G10150_alt_counts</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <th>chrVIII</th>\n",
       "      <th>C</th>\n",
       "      <th>A</th>\n",
       "      <th>P3G06</th>\n",
       "      <th>35.0</th>\n",
       "      <th>31.0</th>\n",
       "      <th>59.0</th>\n",
       "      <th>38.0</th>\n",
       "      <th>42.0</th>\n",
       "      <th>56.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>11.0</th>\n",
       "      <th>0.0</th>\n",
       "      <td>-3.577119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">34</th>\n",
       "      <th>2-micron</th>\n",
       "      <th>G</th>\n",
       "      <th>T</th>\n",
       "      <th>P1E04</th>\n",
       "      <th>74.0</th>\n",
       "      <th>104.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>217.0</th>\n",
       "      <th>426.0</th>\n",
       "      <th>323.0</th>\n",
       "      <th>156.0</th>\n",
       "      <th>236.0</th>\n",
       "      <td>0.392871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrII</th>\n",
       "      <th>G</th>\n",
       "      <th>A</th>\n",
       "      <th>P2C06</th>\n",
       "      <th>4.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>16.0</th>\n",
       "      <th>12.0</th>\n",
       "      <td>-4.563850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">36</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chrIX</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">A</th>\n",
       "      <th>P2F11</th>\n",
       "      <th>24.0</th>\n",
       "      <th>5.0</th>\n",
       "      <th>39.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>34.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>5.0</th>\n",
       "      <th>0.0</th>\n",
       "      <td>-11.012777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3F11</th>\n",
       "      <th>44.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>38.0</th>\n",
       "      <th>4.0</th>\n",
       "      <th>18.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>9.0</th>\n",
       "      <th>0.0</th>\n",
       "      <td>-11.012777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">1525964</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chrIV</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">C</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">T</th>\n",
       "      <th>P3G10</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>5.0</th>\n",
       "      <th>12.0</th>\n",
       "      <th>7.0</th>\n",
       "      <th>18.0</th>\n",
       "      <th>0.0</th>\n",
       "      <td>8.835324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3G11</th>\n",
       "      <th>2.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>6.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>4.0</th>\n",
       "      <th>0.0</th>\n",
       "      <td>8.835324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">1525998</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">chrIV</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">T</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">C</th>\n",
       "      <th>P3B07</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>8.0</th>\n",
       "      <th>0.0</th>\n",
       "      <td>-14.496006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P3B10</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>4.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>3.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <td>-14.496006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526031</th>\n",
       "      <th>chrIV</th>\n",
       "      <th>A</th>\n",
       "      <th>G</th>\n",
       "      <th>P3D05</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>2.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>1.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>0.0</th>\n",
       "      <th>32.0</th>\n",
       "      <th>6.0</th>\n",
       "      <td>-5.155277</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133538 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                                                                                                                                                                                                     accumulate_gain\n",
       "Position CHROM    REF ALT population_name G70_ref_counts G1410_ref_counts G2640_ref_counts G5150_ref_counts G7530_ref_counts G10150_ref_counts G70_alt_counts G1410_alt_counts G2640_alt_counts G5150_alt_counts G7530_alt_counts G10150_alt_counts                 \n",
       "31       chrVIII  C   A   P3G06           35.0           31.0             59.0             38.0             42.0             56.0              0.0            0.0              0.0              0.0              11.0             0.0                      -3.577119\n",
       "34       2-micron G   T   P1E04           74.0           104.0            3.0              0.0              0.0              1.0               0.0            217.0            426.0            323.0            156.0            236.0                     0.392871\n",
       "         chrII    G   A   P2C06           4.0            0.0              1.0              0.0              2.0              1.0               0.0            0.0              0.0              0.0              16.0             12.0                     -4.563850\n",
       "36       chrIX    C   A   P2F11           24.0           5.0              39.0             2.0              34.0             2.0               0.0            0.0              0.0              0.0              5.0              0.0                     -11.012777\n",
       "                          P3F11           44.0           1.0              38.0             4.0              18.0             2.0               0.0            0.0              0.0              0.0              9.0              0.0                     -11.012777\n",
       "...                                                                                                                                                                                                                                                              ...\n",
       "1525964  chrIV    C   T   P3G10           0.0            0.0              0.0              1.0              2.0              0.0               3.0            5.0              12.0             7.0              18.0             0.0                       8.835324\n",
       "                          P3G11           2.0            0.0              1.0              1.0              0.0              0.0               0.0            0.0              6.0              0.0              4.0              0.0                       8.835324\n",
       "1525998  chrIV    T   C   P3B07           0.0            0.0              1.0              1.0              3.0              0.0               2.0            0.0              0.0              1.0              8.0              0.0                     -14.496006\n",
       "                          P3B10           0.0            0.0              0.0              4.0              2.0              1.0               0.0            0.0              3.0              2.0              0.0              0.0                     -14.496006\n",
       "1526031  chrIV    A   G   P3D05           0.0            0.0              0.0              2.0              0.0              1.0               0.0            0.0              0.0              0.0              32.0             6.0                      -5.155277\n",
       "\n",
       "[133538 rows x 1 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ag_data  = df_complete.groupby(['Position', 'CHROM',\n",
    "       'REF', 'ALT', 'population_name', 'G70_ref_counts', 'G1410_ref_counts',\n",
    "       'G2640_ref_counts', 'G5150_ref_counts', 'G7530_ref_counts',\n",
    "       'G10150_ref_counts', 'G70_alt_counts', 'G1410_alt_counts',\n",
    "       'G2640_alt_counts', 'G5150_alt_counts', 'G7530_alt_counts',\n",
    "       'G10150_alt_counts'])['accumulate_gain'].mean()\n",
    "ag_data_group_by = ag_data.reset_index()\n",
    "ag_data_group_by = pd.DataFrame(ag_data)\n",
    "ag_data_group_by"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_complete[df_complete.duplicated(key_columns)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>population_name</th>\n",
       "      <th>G70_ref_counts</th>\n",
       "      <th>G1410_ref_counts</th>\n",
       "      <th>G2640_ref_counts</th>\n",
       "      <th>G5150_ref_counts</th>\n",
       "      <th>G7530_ref_counts</th>\n",
       "      <th>G10150_ref_counts</th>\n",
       "      <th>G70_alt_counts</th>\n",
       "      <th>G1410_alt_counts</th>\n",
       "      <th>G2640_alt_counts</th>\n",
       "      <th>G5150_alt_counts</th>\n",
       "      <th>G7530_alt_counts</th>\n",
       "      <th>G10150_alt_counts</th>\n",
       "      <th>accumulate_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>31</td>\n",
       "      <td>chrVIII</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P3G06</td>\n",
       "      <td>35.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-3.577119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>34</td>\n",
       "      <td>2-micron</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>P1E04</td>\n",
       "      <td>74.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>217.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>323.0</td>\n",
       "      <td>156.0</td>\n",
       "      <td>236.0</td>\n",
       "      <td>0.392871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>34</td>\n",
       "      <td>chrII</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>P2C06</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>-4.563850</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>36</td>\n",
       "      <td>chrIX</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P2F11</td>\n",
       "      <td>24.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-11.012777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>36</td>\n",
       "      <td>chrIX</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>P3F11</td>\n",
       "      <td>44.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-11.012777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133533</th>\n",
       "      <td>1525964</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>P3G10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.835324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133534</th>\n",
       "      <td>1525964</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>P3G11</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.835324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133535</th>\n",
       "      <td>1525998</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P3B07</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-14.496006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133536</th>\n",
       "      <td>1525998</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>P3B10</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-14.496006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133537</th>\n",
       "      <td>1526031</td>\n",
       "      <td>chrIV</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>P3D05</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>-5.155277</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>133538 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Position     CHROM REF ALT population_name  G70_ref_counts  \\\n",
       "0             31   chrVIII   C   A           P3G06            35.0   \n",
       "1             34  2-micron   G   T           P1E04            74.0   \n",
       "2             34     chrII   G   A           P2C06             4.0   \n",
       "3             36     chrIX   C   A           P2F11            24.0   \n",
       "4             36     chrIX   C   A           P3F11            44.0   \n",
       "...          ...       ...  ..  ..             ...             ...   \n",
       "133533   1525964     chrIV   C   T           P3G10             0.0   \n",
       "133534   1525964     chrIV   C   T           P3G11             2.0   \n",
       "133535   1525998     chrIV   T   C           P3B07             0.0   \n",
       "133536   1525998     chrIV   T   C           P3B10             0.0   \n",
       "133537   1526031     chrIV   A   G           P3D05             0.0   \n",
       "\n",
       "        G1410_ref_counts  G2640_ref_counts  G5150_ref_counts  \\\n",
       "0                   31.0              59.0              38.0   \n",
       "1                  104.0               3.0               0.0   \n",
       "2                    0.0               1.0               0.0   \n",
       "3                    5.0              39.0               2.0   \n",
       "4                    1.0              38.0               4.0   \n",
       "...                  ...               ...               ...   \n",
       "133533               0.0               0.0               1.0   \n",
       "133534               0.0               1.0               1.0   \n",
       "133535               0.0               1.0               1.0   \n",
       "133536               0.0               0.0               4.0   \n",
       "133537               0.0               0.0               2.0   \n",
       "\n",
       "        G7530_ref_counts  G10150_ref_counts  G70_alt_counts  G1410_alt_counts  \\\n",
       "0                   42.0               56.0             0.0               0.0   \n",
       "1                    0.0                1.0             0.0             217.0   \n",
       "2                    2.0                1.0             0.0               0.0   \n",
       "3                   34.0                2.0             0.0               0.0   \n",
       "4                   18.0                2.0             0.0               0.0   \n",
       "...                  ...                ...             ...               ...   \n",
       "133533               2.0                0.0             3.0               5.0   \n",
       "133534               0.0                0.0             0.0               0.0   \n",
       "133535               3.0                0.0             2.0               0.0   \n",
       "133536               2.0                1.0             0.0               0.0   \n",
       "133537               0.0                1.0             0.0               0.0   \n",
       "\n",
       "        G2640_alt_counts  G5150_alt_counts  G7530_alt_counts  \\\n",
       "0                    0.0               0.0              11.0   \n",
       "1                  426.0             323.0             156.0   \n",
       "2                    0.0               0.0              16.0   \n",
       "3                    0.0               0.0               5.0   \n",
       "4                    0.0               0.0               9.0   \n",
       "...                  ...               ...               ...   \n",
       "133533              12.0               7.0              18.0   \n",
       "133534               6.0               0.0               4.0   \n",
       "133535               0.0               1.0               8.0   \n",
       "133536               3.0               2.0               0.0   \n",
       "133537               0.0               0.0              32.0   \n",
       "\n",
       "        G10150_alt_counts  accumulate_gain  \n",
       "0                     0.0        -3.577119  \n",
       "1                   236.0         0.392871  \n",
       "2                    12.0        -4.563850  \n",
       "3                     0.0       -11.012777  \n",
       "4                     0.0       -11.012777  \n",
       "...                   ...              ...  \n",
       "133533                0.0         8.835324  \n",
       "133534                0.0         8.835324  \n",
       "133535                0.0       -14.496006  \n",
       "133536                0.0       -14.496006  \n",
       "133537                6.0        -5.155277  \n",
       "\n",
       "[133538 rows x 18 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ag_data_group_by.reset_index(inplace=True)\n",
    "ag_data_group_by"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ag_data_group_by.to_csv(\"Gain_calulated_unfiltered_vcf.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "amr",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
