{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ESM1b analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 1) The E.coli LTEE data was retrived from the authors github page -> https://github.com/benjaminhgood/LTEE-metagenomic\n",
    "#### 2) The path of the folder is data_files , from here the all population annotated_timecourse text was taken and a masterhseet was prepared with only missense mutations.\n",
    "#### 3) Inorder to process this file for ESM1b analysis we need all variants WT protien sequence. We retrived the DNA and protein sequence from NCBI --> https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/017/985/GCF_000017985.1_ASM1798v1/ . \n",
    "#### 4) The NCBI files such as cds_genomics.fna and translate_cds.faa along with feature table.txt was important for this analysis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To retrive the DNA sequence for each variants, we need to do the following steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                col1  \\\n",
      "0  >lcl|CP000819.1_cds_ACT37694.1_1 [gene=thrL] [...   \n",
      "1  >lcl|CP000819.1_cds_ACT37695.1_2 [gene=thrA] [...   \n",
      "2  >lcl|CP000819.1_cds_ACT37696.1_3 [gene=thrB] [...   \n",
      "3  >lcl|CP000819.1_cds_ACT37697.1_4 [gene=thrC] [...   \n",
      "4  >lcl|CP000819.1_cds_ACT37698.1_5 [gene=yaaX] [...   \n",
      "\n",
      "                                                col2  \n",
      "0  ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...  \n",
      "1  ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...  \n",
      "2  ATGGTTAAAGTTTATGCCCCGGCTTCCAGTGCCAATATGAGCGTCG...  \n",
      "3  ATGAAACTCTACAATCTGAAAGATCACAATGAGCAGGTCAGCTTTG...  \n",
      "4  ATGAAAAAGATGCAATCTATCGTACTCGCACTTTCCCTGGTTCTGG...  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Step 1: Open and read the text file\n",
    "file_path = \"/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/NCBI_data/GCA_000017985.1_ASM1798v1_cds_from_genomic.fna\"  # Replace with your file path\n",
    "data = []\n",
    "\n",
    "try:\n",
    "    with open(file_path, 'r') as file:\n",
    "        header = None\n",
    "        sequence = []\n",
    "\n",
    "        for line in file:\n",
    "            if line.startswith(\">\"):\n",
    "                # If a new header is encountered, save the previous sequence (if any)\n",
    "                if header is not None:\n",
    "                    data.append([header, ''.join(sequence)])\n",
    "                header = line.strip()\n",
    "                sequence = []\n",
    "            else:\n",
    "                sequence.append(line.strip())\n",
    "\n",
    "        # Add the last sequence (if any)\n",
    "        if header is not None:\n",
    "            data.append([header, ''.join(sequence)])\n",
    "\n",
    "    # Step 2: Create a DataFrame\n",
    "    df = pd.DataFrame(data, columns=['col1', 'col2'])\n",
    "\n",
    "    # Step 3: Display the DataFrame\n",
    "    print(df.head())\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"File '{file_path}' not found.\")\n",
    "except IOError:\n",
    "    print(f\"An error occurred while reading the file '{file_path}'.\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                   col1  \\\n",
      "0     >lcl|CP000819.1_cds_ACT37694.1_1 [gene=thrL] [...   \n",
      "1     >lcl|CP000819.1_cds_ACT37695.1_2 [gene=thrA] [...   \n",
      "2     >lcl|CP000819.1_cds_ACT37696.1_3 [gene=thrB] [...   \n",
      "3     >lcl|CP000819.1_cds_ACT37697.1_4 [gene=thrC] [...   \n",
      "4     >lcl|CP000819.1_cds_ACT37698.1_5 [gene=yaaX] [...   \n",
      "...                                                 ...   \n",
      "4204  >lcl|CP000819.1_cds_ACT41898.1_4205 [gene=creB...   \n",
      "4205  >lcl|CP000819.1_cds_ACT41899.1_4206 [gene=creC...   \n",
      "4206  >lcl|CP000819.1_cds_ACT41900.1_4207 [gene=arcA...   \n",
      "4207  >lcl|CP000819.1_cds_ACT41901.1_4208 [gene=yjjY...   \n",
      "4208  >lcl|CP000819.1_cds_ACT41902.1_4209 [gene=lasT...   \n",
      "\n",
      "                                                   col2  \\\n",
      "0     ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...   \n",
      "1     ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...   \n",
      "2     ATGGTTAAAGTTTATGCCCCGGCTTCCAGTGCCAATATGAGCGTCG...   \n",
      "3     ATGAAACTCTACAATCTGAAAGATCACAATGAGCAGGTCAGCTTTG...   \n",
      "4     ATGAAAAAGATGCAATCTATCGTACTCGCACTTTCCCTGGTTCTGG...   \n",
      "...                                                 ...   \n",
      "4204  ATGCAACGGGAAACGGTCTGGTTAGTGGAAGATGAGCAAGGGATAG...   \n",
      "4205  ATGCGTATCGGCATGCGGTTGCTGCTGGGCTATTTTTTACTGGTGG...   \n",
      "4206  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
      "4207  ATGACTAAAGTACGTAATTGCGTTCTTGATGCACTTTCCATCAACG...   \n",
      "4208  ATGCGTATAACGATTATTCTGGTCGCACCCGCCAGAGCAGAAAATA...   \n",
      "\n",
      "                               Attribute_0 Attribute_1 Attribute_2  \\\n",
      "0        >lcl|CP000819.1_cds_ACT37694.1_1    gene=thrL               \n",
      "1        >lcl|CP000819.1_cds_ACT37695.1_2    gene=thrA               \n",
      "2        >lcl|CP000819.1_cds_ACT37696.1_3    gene=thrB               \n",
      "3        >lcl|CP000819.1_cds_ACT37697.1_4    gene=thrC               \n",
      "4        >lcl|CP000819.1_cds_ACT37698.1_5    gene=yaaX               \n",
      "...                                    ...         ...         ...   \n",
      "4204  >lcl|CP000819.1_cds_ACT41898.1_4205    gene=creB               \n",
      "4205  >lcl|CP000819.1_cds_ACT41899.1_4206    gene=creC               \n",
      "4206  >lcl|CP000819.1_cds_ACT41900.1_4207    gene=arcA               \n",
      "4207  >lcl|CP000819.1_cds_ACT41901.1_4208    gene=yjjY               \n",
      "4208  >lcl|CP000819.1_cds_ACT41902.1_4209    gene=lasT               \n",
      "\n",
      "              Attribute_3 Attribute_4  \\\n",
      "0     locus_tag=ECB_00001               \n",
      "1     locus_tag=ECB_00002               \n",
      "2     locus_tag=ECB_00003               \n",
      "3     locus_tag=ECB_00004               \n",
      "4     locus_tag=ECB_00005               \n",
      "...                   ...         ...   \n",
      "4204  locus_tag=ECB_04274               \n",
      "4205  locus_tag=ECB_04275               \n",
      "4206  locus_tag=ECB_04277               \n",
      "4207  locus_tag=ECB_04278               \n",
      "4208  locus_tag=ECB_04279               \n",
      "\n",
      "                                            Attribute_5 Attribute_6  \\\n",
      "0                     protein=thr operon leader peptide               \n",
      "1     protein=bifunctional aspartokinase I/homeserin...               \n",
      "2                             protein=homoserine kinase               \n",
      "3                            protein=threonine synthase               \n",
      "4                          protein=hypothetical protein               \n",
      "...                                                 ...         ...   \n",
      "4204  protein=DNA-binding response regulator in two-...               \n",
      "4205  protein=sensory histidine kinase in two-compon...               \n",
      "4206  protein=DNA-binding response regulator in two-...               \n",
      "4207                       protein=hypothetical protein               \n",
      "4208           protein=predicted rRNA methyltransferase               \n",
      "\n",
      "                Attribute_7 Attribute_8  \\\n",
      "0     protein_id=ACT37694.1               \n",
      "1     protein_id=ACT37695.1               \n",
      "2     protein_id=ACT37696.1               \n",
      "3     protein_id=ACT37697.1               \n",
      "4     protein_id=ACT37698.1               \n",
      "...                     ...         ...   \n",
      "4204  protein_id=ACT41898.1               \n",
      "4205  protein_id=ACT41899.1               \n",
      "4206  protein_id=ACT41900.1               \n",
      "4207  protein_id=ACT41901.1               \n",
      "4208  protein_id=ACT41902.1               \n",
      "\n",
      "                                Attribute_9 Attribute_10 Attribute_11  \\\n",
      "0                         location=190..255                 gbkey=CDS   \n",
      "1                        location=336..2798                 gbkey=CDS   \n",
      "2                       location=2800..3732                 gbkey=CDS   \n",
      "3                       location=3733..5019                 gbkey=CDS   \n",
      "4                       location=5232..5528                 gbkey=CDS   \n",
      "...                                     ...          ...          ...   \n",
      "4204              location=4624167..4624856                 gbkey=CDS   \n",
      "4205              location=4624856..4626280                 gbkey=CDS   \n",
      "4206  location=complement(4627750..4628466)                 gbkey=CDS   \n",
      "4207              location=4628562..4628702                 gbkey=CDS   \n",
      "4208              location=4629102..4629788                 gbkey=CDS   \n",
      "\n",
      "     Attribute_12 Attribute_13 Attribute_14  \n",
      "0                         None         None  \n",
      "1                         None         None  \n",
      "2                         None         None  \n",
      "3                         None         None  \n",
      "4                         None         None  \n",
      "...           ...          ...          ...  \n",
      "4204                      None         None  \n",
      "4205                      None         None  \n",
      "4206                      None         None  \n",
      "4207                      None         None  \n",
      "4208                      None         None  \n",
      "\n",
      "[4209 rows x 17 columns]\n"
     ]
    }
   ],
   "source": [
    "# Process the first row to extract cols \n",
    "import pandas as pd\n",
    "\n",
    "# Define a regular expression pattern to match attributes in square brackets\n",
    "pattern = r'\\[([^]]+)\\]'\n",
    "\n",
    "# Split the \"col2\" column by the attribute pattern and expand it into multiple columns\n",
    "split_columns = df['col1'].str.split(pattern, expand=True)\n",
    "\n",
    "# Rename the new columns for clarity (optional)\n",
    "split_columns.columns = [f'Attribute_{i}' for i in range(len(split_columns.columns))]\n",
    "\n",
    "# Concatenate the split_columns DataFrame with the original DataFrame\n",
    "result_df = pd.concat([df, split_columns], axis=1)\n",
    "\n",
    "# Drop the original \"col2\" column if needed\n",
    "# result_df = result_df.drop(columns='col2')\n",
    "\n",
    "# Print the result\n",
    "print(result_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename columns\n",
    "# Assuming you want to rename \"col2\" to \"new_column_name\"\n",
    "result_df.columns = ['Info' if col == 'col1' else col for col in result_df.columns]\n",
    "# # Assuming you want to rename \"col2\" to \"new_column_name\"\n",
    "result_df.columns = ['DNA_Seq' if col == 'col2' else col for col in result_df.columns]\n",
    "result_df.columns = ['Function' if col == 'Attribute_5' else col for col in result_df.columns]\n",
    "result_df.columns = ['protein_id' if col == 'Attribute_7' else col for col in result_df.columns]\n",
    "result_df.columns = ['Location' if col == 'Attribute_9' else col for col in result_df.columns]\n",
    "result_df.columns = ['locus_tag' if col == 'Attribute_3' else col for col in result_df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split \"col2\" based on \"=\" and create new columns C and D\n",
    "result_df[['C', 'D']] = result_df['Attribute_1'].str.split('=', n=1, expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.columns = ['Gene' if col == 'D' else col for col in result_df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                   Info  \\\n",
      "0     >lcl|CP000819.1_cds_ACT37694.1_1 [gene=thrL] [...   \n",
      "1     >lcl|CP000819.1_cds_ACT37695.1_2 [gene=thrA] [...   \n",
      "2     >lcl|CP000819.1_cds_ACT37696.1_3 [gene=thrB] [...   \n",
      "3     >lcl|CP000819.1_cds_ACT37697.1_4 [gene=thrC] [...   \n",
      "4     >lcl|CP000819.1_cds_ACT37698.1_5 [gene=yaaX] [...   \n",
      "...                                                 ...   \n",
      "4204  >lcl|CP000819.1_cds_ACT41898.1_4205 [gene=creB...   \n",
      "4205  >lcl|CP000819.1_cds_ACT41899.1_4206 [gene=creC...   \n",
      "4206  >lcl|CP000819.1_cds_ACT41900.1_4207 [gene=arcA...   \n",
      "4207  >lcl|CP000819.1_cds_ACT41901.1_4208 [gene=yjjY...   \n",
      "4208  >lcl|CP000819.1_cds_ACT41902.1_4209 [gene=lasT...   \n",
      "\n",
      "                                                DNA_Seq  \\\n",
      "0     ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...   \n",
      "1     ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...   \n",
      "2     ATGGTTAAAGTTTATGCCCCGGCTTCCAGTGCCAATATGAGCGTCG...   \n",
      "3     ATGAAACTCTACAATCTGAAAGATCACAATGAGCAGGTCAGCTTTG...   \n",
      "4     ATGAAAAAGATGCAATCTATCGTACTCGCACTTTCCCTGGTTCTGG...   \n",
      "...                                                 ...   \n",
      "4204  ATGCAACGGGAAACGGTCTGGTTAGTGGAAGATGAGCAAGGGATAG...   \n",
      "4205  ATGCGTATCGGCATGCGGTTGCTGCTGGGCTATTTTTTACTGGTGG...   \n",
      "4206  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
      "4207  ATGACTAAAGTACGTAATTGCGTTCTTGATGCACTTTCCATCAACG...   \n",
      "4208  ATGCGTATAACGATTATTCTGGTCGCACCCGCCAGAGCAGAAAATA...   \n",
      "\n",
      "                               Attribute_0 Attribute_1 Attribute_2  \\\n",
      "0        >lcl|CP000819.1_cds_ACT37694.1_1    gene=thrL               \n",
      "1        >lcl|CP000819.1_cds_ACT37695.1_2    gene=thrA               \n",
      "2        >lcl|CP000819.1_cds_ACT37696.1_3    gene=thrB               \n",
      "3        >lcl|CP000819.1_cds_ACT37697.1_4    gene=thrC               \n",
      "4        >lcl|CP000819.1_cds_ACT37698.1_5    gene=yaaX               \n",
      "...                                    ...         ...         ...   \n",
      "4204  >lcl|CP000819.1_cds_ACT41898.1_4205    gene=creB               \n",
      "4205  >lcl|CP000819.1_cds_ACT41899.1_4206    gene=creC               \n",
      "4206  >lcl|CP000819.1_cds_ACT41900.1_4207    gene=arcA               \n",
      "4207  >lcl|CP000819.1_cds_ACT41901.1_4208    gene=yjjY               \n",
      "4208  >lcl|CP000819.1_cds_ACT41902.1_4209    gene=lasT               \n",
      "\n",
      "                locus_tag Attribute_4  \\\n",
      "0     locus_tag=ECB_00001               \n",
      "1     locus_tag=ECB_00002               \n",
      "2     locus_tag=ECB_00003               \n",
      "3     locus_tag=ECB_00004               \n",
      "4     locus_tag=ECB_00005               \n",
      "...                   ...         ...   \n",
      "4204  locus_tag=ECB_04274               \n",
      "4205  locus_tag=ECB_04275               \n",
      "4206  locus_tag=ECB_04277               \n",
      "4207  locus_tag=ECB_04278               \n",
      "4208  locus_tag=ECB_04279               \n",
      "\n",
      "                                               Function Attribute_6  \\\n",
      "0                     protein=thr operon leader peptide               \n",
      "1     protein=bifunctional aspartokinase I/homeserin...               \n",
      "2                             protein=homoserine kinase               \n",
      "3                            protein=threonine synthase               \n",
      "4                          protein=hypothetical protein               \n",
      "...                                                 ...         ...   \n",
      "4204  protein=DNA-binding response regulator in two-...               \n",
      "4205  protein=sensory histidine kinase in two-compon...               \n",
      "4206  protein=DNA-binding response regulator in two-...               \n",
      "4207                       protein=hypothetical protein               \n",
      "4208           protein=predicted rRNA methyltransferase               \n",
      "\n",
      "                 protein_id Attribute_8  \\\n",
      "0     protein_id=ACT37694.1               \n",
      "1     protein_id=ACT37695.1               \n",
      "2     protein_id=ACT37696.1               \n",
      "3     protein_id=ACT37697.1               \n",
      "4     protein_id=ACT37698.1               \n",
      "...                     ...         ...   \n",
      "4204  protein_id=ACT41898.1               \n",
      "4205  protein_id=ACT41899.1               \n",
      "4206  protein_id=ACT41900.1               \n",
      "4207  protein_id=ACT41901.1               \n",
      "4208  protein_id=ACT41902.1               \n",
      "\n",
      "                                   Location Attribute_10 Attribute_11  \\\n",
      "0                         location=190..255                 gbkey=CDS   \n",
      "1                        location=336..2798                 gbkey=CDS   \n",
      "2                       location=2800..3732                 gbkey=CDS   \n",
      "3                       location=3733..5019                 gbkey=CDS   \n",
      "4                       location=5232..5528                 gbkey=CDS   \n",
      "...                                     ...          ...          ...   \n",
      "4204              location=4624167..4624856                 gbkey=CDS   \n",
      "4205              location=4624856..4626280                 gbkey=CDS   \n",
      "4206  location=complement(4627750..4628466)                 gbkey=CDS   \n",
      "4207              location=4628562..4628702                 gbkey=CDS   \n",
      "4208              location=4629102..4629788                 gbkey=CDS   \n",
      "\n",
      "     Attribute_12 Attribute_13 Attribute_14     C  Gene contains_space  \n",
      "0                         None         None  gene  thrL             No  \n",
      "1                         None         None  gene  thrA             No  \n",
      "2                         None         None  gene  thrB             No  \n",
      "3                         None         None  gene  thrC             No  \n",
      "4                         None         None  gene  yaaX             No  \n",
      "...           ...          ...          ...   ...   ...            ...  \n",
      "4204                      None         None  gene  creB             No  \n",
      "4205                      None         None  gene  creC             No  \n",
      "4206                      None         None  gene  arcA             No  \n",
      "4207                      None         None  gene  yjjY             No  \n",
      "4208                      None         None  gene  lasT             No  \n",
      "\n",
      "[4209 rows x 20 columns]\n"
     ]
    }
   ],
   "source": [
    "# Check for spaces between characters in each row of \"col2\"\n",
    "result_df['contains_space'] = result_df['Gene'].str.contains(r'\\s')\n",
    "\n",
    "# Replace True with \"Yes\" and False with \"No\" in the \"contains_\n",
    "# \" column\n",
    "result_df['contains_space'] = result_df['contains_space'].replace({True: 'Yes', False: 'No'})\n",
    "\n",
    "# Print the DataFrame\n",
    "print(result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['No'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if any of the character contains spaces \n",
    "result_df['contains_space'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Info', 'DNA_Seq', 'Attribute_0', 'Attribute_1', 'Attribute_2',\n",
       "       'locus_tag', 'Attribute_4', 'Function', 'Attribute_6', 'protein_id',\n",
       "       'Attribute_8', 'Location', 'Attribute_10', 'Attribute_11',\n",
       "       'Attribute_12', 'Attribute_13', 'Attribute_14', 'C', 'Gene',\n",
       "       'contains_space'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the columns you want to keep\n",
    "DNA = result_df[['Info', 'DNA_Seq', 'Attribute_0', 'Attribute_1', 'Attribute_2',\n",
    "       'locus_tag', 'Attribute_4', 'Function', 'Attribute_6', 'protein_id',\n",
    "       'Attribute_8', 'Location', 'Attribute_10', 'Attribute_11',\n",
    "       'Attribute_12', 'Attribute_13', 'Attribute_14', 'C', 'Gene',]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame saved to /home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/DNA_seq.xlsx\n"
     ]
    }
   ],
   "source": [
    "# Write in local \n",
    "# Specify the file path where you want to save the Excel file\n",
    "excel_file_path = '/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_seq.xlsx'\n",
    "\n",
    "# Write the DataFrame to an Excel file\n",
    "DNA.to_excel(excel_file_path, index=False)\n",
    "\n",
    "print(f'DataFrame saved to {excel_file_path}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### To retrive the Protein sequence for each variants, we need to do the following steps."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                col1  \\\n",
      "0  >lcl|CP000819.1_prot_ACT37694.1_1 [gene=thrL] ...   \n",
      "1  >lcl|CP000819.1_prot_ACT37695.1_2 [gene=thrA] ...   \n",
      "2  >lcl|CP000819.1_prot_ACT37696.1_3 [gene=thrB] ...   \n",
      "3  >lcl|CP000819.1_prot_ACT37697.1_4 [gene=thrC] ...   \n",
      "4  >lcl|CP000819.1_prot_ACT37698.1_5 [gene=yaaX] ...   \n",
      "\n",
      "                                                col2  \n",
      "0                              MKRISTTITTTITITTGNGAG  \n",
      "1  MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  \n",
      "2  MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...  \n",
      "3  MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...  \n",
      "4  MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYW...  \n"
     ]
    }
   ],
   "source": [
    "# Read the prt file from ncbi \n",
    "import pandas as pd\n",
    "\n",
    "# Step 1: Open and read the text file\n",
    "file_path = \"/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/NCBI_data/GCA_000017985.1_ASM1798v1_translated_cds.faa\"  # Replace with your file path\n",
    "data = []\n",
    "\n",
    "try:\n",
    "    with open(file_path, 'r') as file:\n",
    "        header = None\n",
    "        sequence = []\n",
    "\n",
    "        for line in file:\n",
    "            if line.startswith(\">\"):\n",
    "                # If a new header is encountered, save the previous sequence (if any)\n",
    "                if header is not None:\n",
    "                    data.append([header, ''.join(sequence)])\n",
    "                header = line.strip()\n",
    "                sequence = []\n",
    "            else:\n",
    "                sequence.append(line.strip())\n",
    "\n",
    "        # Add the last sequence (if any)\n",
    "        if header is not None:\n",
    "            data.append([header, ''.join(sequence)])\n",
    "\n",
    "    # Step 2: Create a DataFrame\n",
    "    df = pd.DataFrame(data, columns=['col1', 'col2'])\n",
    "\n",
    "    # Step 3: Display the DataFrame\n",
    "    print(df.head())\n",
    "\n",
    "except FileNotFoundError:\n",
    "    print(f\"File '{file_path}' not found.\")\n",
    "except IOError:\n",
    "    print(f\"An error occurred while reading the file '{file_path}'.\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                   col1  \\\n",
      "0     >lcl|CP000819.1_prot_ACT37694.1_1 [gene=thrL] ...   \n",
      "1     >lcl|CP000819.1_prot_ACT37695.1_2 [gene=thrA] ...   \n",
      "2     >lcl|CP000819.1_prot_ACT37696.1_3 [gene=thrB] ...   \n",
      "3     >lcl|CP000819.1_prot_ACT37697.1_4 [gene=thrC] ...   \n",
      "4     >lcl|CP000819.1_prot_ACT37698.1_5 [gene=yaaX] ...   \n",
      "...                                                 ...   \n",
      "4204  >lcl|CP000819.1_prot_ACT41898.1_4205 [gene=cre...   \n",
      "4205  >lcl|CP000819.1_prot_ACT41899.1_4206 [gene=cre...   \n",
      "4206  >lcl|CP000819.1_prot_ACT41900.1_4207 [gene=arc...   \n",
      "4207  >lcl|CP000819.1_prot_ACT41901.1_4208 [gene=yjj...   \n",
      "4208  >lcl|CP000819.1_prot_ACT41902.1_4209 [gene=las...   \n",
      "\n",
      "                                                   col2  \\\n",
      "0                                 MKRISTTITTTITITTGNGAG   \n",
      "1     MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
      "2     MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...   \n",
      "3     MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...   \n",
      "4     MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYW...   \n",
      "...                                                 ...   \n",
      "4204  MQRETVWLVEDEQGIADTLVYMLQQEGFAVEVFERGLPVLDKARQQ...   \n",
      "4205  MRIGMRLLLGYFLLVAVAAWFVLAIFVKEVKPGVRRATEGTLIDTA...   \n",
      "4206  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
      "4207     MTKVRNCVLDALSINVNNIISLVVGTFPQDPTVSKTAVILTILTAT   \n",
      "4208  MRITIILVAPARAENIGAAARAMKTMGFSELRIVDSQAHLEPATRW...   \n",
      "\n",
      "                                Attribute_0 Attribute_1 Attribute_2  \\\n",
      "0        >lcl|CP000819.1_prot_ACT37694.1_1    gene=thrL               \n",
      "1        >lcl|CP000819.1_prot_ACT37695.1_2    gene=thrA               \n",
      "2        >lcl|CP000819.1_prot_ACT37696.1_3    gene=thrB               \n",
      "3        >lcl|CP000819.1_prot_ACT37697.1_4    gene=thrC               \n",
      "4        >lcl|CP000819.1_prot_ACT37698.1_5    gene=yaaX               \n",
      "...                                     ...         ...         ...   \n",
      "4204  >lcl|CP000819.1_prot_ACT41898.1_4205    gene=creB               \n",
      "4205  >lcl|CP000819.1_prot_ACT41899.1_4206    gene=creC               \n",
      "4206  >lcl|CP000819.1_prot_ACT41900.1_4207    gene=arcA               \n",
      "4207  >lcl|CP000819.1_prot_ACT41901.1_4208    gene=yjjY               \n",
      "4208  >lcl|CP000819.1_prot_ACT41902.1_4209    gene=lasT               \n",
      "\n",
      "              Attribute_3 Attribute_4  \\\n",
      "0     locus_tag=ECB_00001               \n",
      "1     locus_tag=ECB_00002               \n",
      "2     locus_tag=ECB_00003               \n",
      "3     locus_tag=ECB_00004               \n",
      "4     locus_tag=ECB_00005               \n",
      "...                   ...         ...   \n",
      "4204  locus_tag=ECB_04274               \n",
      "4205  locus_tag=ECB_04275               \n",
      "4206  locus_tag=ECB_04277               \n",
      "4207  locus_tag=ECB_04278               \n",
      "4208  locus_tag=ECB_04279               \n",
      "\n",
      "                                            Attribute_5 Attribute_6  \\\n",
      "0                     protein=thr operon leader peptide               \n",
      "1     protein=bifunctional aspartokinase I/homeserin...               \n",
      "2                             protein=homoserine kinase               \n",
      "3                            protein=threonine synthase               \n",
      "4                          protein=hypothetical protein               \n",
      "...                                                 ...         ...   \n",
      "4204  protein=DNA-binding response regulator in two-...               \n",
      "4205  protein=sensory histidine kinase in two-compon...               \n",
      "4206  protein=DNA-binding response regulator in two-...               \n",
      "4207                       protein=hypothetical protein               \n",
      "4208           protein=predicted rRNA methyltransferase               \n",
      "\n",
      "                Attribute_7 Attribute_8  \\\n",
      "0     protein_id=ACT37694.1               \n",
      "1     protein_id=ACT37695.1               \n",
      "2     protein_id=ACT37696.1               \n",
      "3     protein_id=ACT37697.1               \n",
      "4     protein_id=ACT37698.1               \n",
      "...                     ...         ...   \n",
      "4204  protein_id=ACT41898.1               \n",
      "4205  protein_id=ACT41899.1               \n",
      "4206  protein_id=ACT41900.1               \n",
      "4207  protein_id=ACT41901.1               \n",
      "4208  protein_id=ACT41902.1               \n",
      "\n",
      "                                Attribute_9 Attribute_10 Attribute_11  \\\n",
      "0                         location=190..255                 gbkey=CDS   \n",
      "1                        location=336..2798                 gbkey=CDS   \n",
      "2                       location=2800..3732                 gbkey=CDS   \n",
      "3                       location=3733..5019                 gbkey=CDS   \n",
      "4                       location=5232..5528                 gbkey=CDS   \n",
      "...                                     ...          ...          ...   \n",
      "4204              location=4624167..4624856                 gbkey=CDS   \n",
      "4205              location=4624856..4626280                 gbkey=CDS   \n",
      "4206  location=complement(4627750..4628466)                 gbkey=CDS   \n",
      "4207              location=4628562..4628702                 gbkey=CDS   \n",
      "4208              location=4629102..4629788                 gbkey=CDS   \n",
      "\n",
      "     Attribute_12 Attribute_13 Attribute_14  \n",
      "0                         None         None  \n",
      "1                         None         None  \n",
      "2                         None         None  \n",
      "3                         None         None  \n",
      "4                         None         None  \n",
      "...           ...          ...          ...  \n",
      "4204                      None         None  \n",
      "4205                      None         None  \n",
      "4206                      None         None  \n",
      "4207                      None         None  \n",
      "4208                      None         None  \n",
      "\n",
      "[4209 rows x 17 columns]\n"
     ]
    }
   ],
   "source": [
    "# process the rows \n",
    "import pandas as pd\n",
    "\n",
    "# Define a regular expression pattern to match attributes in square brackets\n",
    "pattern = r'\\[([^]]+)\\]'\n",
    "\n",
    "# Split the \"col2\" column by the attribute pattern and expand it into multiple columns\n",
    "split_columns = df['col1'].str.split(pattern, expand=True)\n",
    "\n",
    "# Rename the new columns for clarity (optional)\n",
    "split_columns.columns = [f'Attribute_{i}' for i in range(len(split_columns.columns))]\n",
    "\n",
    "# Concatenate the split_columns DataFrame with the original DataFrame\n",
    "result_df = pd.concat([df, split_columns], axis=1)\n",
    "\n",
    "# Drop the original \"col2\" column if needed\n",
    "# result_df = result_df.drop(columns='col2')\n",
    "\n",
    "# Print the result\n",
    "print(result_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename the rows \n",
    "# Assuming you want to rename \"col2\" to \"new_column_name\"\n",
    "result_df.columns = ['Info' if col == 'col1' else col for col in result_df.columns]\n",
    "# # Assuming you want to rename \"col2\" to \"new_column_name\"\n",
    "result_df.columns = ['Prt_Seq' if col == 'col2' else col for col in result_df.columns]\n",
    "result_df.columns = ['Function' if col == 'Attribute_5' else col for col in result_df.columns]\n",
    "result_df.columns = ['protein_id' if col == 'Attribute_7' else col for col in result_df.columns]\n",
    "result_df.columns = ['Location' if col == 'Attribute_9' else col for col in result_df.columns]\n",
    "result_df.columns = ['locus_tag' if col == 'Attribute_3' else col for col in result_df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Split \"col2\" based on \"=\" and create new columns C and D\n",
    "result_df[['C', 'D']] = result_df['Attribute_1'].str.split('=', n=1, expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.columns = ['Gene' if col == 'D' else col for col in result_df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                   Info  \\\n",
      "0     >lcl|CP000819.1_prot_ACT37694.1_1 [gene=thrL] ...   \n",
      "1     >lcl|CP000819.1_prot_ACT37695.1_2 [gene=thrA] ...   \n",
      "2     >lcl|CP000819.1_prot_ACT37696.1_3 [gene=thrB] ...   \n",
      "3     >lcl|CP000819.1_prot_ACT37697.1_4 [gene=thrC] ...   \n",
      "4     >lcl|CP000819.1_prot_ACT37698.1_5 [gene=yaaX] ...   \n",
      "...                                                 ...   \n",
      "4204  >lcl|CP000819.1_prot_ACT41898.1_4205 [gene=cre...   \n",
      "4205  >lcl|CP000819.1_prot_ACT41899.1_4206 [gene=cre...   \n",
      "4206  >lcl|CP000819.1_prot_ACT41900.1_4207 [gene=arc...   \n",
      "4207  >lcl|CP000819.1_prot_ACT41901.1_4208 [gene=yjj...   \n",
      "4208  >lcl|CP000819.1_prot_ACT41902.1_4209 [gene=las...   \n",
      "\n",
      "                                                Prt_Seq  \\\n",
      "0                                 MKRISTTITTTITITTGNGAG   \n",
      "1     MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
      "2     MVKVYAPASSANMSVGFDVLGAAVTPVDGALLGDVVTVEAAETFSL...   \n",
      "3     MKLYNLKDHNEQVSFAQAVTQGLGKNQGLFFPHDLPEFSLTEIDEM...   \n",
      "4     MKKMQSIVLALSLVLVAPMAAQAAEITLVPSVKLQIGDRDNRGYYW...   \n",
      "...                                                 ...   \n",
      "4204  MQRETVWLVEDEQGIADTLVYMLQQEGFAVEVFERGLPVLDKARQQ...   \n",
      "4205  MRIGMRLLLGYFLLVAVAAWFVLAIFVKEVKPGVRRATEGTLIDTA...   \n",
      "4206  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
      "4207     MTKVRNCVLDALSINVNNIISLVVGTFPQDPTVSKTAVILTILTAT   \n",
      "4208  MRITIILVAPARAENIGAAARAMKTMGFSELRIVDSQAHLEPATRW...   \n",
      "\n",
      "                                Attribute_0 Attribute_1 Attribute_2  \\\n",
      "0        >lcl|CP000819.1_prot_ACT37694.1_1    gene=thrL               \n",
      "1        >lcl|CP000819.1_prot_ACT37695.1_2    gene=thrA               \n",
      "2        >lcl|CP000819.1_prot_ACT37696.1_3    gene=thrB               \n",
      "3        >lcl|CP000819.1_prot_ACT37697.1_4    gene=thrC               \n",
      "4        >lcl|CP000819.1_prot_ACT37698.1_5    gene=yaaX               \n",
      "...                                     ...         ...         ...   \n",
      "4204  >lcl|CP000819.1_prot_ACT41898.1_4205    gene=creB               \n",
      "4205  >lcl|CP000819.1_prot_ACT41899.1_4206    gene=creC               \n",
      "4206  >lcl|CP000819.1_prot_ACT41900.1_4207    gene=arcA               \n",
      "4207  >lcl|CP000819.1_prot_ACT41901.1_4208    gene=yjjY               \n",
      "4208  >lcl|CP000819.1_prot_ACT41902.1_4209    gene=lasT               \n",
      "\n",
      "                locus_tag Attribute_4  \\\n",
      "0     locus_tag=ECB_00001               \n",
      "1     locus_tag=ECB_00002               \n",
      "2     locus_tag=ECB_00003               \n",
      "3     locus_tag=ECB_00004               \n",
      "4     locus_tag=ECB_00005               \n",
      "...                   ...         ...   \n",
      "4204  locus_tag=ECB_04274               \n",
      "4205  locus_tag=ECB_04275               \n",
      "4206  locus_tag=ECB_04277               \n",
      "4207  locus_tag=ECB_04278               \n",
      "4208  locus_tag=ECB_04279               \n",
      "\n",
      "                                               Function Attribute_6  \\\n",
      "0                     protein=thr operon leader peptide               \n",
      "1     protein=bifunctional aspartokinase I/homeserin...               \n",
      "2                             protein=homoserine kinase               \n",
      "3                            protein=threonine synthase               \n",
      "4                          protein=hypothetical protein               \n",
      "...                                                 ...         ...   \n",
      "4204  protein=DNA-binding response regulator in two-...               \n",
      "4205  protein=sensory histidine kinase in two-compon...               \n",
      "4206  protein=DNA-binding response regulator in two-...               \n",
      "4207                       protein=hypothetical protein               \n",
      "4208           protein=predicted rRNA methyltransferase               \n",
      "\n",
      "                 protein_id Attribute_8  \\\n",
      "0     protein_id=ACT37694.1               \n",
      "1     protein_id=ACT37695.1               \n",
      "2     protein_id=ACT37696.1               \n",
      "3     protein_id=ACT37697.1               \n",
      "4     protein_id=ACT37698.1               \n",
      "...                     ...         ...   \n",
      "4204  protein_id=ACT41898.1               \n",
      "4205  protein_id=ACT41899.1               \n",
      "4206  protein_id=ACT41900.1               \n",
      "4207  protein_id=ACT41901.1               \n",
      "4208  protein_id=ACT41902.1               \n",
      "\n",
      "                                   Location Attribute_10 Attribute_11  \\\n",
      "0                         location=190..255                 gbkey=CDS   \n",
      "1                        location=336..2798                 gbkey=CDS   \n",
      "2                       location=2800..3732                 gbkey=CDS   \n",
      "3                       location=3733..5019                 gbkey=CDS   \n",
      "4                       location=5232..5528                 gbkey=CDS   \n",
      "...                                     ...          ...          ...   \n",
      "4204              location=4624167..4624856                 gbkey=CDS   \n",
      "4205              location=4624856..4626280                 gbkey=CDS   \n",
      "4206  location=complement(4627750..4628466)                 gbkey=CDS   \n",
      "4207              location=4628562..4628702                 gbkey=CDS   \n",
      "4208              location=4629102..4629788                 gbkey=CDS   \n",
      "\n",
      "     Attribute_12 Attribute_13 Attribute_14     C  Gene contains_space  \n",
      "0                         None         None  gene  thrL             No  \n",
      "1                         None         None  gene  thrA             No  \n",
      "2                         None         None  gene  thrB             No  \n",
      "3                         None         None  gene  thrC             No  \n",
      "4                         None         None  gene  yaaX             No  \n",
      "...           ...          ...          ...   ...   ...            ...  \n",
      "4204                      None         None  gene  creB             No  \n",
      "4205                      None         None  gene  creC             No  \n",
      "4206                      None         None  gene  arcA             No  \n",
      "4207                      None         None  gene  yjjY             No  \n",
      "4208                      None         None  gene  lasT             No  \n",
      "\n",
      "[4209 rows x 20 columns]\n"
     ]
    }
   ],
   "source": [
    "# Check for spaces between characters in each row of \"col2\"\n",
    "result_df['contains_space'] = result_df['Gene'].str.contains(r'\\s')\n",
    "\n",
    "# Replace True with \"Yes\" and False with \"No\" in the \"contains_space\" column\n",
    "result_df['contains_space'] = result_df['contains_space'].replace({True: 'Yes', False: 'No'})\n",
    "\n",
    "# Print the DataFrame\n",
    "print(result_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['No'], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df['contains_space'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Empty DataFrame\n",
      "Columns: [Info, Prt_Seq, Attribute_0, Attribute_1, Attribute_2, locus_tag, Attribute_4, Function, Attribute_6, protein_id, Attribute_8, Location, Attribute_10, Attribute_11, Attribute_12, Attribute_13, Attribute_14, C, Gene, contains_space]\n",
      "Index: []\n"
     ]
    }
   ],
   "source": [
    "# Check for duplicate values in the \"col2\" column\n",
    "duplicates = result_df[result_df['Gene'].duplicated(keep=False)]\n",
    "\n",
    "# Print the duplicate rows\n",
    "print(duplicates)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the columns you want to keep\n",
    "Prt = result_df[['Info', 'Prt_Seq', 'Attribute_0', 'Attribute_1', 'Attribute_2',\n",
    "       'locus_tag', 'Attribute_4', 'Function', 'Attribute_6', 'protein_id',\n",
    "       'Attribute_8', 'Location', 'Attribute_10', 'Attribute_11',\n",
    "       'Attribute_12', 'Attribute_13', 'Attribute_14', 'C', 'Gene',]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame saved to /home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/Prt_seq.xlsx\n"
     ]
    }
   ],
   "source": [
    "# Write in local \n",
    "# Specify the file path where you want to save the Excel file\n",
    "excel_file_path = '/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/Prt_seq.xlsx'\n",
    "\n",
    "# Write the DataFrame to an Excel file\n",
    "result_df.to_excel(excel_file_path, index=False)\n",
    "\n",
    "print(f'DataFrame saved to {excel_file_path}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now we have DNA and protien sequence for each gene. So merge it with variant file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "df = pd.read_excel('/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Mastersheet.xlsx',sheet_name='missense') #Variant sheet \n",
    "dna = pd.read_excel(\"/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_seq.xlsx\") # DNA sheet \n",
    "prt = pd.read_excel(\"/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/Prt_seq.xlsx\") # Protein sheet "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge DNA with prt firstly then merge thus df with the original sheet \n",
    "DNA_prt = dna.merge(prt, on='Gene', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge DNA with prt firstly then merge thus df with the original sheet \n",
    "Final_df= df.merge(DNA_prt, on='Gene', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame saved to /home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_variant.xlsx\n"
     ]
    }
   ],
   "source": [
    "# Write in local \n",
    "# Specify the file path where you want to save the Excel file\n",
    "excel_file_path = '/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_variant.xlsx'\n",
    "\n",
    "# Write the DataFrame to an Excel file\n",
    "Final_df.to_excel(excel_file_path, index=False)\n",
    "\n",
    "print(f'DataFrame saved to {excel_file_path}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step2 "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Prepare this data as per input format for esm1b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The letter 'U' is present in at least one row of 'Prt_Seq'.\n"
     ]
    }
   ],
   "source": [
    "# Check if U is present in any of the prt seq \n",
    "# Check if 'U' is present in any row of 'Prt_Seq'\n",
    "contains_U = Final_df['Prt_Seq'].str.contains('U')\n",
    "\n",
    "# Print the result\n",
    "if contains_U.any():\n",
    "    print(\"The letter 'U' is present in at least one row of 'Prt_Seq'.\")\n",
    "else:\n",
    "    print(\"The letter 'U' is not present in any row of 'Prt_Seq'.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Position  Gene Allele Annotation  Test.statistic   P-value Ref_allele  \\\n",
      "1054    1522293  fdnG   A->C   missense        1.385120  0.000100          A   \n",
      "1055    1522820  fdnG   A->C   missense        3.572310  0.001300          A   \n",
      "1056    1523598  fdnG   A->C   missense        2.565300  0.000100          A   \n",
      "1057    1524108  fdnG   T->G   missense        2.080730  0.000100          T   \n",
      "2971    4060956  fdoG   T->G   missense        2.337940  0.011356          T   \n",
      "2972    4062141  fdoG   T->G   missense        2.423190  0.002700          T   \n",
      "2973    4062272  fdoG   T->G   missense        1.401510  0.000100          T   \n",
      "3137    4276594  fdhF   T->G   missense        2.645590  0.000200          T   \n",
      "3138    4277719  fdhF   A->C   missense        1.647880  0.000100          A   \n",
      "4022    1523115  fdnG   A->G   missense        3.006440  0.005999          A   \n",
      "4023    1523184  fdnG   A->G   missense        1.021040  0.000300          A   \n",
      "4024    1523997  fdnG   C->T   missense        0.000000  1.000000          C   \n",
      "4025    1524453  fdnG   A->G   missense        0.903494  0.004400          A   \n",
      "4026    1524548  fdnG   T->C   missense        1.169450  0.000100          T   \n",
      "5076    4061467  fdoG   G->A   missense        4.091080  0.000500          G   \n",
      "5077    4061692  fdoG   A->G   missense        0.792414  0.000100          A   \n",
      "5156    4278094  fdhF   G->A   missense        3.255410  0.004500          G   \n",
      "6955    4061284  fdoG   A->G   missense        2.972440  0.001500          A   \n",
      "7992    1522559  fdnG   T->C   missense        5.425350  0.001100          T   \n",
      "7993    1523115  fdnG   A->G   missense        4.265660  0.006499          A   \n",
      "7994    1524245  fdnG   G->A   missense        2.254420  0.023942          G   \n",
      "9571    4062721  fdoG   C->T   missense        3.578140  0.000700          C   \n",
      "9572    4063299  fdoG   T->C   missense        2.782610  0.003000          T   \n",
      "9666    4277855  fdhF   T->C   missense        2.741930  0.006799          T   \n",
      "9667    4277869  fdhF   C->T   missense        2.764680  0.016129          C   \n",
      "9668    4278166  fdhF   C->T   missense        0.973316  0.000100          C   \n",
      "12473   1523624  fdnG   G->A   missense        1.391360  0.061005          G   \n",
      "12474   1524237  fdnG   G->A   missense        1.297990  0.000100          G   \n",
      "15071   4063468  fdoG   C->T   missense        4.827160  0.000100          C   \n",
      "15072   4063474  fdoG   C->T   missense        0.963534  0.000100          C   \n",
      "17600   1521906  fdnG   A->C   missense        0.975737  0.000100          A   \n",
      "17601   1522233  fdnG   T->G   missense        1.184850  0.000100          T   \n",
      "17602   1522236  fdnG   T->G   missense        3.445140  0.000100          T   \n",
      "17603   1523085  fdnG   T->G   missense        2.924440  0.000500          T   \n",
      "17604   1523588  fdnG   T->G   missense        1.858160  0.000100          T   \n",
      "17605   1523648  fdnG   T->G   missense        0.963314  0.000100          T   \n",
      "17606   1523718  fdnG   T->G   missense        2.933740  0.000100          T   \n",
      "17607   1524237  fdnG   G->A   missense        1.249020  0.000400          G   \n",
      "21957   4061045  fdoG   A->C   missense        2.797310  0.000100          A   \n",
      "21958   4061340  fdoG   T->G   missense        2.720390  0.000100          T   \n",
      "21959   4062565  fdoG   T->G   missense        1.416900  0.000100          T   \n",
      "21960   4062799  fdoG   A->C   missense        2.148510  0.009999          A   \n",
      "21961   4062916  fdoG   T->G   missense        1.122310  0.000100          T   \n",
      "21962   4063942  fdoG   T->G   missense        0.958080  0.000100          T   \n",
      "22259   4276126  fdhF   A->C   missense        2.008930  0.000100          A   \n",
      "22260   4276408  fdhF   T->G   missense        2.887930  0.000200          T   \n",
      "22261   4277160  fdhF   A->C   missense        4.403190  0.000100          A   \n",
      "22262   4277215  fdhF   G->C   missense        0.000000  1.000000          G   \n",
      "22263   4277329  fdhF   T->G   missense        3.636720  0.001200          T   \n",
      "22264   4277356  fdhF   A->C   missense        2.708630  0.005099          A   \n",
      "22265   4277708  fdhF   T->G   missense        2.441190  0.004100          T   \n",
      "\n",
      "      Alt_allele label                                             Info_x  \\\n",
      "1054           C    m1  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "1055           C    m1  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "1056           C    m1  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "1057           G    m1  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "2971           G    m1  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "2972           G    m1  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "2973           G    m1  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "3137           G    m1  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "3138           C    m1  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "4022           G    m2  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "4023           G    m2  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "4024           T    m2  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "4025           G    m2  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "4026           C    m2  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "5076           A    m2  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "5077           G    m2  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "5156           A    m2  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "6955           G    m3  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "7992           C    m4  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "7993           G    m4  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "7994           A    m4  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "9571           T    m4  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "9572           C    m4  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "9666           C    m4  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "9667           T    m4  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "9668           T    m4  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "12473          A    p3  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "12474          A    p3  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "15071          T    p5  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "15072          T    p5  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "17600          C    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "17601          G    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "17602          G    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "17603          G    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "17604          G    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "17605          G    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "17606          G    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "17607          A    p6  >lcl|CP000819.1_cds_ACT39104.1_1411 [gene=fdnG...   \n",
      "21957          C    p6  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "21958          G    p6  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "21959          G    p6  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "21960          C    p6  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "21961          G    p6  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "21962          G    p6  >lcl|CP000819.1_cds_ACT41416.1_3723 [gene=fdoG...   \n",
      "22259          C    p6  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "22260          G    p6  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "22261          C    p6  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "22262          C    p6  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "22263          G    p6  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "22264          C    p6  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "22265          G    p6  >lcl|CP000819.1_cds_ACT41586.1_3893 [gene=fdhF...   \n",
      "\n",
      "       ...                         protein_id_y Attribute_8_y  \\\n",
      "1054   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "1055   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "1056   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "1057   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "2971   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "2972   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "2973   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "3137   ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "3138   ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "4022   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "4023   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "4024   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "4025   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "4026   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "5076   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "5077   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "5156   ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "6955   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "7992   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "7993   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "7994   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "9571   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "9572   ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "9666   ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "9667   ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "9668   ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "12473  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "12474  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "15071  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "15072  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17600  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17601  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17602  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17603  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17604  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17605  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17606  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "17607  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "21957  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "21958  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "21959  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "21960  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "21961  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "21962  ...  transl_except=(pos:586..588,aa:Sec)                 \n",
      "22259  ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "22260  ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "22261  ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "22262  ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "22263  ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "22264  ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "22265  ...  transl_except=(pos:418..420,aa:Sec)                 \n",
      "\n",
      "                  Location_y Attribute_10_y  \\\n",
      "1054   protein_id=ACT39104.1                  \n",
      "1055   protein_id=ACT39104.1                  \n",
      "1056   protein_id=ACT39104.1                  \n",
      "1057   protein_id=ACT39104.1                  \n",
      "2971   protein_id=ACT41416.1                  \n",
      "2972   protein_id=ACT41416.1                  \n",
      "2973   protein_id=ACT41416.1                  \n",
      "3137   protein_id=ACT41586.1                  \n",
      "3138   protein_id=ACT41586.1                  \n",
      "4022   protein_id=ACT39104.1                  \n",
      "4023   protein_id=ACT39104.1                  \n",
      "4024   protein_id=ACT39104.1                  \n",
      "4025   protein_id=ACT39104.1                  \n",
      "4026   protein_id=ACT39104.1                  \n",
      "5076   protein_id=ACT41416.1                  \n",
      "5077   protein_id=ACT41416.1                  \n",
      "5156   protein_id=ACT41586.1                  \n",
      "6955   protein_id=ACT41416.1                  \n",
      "7992   protein_id=ACT39104.1                  \n",
      "7993   protein_id=ACT39104.1                  \n",
      "7994   protein_id=ACT39104.1                  \n",
      "9571   protein_id=ACT41416.1                  \n",
      "9572   protein_id=ACT41416.1                  \n",
      "9666   protein_id=ACT41586.1                  \n",
      "9667   protein_id=ACT41586.1                  \n",
      "9668   protein_id=ACT41586.1                  \n",
      "12473  protein_id=ACT39104.1                  \n",
      "12474  protein_id=ACT39104.1                  \n",
      "15071  protein_id=ACT41416.1                  \n",
      "15072  protein_id=ACT41416.1                  \n",
      "17600  protein_id=ACT39104.1                  \n",
      "17601  protein_id=ACT39104.1                  \n",
      "17602  protein_id=ACT39104.1                  \n",
      "17603  protein_id=ACT39104.1                  \n",
      "17604  protein_id=ACT39104.1                  \n",
      "17605  protein_id=ACT39104.1                  \n",
      "17606  protein_id=ACT39104.1                  \n",
      "17607  protein_id=ACT39104.1                  \n",
      "21957  protein_id=ACT41416.1                  \n",
      "21958  protein_id=ACT41416.1                  \n",
      "21959  protein_id=ACT41416.1                  \n",
      "21960  protein_id=ACT41416.1                  \n",
      "21961  protein_id=ACT41416.1                  \n",
      "21962  protein_id=ACT41416.1                  \n",
      "22259  protein_id=ACT41586.1                  \n",
      "22260  protein_id=ACT41586.1                  \n",
      "22261  protein_id=ACT41586.1                  \n",
      "22262  protein_id=ACT41586.1                  \n",
      "22263  protein_id=ACT41586.1                  \n",
      "22264  protein_id=ACT41586.1                  \n",
      "22265  protein_id=ACT41586.1                  \n",
      "\n",
      "                              Attribute_11_y Attribute_12_y Attribute_13_y  \\\n",
      "1054               location=1521551..1524598                     gbkey=CDS   \n",
      "1055               location=1521551..1524598                     gbkey=CDS   \n",
      "1056               location=1521551..1524598                     gbkey=CDS   \n",
      "1057               location=1521551..1524598                     gbkey=CDS   \n",
      "2971   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "2972   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "2973   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "3137   location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "3138   location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "4022               location=1521551..1524598                     gbkey=CDS   \n",
      "4023               location=1521551..1524598                     gbkey=CDS   \n",
      "4024               location=1521551..1524598                     gbkey=CDS   \n",
      "4025               location=1521551..1524598                     gbkey=CDS   \n",
      "4026               location=1521551..1524598                     gbkey=CDS   \n",
      "5076   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "5077   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "5156   location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "6955   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "7992               location=1521551..1524598                     gbkey=CDS   \n",
      "7993               location=1521551..1524598                     gbkey=CDS   \n",
      "7994               location=1521551..1524598                     gbkey=CDS   \n",
      "9571   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "9572   location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "9666   location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "9667   location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "9668   location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "12473              location=1521551..1524598                     gbkey=CDS   \n",
      "12474              location=1521551..1524598                     gbkey=CDS   \n",
      "15071  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "15072  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "17600              location=1521551..1524598                     gbkey=CDS   \n",
      "17601              location=1521551..1524598                     gbkey=CDS   \n",
      "17602              location=1521551..1524598                     gbkey=CDS   \n",
      "17603              location=1521551..1524598                     gbkey=CDS   \n",
      "17604              location=1521551..1524598                     gbkey=CDS   \n",
      "17605              location=1521551..1524598                     gbkey=CDS   \n",
      "17606              location=1521551..1524598                     gbkey=CDS   \n",
      "17607              location=1521551..1524598                     gbkey=CDS   \n",
      "21957  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "21958  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "21959  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "21960  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "21961  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "21962  location=complement(4060919..4063969)                     gbkey=CDS   \n",
      "22259  location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "22260  location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "22261  location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "22262  location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "22263  location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "22264  location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "22265  location=complement(4276041..4278188)                     gbkey=CDS   \n",
      "\n",
      "      Attribute_14_y   C_y contains_space  \n",
      "1054             NaN  gene             No  \n",
      "1055             NaN  gene             No  \n",
      "1056             NaN  gene             No  \n",
      "1057             NaN  gene             No  \n",
      "2971             NaN  gene             No  \n",
      "2972             NaN  gene             No  \n",
      "2973             NaN  gene             No  \n",
      "3137             NaN  gene             No  \n",
      "3138             NaN  gene             No  \n",
      "4022             NaN  gene             No  \n",
      "4023             NaN  gene             No  \n",
      "4024             NaN  gene             No  \n",
      "4025             NaN  gene             No  \n",
      "4026             NaN  gene             No  \n",
      "5076             NaN  gene             No  \n",
      "5077             NaN  gene             No  \n",
      "5156             NaN  gene             No  \n",
      "6955             NaN  gene             No  \n",
      "7992             NaN  gene             No  \n",
      "7993             NaN  gene             No  \n",
      "7994             NaN  gene             No  \n",
      "9571             NaN  gene             No  \n",
      "9572             NaN  gene             No  \n",
      "9666             NaN  gene             No  \n",
      "9667             NaN  gene             No  \n",
      "9668             NaN  gene             No  \n",
      "12473            NaN  gene             No  \n",
      "12474            NaN  gene             No  \n",
      "15071            NaN  gene             No  \n",
      "15072            NaN  gene             No  \n",
      "17600            NaN  gene             No  \n",
      "17601            NaN  gene             No  \n",
      "17602            NaN  gene             No  \n",
      "17603            NaN  gene             No  \n",
      "17604            NaN  gene             No  \n",
      "17605            NaN  gene             No  \n",
      "17606            NaN  gene             No  \n",
      "17607            NaN  gene             No  \n",
      "21957            NaN  gene             No  \n",
      "21958            NaN  gene             No  \n",
      "21959            NaN  gene             No  \n",
      "21960            NaN  gene             No  \n",
      "21961            NaN  gene             No  \n",
      "21962            NaN  gene             No  \n",
      "22259            NaN  gene             No  \n",
      "22260            NaN  gene             No  \n",
      "22261            NaN  gene             No  \n",
      "22262            NaN  gene             No  \n",
      "22263            NaN  gene             No  \n",
      "22264            NaN  gene             No  \n",
      "22265            NaN  gene             No  \n",
      "\n",
      "[51 rows x 46 columns]\n"
     ]
    }
   ],
   "source": [
    "# Check if 'U' is present in any row of 'Prt_Seq'\n",
    "contains_U = Final_df['Prt_Seq'].str.contains('U')\n",
    "\n",
    "# Find the rows where 'U' is present\n",
    "rows_with_U = Final_df.loc[contains_U]\n",
    "\n",
    "# Print the rows containing 'U'\n",
    "print(rows_with_U)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Position  Gene Allele Annotation  Test.statistic   P-value Ref_allele  \\\n",
      "0           241  thrL   A->C   missense        2.887050  0.000100          A   \n",
      "1          1615  thrA   T->G   missense        0.986369  0.000100          T   \n",
      "2          6046  yaaA   C->A   missense        1.681940  0.077982          C   \n",
      "3          6220  yaaA   T->G   missense        2.478710  0.000100          T   \n",
      "4          7406  yaaJ   A->C   missense        3.437140  0.002300          A   \n",
      "...         ...   ...    ...        ...             ...       ...        ...   \n",
      "22852   4627830  arcA   C->T   missense        1.390190  0.012235          C   \n",
      "22853   4627933  arcA   T->G   missense        2.990930  0.005099          T   \n",
      "22854   4628381  arcA   T->G   missense        1.179320  0.000100          T   \n",
      "22855   4628701  yjjY   A->C   missense        5.263680  0.000100          A   \n",
      "22856   4629600  lasT   A->G   missense        1.506350  0.000100          A   \n",
      "\n",
      "      Alt_allele label                                             Info_x  \\\n",
      "0              C    m1  >lcl|CP000819.1_cds_ACT37694.1_1 [gene=thrL] [...   \n",
      "1              G    m1  >lcl|CP000819.1_cds_ACT37695.1_2 [gene=thrA] [...   \n",
      "2              A    m1  >lcl|CP000819.1_cds_ACT37699.1_6 [gene=yaaA] [...   \n",
      "3              G    m1  >lcl|CP000819.1_cds_ACT37699.1_6 [gene=yaaA] [...   \n",
      "4              C    m1  >lcl|CP000819.1_cds_ACT37700.1_7 [gene=yaaJ] [...   \n",
      "...          ...   ...                                                ...   \n",
      "22852          T    p6  >lcl|CP000819.1_cds_ACT41900.1_4207 [gene=arcA...   \n",
      "22853          G    p6  >lcl|CP000819.1_cds_ACT41900.1_4207 [gene=arcA...   \n",
      "22854          G    p6  >lcl|CP000819.1_cds_ACT41900.1_4207 [gene=arcA...   \n",
      "22855          C    p6  >lcl|CP000819.1_cds_ACT41901.1_4208 [gene=yjjY...   \n",
      "22856          G    p6  >lcl|CP000819.1_cds_ACT41902.1_4209 [gene=lasT...   \n",
      "\n",
      "       ...           protein_id_y Attribute_8_y  \\\n",
      "0      ...  protein_id=ACT37694.1                 \n",
      "1      ...  protein_id=ACT37695.1                 \n",
      "2      ...  protein_id=ACT37699.1                 \n",
      "3      ...  protein_id=ACT37699.1                 \n",
      "4      ...  protein_id=ACT37700.1                 \n",
      "...    ...                    ...           ...   \n",
      "22852  ...  protein_id=ACT41900.1                 \n",
      "22853  ...  protein_id=ACT41900.1                 \n",
      "22854  ...  protein_id=ACT41900.1                 \n",
      "22855  ...  protein_id=ACT41901.1                 \n",
      "22856  ...  protein_id=ACT41902.1                 \n",
      "\n",
      "                                  Location_y Attribute_10_y Attribute_11_y  \\\n",
      "0                          location=190..255                     gbkey=CDS   \n",
      "1                         location=336..2798                     gbkey=CDS   \n",
      "2            location=complement(5681..6457)                     gbkey=CDS   \n",
      "3            location=complement(5681..6457)                     gbkey=CDS   \n",
      "4            location=complement(6527..7957)                     gbkey=CDS   \n",
      "...                                      ...            ...            ...   \n",
      "22852  location=complement(4627750..4628466)                     gbkey=CDS   \n",
      "22853  location=complement(4627750..4628466)                     gbkey=CDS   \n",
      "22854  location=complement(4627750..4628466)                     gbkey=CDS   \n",
      "22855              location=4628562..4628702                     gbkey=CDS   \n",
      "22856              location=4629102..4629788                     gbkey=CDS   \n",
      "\n",
      "      Attribute_12_y Attribute_13_y Attribute_14_y   C_y contains_space  \n",
      "0                NaN            NaN            NaN  gene             No  \n",
      "1                NaN            NaN            NaN  gene             No  \n",
      "2                NaN            NaN            NaN  gene             No  \n",
      "3                NaN            NaN            NaN  gene             No  \n",
      "4                NaN            NaN            NaN  gene             No  \n",
      "...              ...            ...            ...   ...            ...  \n",
      "22852            NaN            NaN            NaN  gene             No  \n",
      "22853            NaN            NaN            NaN  gene             No  \n",
      "22854            NaN            NaN            NaN  gene             No  \n",
      "22855            NaN            NaN            NaN  gene             No  \n",
      "22856            NaN            NaN            NaN  gene             No  \n",
      "\n",
      "[22753 rows x 46 columns]\n"
     ]
    }
   ],
   "source": [
    "# Check if 'U' is present in any row of 'col2'\n",
    "contains_U = Final_df['Prt_Seq'].str.contains('U')\n",
    "\n",
    "# Create a new DataFrame without rows containing 'U'\n",
    "Final_df = Final_df[~contains_U]\n",
    "\n",
    "# Display the filtered DataFrame\n",
    "print(Final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop nan \n",
    "Final_df = Final_df.dropna(subset=['Prt_Seq'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of columns you want to keep esm1b input \n",
    "columns_to_keep = ['Position', 'Gene','Prt_Seq']\n",
    "\n",
    "# Select the desired columns and assign them back to the DataFrame\n",
    "Final_df = Final_df[columns_to_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge 'Position' and 'Gene' into a new column 'seq_id' with an underscore separator\n",
    "Final_df['seq_id'] = Final_df['Position'].astype(str) + '_' + df['Gene'].astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of columns you want to keep\n",
    "columns_to_keep = ['seq_id', 'Prt_Seq']\n",
    "\n",
    "# Select the desired columns and assign them back to the DataFrame\n",
    "Final_df = Final_df[columns_to_keep]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "Final_df.columns = ['Prt_Seq_WT' if col == 'Prt_Seq' else col for col in Final_df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data written to /home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/esm1b_input_WT.fasta\n"
     ]
    }
   ],
   "source": [
    "## Esm1B format file \n",
    "import pandas as pd\n",
    "\n",
    "# Read the Excel file\n",
    "# all_data = pd.read_excel(\"/content/drive/MyDrive/compiled_final.xlsx\")\n",
    "\n",
    "# Output file name\n",
    "output_file = '/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/esm1b_input_WT.fasta'\n",
    "\n",
    "# Write to the text file\n",
    "with open(output_file, 'w') as f:\n",
    "    for index, row in Final_df.iterrows():\n",
    "        #f.write(f'>{row[\"seq_id\"]}\\n{row[\"aa_seq_Mut\"]}\\n')\n",
    "        f.write(f'>{row[\"seq_id\"]}\\n{row[\"Prt_Seq_WT\"]}\\n')\n",
    "\n",
    "print('Data written to', output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This was the source of the ESM1b model -> https://github.com/ntranoslab/esm-variants and this was the command line python3 esm_score_missense_mutations.py --input-fasta-file /path/to/input.fasta --output-csv-file /path/to/output.csv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the sheet before transforming it to esm1b input \n",
    "# read the sheet \n",
    "Final_df = pd.read_excel('/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_variant.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This sheet was used to get the location GCA_000017985.1_ASM1798v1_feature_table.txt and was put in the DNA_Prt_variant in excel by vlookup and all the unwanted columns were removed.Also those variants were removed which did not have gene start and stop and genes which had upstream variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "Final_df = pd.read_excel('/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_Strand_variant.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Now take this file to script2.R script to generate the mutant DNA sequence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### After generating Mut DNA seq from script2.R get the file here.\n",
    "### Now convert the WT DNA to WT prt to check if it matches the NCBI submitted Prt seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "data = pd.read_excel(\"~/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_Strand_Variant_MutDNA.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>DNA_Seq</th>\n",
       "      <th>Prt_Seq</th>\n",
       "      <th>Sequence_Mut</th>\n",
       "      <th>Diff_Count</th>\n",
       "      <th>Differences</th>\n",
       "      <th>Diff_Pos</th>\n",
       "      <th>has_space</th>\n",
       "      <th>WT_aa_made</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>241</td>\n",
       "      <td>False</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>336</td>\n",
       "      <td>2798</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>1615</td>\n",
       "      <td>False</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>8236</td>\n",
       "      <td>9189</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>8430</td>\n",
       "      <td>False</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>12527</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>13974</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21765</th>\n",
       "      <td>4621500</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4621384</td>\n",
       "      <td>4621905</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4621789</td>\n",
       "      <td>False</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21766</th>\n",
       "      <td>4623143</td>\n",
       "      <td>rob</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4622601</td>\n",
       "      <td>4623470</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4622928</td>\n",
       "      <td>False</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21767</th>\n",
       "      <td>4627830</td>\n",
       "      <td>arcA</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>GA</td>\n",
       "      <td>4628386</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21768</th>\n",
       "      <td>4627933</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4628283</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21769</th>\n",
       "      <td>4628381</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4627835</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21770 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position  Gene Allele Ref_allele Alt_allele Annotation label    Start  \\\n",
       "0           241  thrL   A->C          A          C   missense    m1      190   \n",
       "1          1615  thrA   T->G          T          G   missense    m1      336   \n",
       "2          8430  talB   T->G          T          G   missense    m1     8236   \n",
       "3         12527  dnaK   A->C          A          C   missense    m1    12161   \n",
       "4         13974  dnaK   A->C          A          C   missense    m1    12161   \n",
       "...         ...   ...    ...        ...        ...        ...   ...      ...   \n",
       "21765   4621500  yjjX   T->G          T          C   missense    p6  4621384   \n",
       "21766   4623143   rob   T->G          T          C   missense    p6  4622601   \n",
       "21767   4627830  arcA   C->T          C          A   missense    p6  4627750   \n",
       "21768   4627933  arcA   T->G          T          C   missense    p6  4627750   \n",
       "21769   4628381  arcA   T->G          T          C   missense    p6  4627750   \n",
       "\n",
       "           End Strand                                            DNA_Seq  \\\n",
       "0          255      +  ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...   \n",
       "1         2798      +  ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...   \n",
       "2         9189      +  ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...   \n",
       "3        14077      +  ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "4        14077      +  ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "...        ...    ...                                                ...   \n",
       "21765  4621905      -  ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...   \n",
       "21766  4623470      -  ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...   \n",
       "21767  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "21768  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "21769  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "\n",
       "                                                 Prt_Seq  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "...                                                  ...   \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...   \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...   \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "\n",
       "                                            Sequence_Mut  Diff_Count  \\\n",
       "0      ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...           1   \n",
       "1      ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...           1   \n",
       "2      ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...           1   \n",
       "3      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...           1   \n",
       "4      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...           1   \n",
       "...                                                  ...         ...   \n",
       "21765  ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...           1   \n",
       "21766  ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...           1   \n",
       "21767  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "21768  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "21769  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "\n",
       "      Differences  Diff_Pos  has_space  \\\n",
       "0              AC       241      False   \n",
       "1              TG      1615      False   \n",
       "2              TG      8430      False   \n",
       "3              AC     12527      False   \n",
       "4              AC     13974      False   \n",
       "...           ...       ...        ...   \n",
       "21765          AC   4621789      False   \n",
       "21766          AC   4622928      False   \n",
       "21767          GA   4628386      False   \n",
       "21768          AC   4628283      False   \n",
       "21769          AC   4627835      False   \n",
       "\n",
       "                                              WT_aa_made  \n",
       "0                                 MKRISTTITTTITITTGNGAG   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...  \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  \n",
       "...                                                  ...  \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...  \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...  \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  \n",
       "\n",
       "[21770 rows x 18 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "import pandas as pd\n",
    "\n",
    "# Define the custom codon table\n",
    "custom_codon_table = CodonTable.unambiguous_dna_by_id[11]  # Change the ID as needed\n",
    "data['WT_aa_made'] = data['DNA_Seq'].apply(lambda cds_sequence: str(Seq(cds_sequence).translate(table=custom_codon_table)).replace('*', ' '))\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>DNA_Seq</th>\n",
       "      <th>Prt_Seq</th>\n",
       "      <th>Sequence_Mut</th>\n",
       "      <th>Diff_Count</th>\n",
       "      <th>Differences</th>\n",
       "      <th>Diff_Pos</th>\n",
       "      <th>has_space</th>\n",
       "      <th>WT_aa_made</th>\n",
       "      <th>comparison</th>\n",
       "      <th>Mut_aa_made</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>241</td>\n",
       "      <td>False</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>True</td>\n",
       "      <td>MKRISTTITTTITITTGHGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>336</td>\n",
       "      <td>2798</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>1615</td>\n",
       "      <td>False</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>True</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>8236</td>\n",
       "      <td>9189</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>8430</td>\n",
       "      <td>False</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>True</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>12527</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>13974</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21765</th>\n",
       "      <td>4621500</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4621384</td>\n",
       "      <td>4621905</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4621789</td>\n",
       "      <td>False</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>True</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21766</th>\n",
       "      <td>4623143</td>\n",
       "      <td>rob</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4622601</td>\n",
       "      <td>4623470</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4622928</td>\n",
       "      <td>False</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>True</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21767</th>\n",
       "      <td>4627830</td>\n",
       "      <td>arcA</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>GA</td>\n",
       "      <td>4628386</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21768</th>\n",
       "      <td>4627933</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4628283</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21769</th>\n",
       "      <td>4628381</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4627835</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21770 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position  Gene Allele Ref_allele Alt_allele Annotation label    Start  \\\n",
       "0           241  thrL   A->C          A          C   missense    m1      190   \n",
       "1          1615  thrA   T->G          T          G   missense    m1      336   \n",
       "2          8430  talB   T->G          T          G   missense    m1     8236   \n",
       "3         12527  dnaK   A->C          A          C   missense    m1    12161   \n",
       "4         13974  dnaK   A->C          A          C   missense    m1    12161   \n",
       "...         ...   ...    ...        ...        ...        ...   ...      ...   \n",
       "21765   4621500  yjjX   T->G          T          C   missense    p6  4621384   \n",
       "21766   4623143   rob   T->G          T          C   missense    p6  4622601   \n",
       "21767   4627830  arcA   C->T          C          A   missense    p6  4627750   \n",
       "21768   4627933  arcA   T->G          T          C   missense    p6  4627750   \n",
       "21769   4628381  arcA   T->G          T          C   missense    p6  4627750   \n",
       "\n",
       "           End Strand                                            DNA_Seq  \\\n",
       "0          255      +  ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...   \n",
       "1         2798      +  ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...   \n",
       "2         9189      +  ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...   \n",
       "3        14077      +  ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "4        14077      +  ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "...        ...    ...                                                ...   \n",
       "21765  4621905      -  ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...   \n",
       "21766  4623470      -  ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...   \n",
       "21767  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "21768  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "21769  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "\n",
       "                                                 Prt_Seq  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "...                                                  ...   \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...   \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...   \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "\n",
       "                                            Sequence_Mut  Diff_Count  \\\n",
       "0      ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...           1   \n",
       "1      ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...           1   \n",
       "2      ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...           1   \n",
       "3      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...           1   \n",
       "4      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...           1   \n",
       "...                                                  ...         ...   \n",
       "21765  ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...           1   \n",
       "21766  ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...           1   \n",
       "21767  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "21768  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "21769  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "\n",
       "      Differences  Diff_Pos  has_space  \\\n",
       "0              AC       241      False   \n",
       "1              TG      1615      False   \n",
       "2              TG      8430      False   \n",
       "3              AC     12527      False   \n",
       "4              AC     13974      False   \n",
       "...           ...       ...        ...   \n",
       "21765          AC   4621789      False   \n",
       "21766          AC   4622928      False   \n",
       "21767          GA   4628386      False   \n",
       "21768          AC   4628283      False   \n",
       "21769          AC   4627835      False   \n",
       "\n",
       "                                              WT_aa_made  comparison  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG        True   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...        True   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...        True   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...        True   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...        True   \n",
       "...                                                  ...         ...   \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...        True   \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...        True   \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "\n",
       "                                             Mut_aa_made  \n",
       "0                                 MKRISTTITTTITITTGHGAG   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...  \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  \n",
       "...                                                  ...  \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...  \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...  \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...  \n",
       "\n",
       "[21770 rows x 20 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.Data import CodonTable\n",
    "import pandas as pd\n",
    "\n",
    "# Define the custom codon table\n",
    "custom_codon_table = CodonTable.unambiguous_dna_by_id[11]  # Change the ID as needed\n",
    "data['Mut_aa_made'] = data['Sequence_Mut'].apply(lambda cds_sequence: str(Seq(cds_sequence).translate(table=custom_codon_table)).replace('*', ' '))\n",
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove any empty spcae \n",
    "data['WT_aa_made'] = data['WT_aa_made'].str.replace(' ', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove any empty spcae \n",
    "data['Prt_Seq'] = data['Prt_Seq'].str.replace(' ', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Compare sequences and create a new column\n",
    "data['comparison'] = data.apply(lambda row: row['Prt_Seq'] == row['WT_aa_made'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ True, False])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['comparison'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of 'False' values in comparison: 2216\n"
     ]
    }
   ],
   "source": [
    "# Count the number of 'True' values in 'comparison'\n",
    "false_count = (data['comparison'] == False).sum()\n",
    "\n",
    "print(\"Number of 'False' values in comparison:\", false_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of 'True' values in comparison: 19554\n"
     ]
    }
   ],
   "source": [
    "# Count the number of 'True' values in 'comparison'\n",
    "true_count = (data['comparison'] == True).sum()\n",
    "\n",
    "print(\"Number of 'True' values in comparison:\", true_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset rows with 'True' values into a new DataFrame\n",
    "true_df = data[data['comparison'] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>DNA_Seq</th>\n",
       "      <th>Prt_Seq</th>\n",
       "      <th>Sequence_Mut</th>\n",
       "      <th>Diff_Count</th>\n",
       "      <th>Differences</th>\n",
       "      <th>Diff_Pos</th>\n",
       "      <th>has_space</th>\n",
       "      <th>WT_aa_made</th>\n",
       "      <th>comparison</th>\n",
       "      <th>Mut_aa_made</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>241</td>\n",
       "      <td>False</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>True</td>\n",
       "      <td>MKRISTTITTTITITTGHGAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>336</td>\n",
       "      <td>2798</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>1615</td>\n",
       "      <td>False</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>True</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>8236</td>\n",
       "      <td>9189</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>8430</td>\n",
       "      <td>False</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>True</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>12527</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>13974</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21765</th>\n",
       "      <td>4621500</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4621384</td>\n",
       "      <td>4621905</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4621789</td>\n",
       "      <td>False</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>True</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21766</th>\n",
       "      <td>4623143</td>\n",
       "      <td>rob</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4622601</td>\n",
       "      <td>4623470</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4622928</td>\n",
       "      <td>False</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>True</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21767</th>\n",
       "      <td>4627830</td>\n",
       "      <td>arcA</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>GA</td>\n",
       "      <td>4628386</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21768</th>\n",
       "      <td>4627933</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4628283</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21769</th>\n",
       "      <td>4628381</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4627835</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19554 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position  Gene Allele Ref_allele Alt_allele Annotation label    Start  \\\n",
       "0           241  thrL   A->C          A          C   missense    m1      190   \n",
       "1          1615  thrA   T->G          T          G   missense    m1      336   \n",
       "2          8430  talB   T->G          T          G   missense    m1     8236   \n",
       "3         12527  dnaK   A->C          A          C   missense    m1    12161   \n",
       "4         13974  dnaK   A->C          A          C   missense    m1    12161   \n",
       "...         ...   ...    ...        ...        ...        ...   ...      ...   \n",
       "21765   4621500  yjjX   T->G          T          C   missense    p6  4621384   \n",
       "21766   4623143   rob   T->G          T          C   missense    p6  4622601   \n",
       "21767   4627830  arcA   C->T          C          A   missense    p6  4627750   \n",
       "21768   4627933  arcA   T->G          T          C   missense    p6  4627750   \n",
       "21769   4628381  arcA   T->G          T          C   missense    p6  4627750   \n",
       "\n",
       "           End Strand                                            DNA_Seq  \\\n",
       "0          255      +  ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...   \n",
       "1         2798      +  ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...   \n",
       "2         9189      +  ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...   \n",
       "3        14077      +  ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "4        14077      +  ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "...        ...    ...                                                ...   \n",
       "21765  4621905      -  ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...   \n",
       "21766  4623470      -  ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...   \n",
       "21767  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "21768  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "21769  4628466      -  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "\n",
       "                                                 Prt_Seq  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "...                                                  ...   \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...   \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...   \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "\n",
       "                                            Sequence_Mut  Diff_Count  \\\n",
       "0      ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...           1   \n",
       "1      ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...           1   \n",
       "2      ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...           1   \n",
       "3      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...           1   \n",
       "4      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...           1   \n",
       "...                                                  ...         ...   \n",
       "21765  ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...           1   \n",
       "21766  ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...           1   \n",
       "21767  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "21768  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "21769  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...           1   \n",
       "\n",
       "      Differences  Diff_Pos  has_space  \\\n",
       "0              AC       241      False   \n",
       "1              TG      1615      False   \n",
       "2              TG      8430      False   \n",
       "3              AC     12527      False   \n",
       "4              AC     13974      False   \n",
       "...           ...       ...        ...   \n",
       "21765          AC   4621789      False   \n",
       "21766          AC   4622928      False   \n",
       "21767          GA   4628386      False   \n",
       "21768          AC   4628283      False   \n",
       "21769          AC   4627835      False   \n",
       "\n",
       "                                              WT_aa_made  comparison  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG        True   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...        True   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...        True   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...        True   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...        True   \n",
       "...                                                  ...         ...   \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...        True   \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...        True   \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "\n",
       "                                             Mut_aa_made  \n",
       "0                                 MKRISTTITTTITITTGHGAG   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...  \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  \n",
       "...                                                  ...  \n",
       "21765  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...  \n",
       "21766  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...  \n",
       "21767  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  \n",
       "21768  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  \n",
       "21769  MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...  \n",
       "\n",
       "[19554 rows x 20 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "true_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some seq after conversion did not match the NCBI submitted seq , Therefore remove those variants "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame saved to /home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_Strand_Variant_MutDNA_True.xlsx\n"
     ]
    }
   ],
   "source": [
    "# Write in local \n",
    "# Specify the file path where you want to save the Excel file\n",
    "excel_file_path = '/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_Strand_Variant_MutDNA_True.xlsx'\n",
    "\n",
    "# Write the DataFrame to an Excel file\n",
    "true_df.to_excel(excel_file_path, index=False)\n",
    "\n",
    "print(f'DataFrame saved to {excel_file_path}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Take this sheet to R to get the SNP between WT prt and Mut prt "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### All elements in place we can pull the ESM1b score to the variant mastersheet "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the esm1b output score \n",
    "import pandas as pd\n",
    "score = pd.read_csv('/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/esm1b_output.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make universal ID \n",
    "score['universal_id'] = score['seq_id'].astype(str) + '_' + score['mut_name'].astype(str)\n",
    "score['universal_id'] = score['universal_id'].str.replace(' ', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq_id</th>\n",
       "      <th>mut_name</th>\n",
       "      <th>esm_score</th>\n",
       "      <th>universal_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241_thrL</td>\n",
       "      <td>M1K</td>\n",
       "      <td>-7.278638</td>\n",
       "      <td>241_thrL_M1K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>241_thrL</td>\n",
       "      <td>M1R</td>\n",
       "      <td>-7.132564</td>\n",
       "      <td>241_thrL_M1R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>241_thrL</td>\n",
       "      <td>M1H</td>\n",
       "      <td>-8.481606</td>\n",
       "      <td>241_thrL_M1H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>241_thrL</td>\n",
       "      <td>M1E</td>\n",
       "      <td>-7.796053</td>\n",
       "      <td>241_thrL_M1E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>241_thrL</td>\n",
       "      <td>M1D</td>\n",
       "      <td>-8.461714</td>\n",
       "      <td>241_thrL_M1D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308506095</th>\n",
       "      <td>4629600_lasT</td>\n",
       "      <td>K228M</td>\n",
       "      <td>-8.520492</td>\n",
       "      <td>4629600_lasT_K228M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308506096</th>\n",
       "      <td>4629600_lasT</td>\n",
       "      <td>K228P</td>\n",
       "      <td>-5.068690</td>\n",
       "      <td>4629600_lasT_K228P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308506097</th>\n",
       "      <td>4629600_lasT</td>\n",
       "      <td>K228Y</td>\n",
       "      <td>-7.547709</td>\n",
       "      <td>4629600_lasT_K228Y</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308506098</th>\n",
       "      <td>4629600_lasT</td>\n",
       "      <td>K228F</td>\n",
       "      <td>-7.534645</td>\n",
       "      <td>4629600_lasT_K228F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308506099</th>\n",
       "      <td>4629600_lasT</td>\n",
       "      <td>K228W</td>\n",
       "      <td>-8.726114</td>\n",
       "      <td>4629600_lasT_K228W</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>308506100 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seq_id mut_name  esm_score        universal_id\n",
       "0              241_thrL      M1K  -7.278638        241_thrL_M1K\n",
       "1              241_thrL      M1R  -7.132564        241_thrL_M1R\n",
       "2              241_thrL      M1H  -8.481606        241_thrL_M1H\n",
       "3              241_thrL      M1E  -7.796053        241_thrL_M1E\n",
       "4              241_thrL      M1D  -8.461714        241_thrL_M1D\n",
       "...                 ...      ...        ...                 ...\n",
       "308506095  4629600_lasT    K228M  -8.520492  4629600_lasT_K228M\n",
       "308506096  4629600_lasT    K228P  -5.068690  4629600_lasT_K228P\n",
       "308506097  4629600_lasT    K228Y  -7.547709  4629600_lasT_K228Y\n",
       "308506098  4629600_lasT    K228F  -7.534645  4629600_lasT_K228F\n",
       "308506099  4629600_lasT    K228W  -8.726114  4629600_lasT_K228W\n",
       "\n",
       "[308506100 rows x 4 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read the dataframe with SNP preporcesd in r \n",
    "true_df = pd.read_excel('/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/DNA_Prt_Strand_Variant_MutDNA_True_SNP.xlsx')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make Seq Id in mastersheet to pull scores \n",
    "true_df['seq_id'] = true_df['Position'].astype(str) + '_' + true_df['Gene'].astype(str)\n",
    "true_df['universal_id'] = true_df['seq_id'].astype(str) + '_' + true_df['SNP'].astype(str)\n",
    "true_df['universal_id'] = true_df['universal_id'].str.replace(' ', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>...</th>\n",
       "      <th>Diff_Count</th>\n",
       "      <th>Differences</th>\n",
       "      <th>Diff_Pos</th>\n",
       "      <th>has_space</th>\n",
       "      <th>WT_aa_made</th>\n",
       "      <th>comparison</th>\n",
       "      <th>Mut_aa_made</th>\n",
       "      <th>SNP</th>\n",
       "      <th>seq_id</th>\n",
       "      <th>universal_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>241</td>\n",
       "      <td>False</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>True</td>\n",
       "      <td>MKRISTTITTTITITTGHGAG</td>\n",
       "      <td>N18H</td>\n",
       "      <td>241_thrL</td>\n",
       "      <td>241_thrL_N18H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>336</td>\n",
       "      <td>2798</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>1615</td>\n",
       "      <td>False</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>True</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>I427S</td>\n",
       "      <td>1615_thrA</td>\n",
       "      <td>1615_thrA_I427S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>8236</td>\n",
       "      <td>9189</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>TG</td>\n",
       "      <td>8430</td>\n",
       "      <td>False</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>True</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>D65E</td>\n",
       "      <td>8430_talB</td>\n",
       "      <td>8430_talB_D65E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>12527</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>M123L</td>\n",
       "      <td>12527_dnaK</td>\n",
       "      <td>12527_dnaK_M123L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>13974</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>Q605P</td>\n",
       "      <td>13974_dnaK</td>\n",
       "      <td>13974_dnaK_Q605P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19549</th>\n",
       "      <td>4621500</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4621384</td>\n",
       "      <td>4621905</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4621789</td>\n",
       "      <td>False</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>True</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>K136Q</td>\n",
       "      <td>4621500_yjjX</td>\n",
       "      <td>4621500_yjjX_K136Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19550</th>\n",
       "      <td>4623143</td>\n",
       "      <td>rob</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4622601</td>\n",
       "      <td>4623470</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4622928</td>\n",
       "      <td>False</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>True</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>S110R</td>\n",
       "      <td>4623143_rob</td>\n",
       "      <td>4623143_rob_S110R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19551</th>\n",
       "      <td>4627830</td>\n",
       "      <td>arcA</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>GA</td>\n",
       "      <td>4628386</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>E213K</td>\n",
       "      <td>4627830_arcA</td>\n",
       "      <td>4627830_arcA_E213K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19552</th>\n",
       "      <td>4627933</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4628283</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>K178N</td>\n",
       "      <td>4627933_arcA</td>\n",
       "      <td>4627933_arcA_K178N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19553</th>\n",
       "      <td>4628381</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>AC</td>\n",
       "      <td>4627835</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...</td>\n",
       "      <td>D29A</td>\n",
       "      <td>4628381_arcA</td>\n",
       "      <td>4628381_arcA_D29A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19554 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position  Gene Allele Ref_allele Alt_allele Annotation label    Start  \\\n",
       "0           241  thrL   A->C          A          C   missense    m1      190   \n",
       "1          1615  thrA   T->G          T          G   missense    m1      336   \n",
       "2          8430  talB   T->G          T          G   missense    m1     8236   \n",
       "3         12527  dnaK   A->C          A          C   missense    m1    12161   \n",
       "4         13974  dnaK   A->C          A          C   missense    m1    12161   \n",
       "...         ...   ...    ...        ...        ...        ...   ...      ...   \n",
       "19549   4621500  yjjX   T->G          T          C   missense    p6  4621384   \n",
       "19550   4623143   rob   T->G          T          C   missense    p6  4622601   \n",
       "19551   4627830  arcA   C->T          C          A   missense    p6  4627750   \n",
       "19552   4627933  arcA   T->G          T          C   missense    p6  4627750   \n",
       "19553   4628381  arcA   T->G          T          C   missense    p6  4627750   \n",
       "\n",
       "           End Strand  ... Diff_Count Differences Diff_Pos  has_space  \\\n",
       "0          255      +  ...          1          AC      241      False   \n",
       "1         2798      +  ...          1          TG     1615      False   \n",
       "2         9189      +  ...          1          TG     8430      False   \n",
       "3        14077      +  ...          1          AC    12527      False   \n",
       "4        14077      +  ...          1          AC    13974      False   \n",
       "...        ...    ...  ...        ...         ...      ...        ...   \n",
       "19549  4621905      -  ...          1          AC  4621789      False   \n",
       "19550  4623470      -  ...          1          AC  4622928      False   \n",
       "19551  4628466      -  ...          1          GA  4628386      False   \n",
       "19552  4628466      -  ...          1          AC  4628283      False   \n",
       "19553  4628466      -  ...          1          AC  4627835      False   \n",
       "\n",
       "                                              WT_aa_made  comparison  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG        True   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...        True   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...        True   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...        True   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...        True   \n",
       "...                                                  ...         ...   \n",
       "19549  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...        True   \n",
       "19550  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...        True   \n",
       "19551  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "19552  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "19553  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...        True   \n",
       "\n",
       "                                             Mut_aa_made    SNP        seq_id  \\\n",
       "0                                  MKRISTTITTTITITTGHGAG   N18H      241_thrL   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  I427S     1615_thrA   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   D65E     8430_talB   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  M123L    12527_dnaK   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  Q605P    13974_dnaK   \n",
       "...                                                  ...    ...           ...   \n",
       "19549  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...  K136Q  4621500_yjjX   \n",
       "19550  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...  S110R   4623143_rob   \n",
       "19551  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  E213K  4627830_arcA   \n",
       "19552  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  K178N  4627933_arcA   \n",
       "19553  MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...   D29A  4628381_arcA   \n",
       "\n",
       "             universal_id  \n",
       "0           241_thrL_N18H  \n",
       "1         1615_thrA_I427S  \n",
       "2          8430_talB_D65E  \n",
       "3        12527_dnaK_M123L  \n",
       "4        13974_dnaK_Q605P  \n",
       "...                   ...  \n",
       "19549  4621500_yjjX_K136Q  \n",
       "19550   4623143_rob_S110R  \n",
       "19551  4627830_arcA_E213K  \n",
       "19552  4627933_arcA_K178N  \n",
       "19553   4628381_arcA_D29A  \n",
       "\n",
       "[19554 rows x 23 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "true_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge \n",
    "final_df = true_df.merge(score, on='universal_id', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>...</th>\n",
       "      <th>has_space</th>\n",
       "      <th>WT_aa_made</th>\n",
       "      <th>comparison</th>\n",
       "      <th>Mut_aa_made</th>\n",
       "      <th>SNP</th>\n",
       "      <th>seq_id_x</th>\n",
       "      <th>universal_id</th>\n",
       "      <th>seq_id_y</th>\n",
       "      <th>mut_name</th>\n",
       "      <th>esm_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>True</td>\n",
       "      <td>MKRISTTITTTITITTGHGAG</td>\n",
       "      <td>N18H</td>\n",
       "      <td>241_thrL</td>\n",
       "      <td>241_thrL_N18H</td>\n",
       "      <td>241_thrL</td>\n",
       "      <td>N18H</td>\n",
       "      <td>-4.279933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>336</td>\n",
       "      <td>2798</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>True</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>I427S</td>\n",
       "      <td>1615_thrA</td>\n",
       "      <td>1615_thrA_I427S</td>\n",
       "      <td>1615_thrA</td>\n",
       "      <td>I427S</td>\n",
       "      <td>-11.530375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>8236</td>\n",
       "      <td>9189</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>True</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>D65E</td>\n",
       "      <td>8430_talB</td>\n",
       "      <td>8430_talB_D65E</td>\n",
       "      <td>8430_talB</td>\n",
       "      <td>D65E</td>\n",
       "      <td>-5.331029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>M123L</td>\n",
       "      <td>12527_dnaK</td>\n",
       "      <td>12527_dnaK_M123L</td>\n",
       "      <td>12527_dnaK</td>\n",
       "      <td>M123L</td>\n",
       "      <td>-4.669276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>Q605P</td>\n",
       "      <td>13974_dnaK</td>\n",
       "      <td>13974_dnaK_Q605P</td>\n",
       "      <td>13974_dnaK</td>\n",
       "      <td>Q605P</td>\n",
       "      <td>-7.992450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24078</th>\n",
       "      <td>4623143</td>\n",
       "      <td>rob</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4622601</td>\n",
       "      <td>4623470</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>True</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>S110R</td>\n",
       "      <td>4623143_rob</td>\n",
       "      <td>4623143_rob_S110R</td>\n",
       "      <td>4623143_rob</td>\n",
       "      <td>S110R</td>\n",
       "      <td>-5.109406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24079</th>\n",
       "      <td>4627830</td>\n",
       "      <td>arcA</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>E213K</td>\n",
       "      <td>4627830_arcA</td>\n",
       "      <td>4627830_arcA_E213K</td>\n",
       "      <td>4627830_arcA</td>\n",
       "      <td>E213K</td>\n",
       "      <td>-6.002205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24080</th>\n",
       "      <td>4627830</td>\n",
       "      <td>arcA</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>E213K</td>\n",
       "      <td>4627830_arcA</td>\n",
       "      <td>4627830_arcA_E213K</td>\n",
       "      <td>4627830_arcA</td>\n",
       "      <td>E213K</td>\n",
       "      <td>-6.002205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24081</th>\n",
       "      <td>4627933</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>K178N</td>\n",
       "      <td>4627933_arcA</td>\n",
       "      <td>4627933_arcA_K178N</td>\n",
       "      <td>4627933_arcA</td>\n",
       "      <td>K178N</td>\n",
       "      <td>-4.694204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24082</th>\n",
       "      <td>4628381</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>True</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...</td>\n",
       "      <td>D29A</td>\n",
       "      <td>4628381_arcA</td>\n",
       "      <td>4628381_arcA_D29A</td>\n",
       "      <td>4628381_arcA</td>\n",
       "      <td>D29A</td>\n",
       "      <td>-3.082379</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>24083 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position  Gene Allele Ref_allele Alt_allele Annotation label    Start  \\\n",
       "0           241  thrL   A->C          A          C   missense    m1      190   \n",
       "1          1615  thrA   T->G          T          G   missense    m1      336   \n",
       "2          8430  talB   T->G          T          G   missense    m1     8236   \n",
       "3         12527  dnaK   A->C          A          C   missense    m1    12161   \n",
       "4         13974  dnaK   A->C          A          C   missense    m1    12161   \n",
       "...         ...   ...    ...        ...        ...        ...   ...      ...   \n",
       "24078   4623143   rob   T->G          T          C   missense    p6  4622601   \n",
       "24079   4627830  arcA   C->T          C          A   missense    p6  4627750   \n",
       "24080   4627830  arcA   C->T          C          A   missense    p6  4627750   \n",
       "24081   4627933  arcA   T->G          T          C   missense    p6  4627750   \n",
       "24082   4628381  arcA   T->G          T          C   missense    p6  4627750   \n",
       "\n",
       "           End Strand  ... has_space  \\\n",
       "0          255      +  ...     False   \n",
       "1         2798      +  ...     False   \n",
       "2         9189      +  ...     False   \n",
       "3        14077      +  ...     False   \n",
       "4        14077      +  ...     False   \n",
       "...        ...    ...  ...       ...   \n",
       "24078  4623470      -  ...     False   \n",
       "24079  4628466      -  ...     False   \n",
       "24080  4628466      -  ...     False   \n",
       "24081  4628466      -  ...     False   \n",
       "24082  4628466      -  ...     False   \n",
       "\n",
       "                                              WT_aa_made comparison  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG       True   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...       True   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...       True   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...       True   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...       True   \n",
       "...                                                  ...        ...   \n",
       "24078  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...       True   \n",
       "24079  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...       True   \n",
       "24080  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...       True   \n",
       "24081  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...       True   \n",
       "24082  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...       True   \n",
       "\n",
       "                                             Mut_aa_made    SNP      seq_id_x  \\\n",
       "0                                  MKRISTTITTTITITTGHGAG   N18H      241_thrL   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  I427S     1615_thrA   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   D65E     8430_talB   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  M123L    12527_dnaK   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  Q605P    13974_dnaK   \n",
       "...                                                  ...    ...           ...   \n",
       "24078  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...  S110R   4623143_rob   \n",
       "24079  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  E213K  4627830_arcA   \n",
       "24080  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  E213K  4627830_arcA   \n",
       "24081  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  K178N  4627933_arcA   \n",
       "24082  MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...   D29A  4628381_arcA   \n",
       "\n",
       "             universal_id      seq_id_y  mut_name  esm_score  \n",
       "0           241_thrL_N18H      241_thrL      N18H  -4.279933  \n",
       "1         1615_thrA_I427S     1615_thrA     I427S -11.530375  \n",
       "2          8430_talB_D65E     8430_talB      D65E  -5.331029  \n",
       "3        12527_dnaK_M123L    12527_dnaK     M123L  -4.669276  \n",
       "4        13974_dnaK_Q605P    13974_dnaK     Q605P  -7.992450  \n",
       "...                   ...           ...       ...        ...  \n",
       "24078   4623143_rob_S110R   4623143_rob     S110R  -5.109406  \n",
       "24079  4627830_arcA_E213K  4627830_arcA     E213K  -6.002205  \n",
       "24080  4627830_arcA_E213K  4627830_arcA     E213K  -6.002205  \n",
       "24081  4627933_arcA_K178N  4627933_arcA     K178N  -4.694204  \n",
       "24082   4628381_arcA_D29A  4628381_arcA      D29A  -3.082379  \n",
       "\n",
       "[24083 rows x 26 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate rows based on 'universal_id': 5765\n"
     ]
    }
   ],
   "source": [
    "# Count duplicate rows based on 'key_col'\n",
    "duplicate_count = final_df.duplicated(subset=['universal_id']).sum()\n",
    "\n",
    "print(\"Number of duplicate rows based on 'universal_id':\", duplicate_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of duplicate rows based on 'universal_id': 5765\n",
      "DataFrame after dropping duplicates:\n",
      "       Position  Gene Allele Ref_allele Alt_allele Annotation label    Start  \\\n",
      "0           241  thrL   A->C          A          C   missense    m1      190   \n",
      "1          1615  thrA   T->G          T          G   missense    m1      336   \n",
      "2          8430  talB   T->G          T          G   missense    m1     8236   \n",
      "3         12527  dnaK   A->C          A          C   missense    m1    12161   \n",
      "4         13974  dnaK   A->C          A          C   missense    m1    12161   \n",
      "...         ...   ...    ...        ...        ...        ...   ...      ...   \n",
      "24076   4612040  lplA   T->G          T          C   missense    p6  4611315   \n",
      "24077   4621500  yjjX   T->G          T          C   missense    p6  4621384   \n",
      "24078   4623143   rob   T->G          T          C   missense    p6  4622601   \n",
      "24081   4627933  arcA   T->G          T          C   missense    p6  4627750   \n",
      "24082   4628381  arcA   T->G          T          C   missense    p6  4627750   \n",
      "\n",
      "           End Strand  ... has_space  \\\n",
      "0          255      +  ...     False   \n",
      "1         2798      +  ...     False   \n",
      "2         9189      +  ...     False   \n",
      "3        14077      +  ...     False   \n",
      "4        14077      +  ...     False   \n",
      "...        ...    ...  ...       ...   \n",
      "24076  4612331      -  ...     False   \n",
      "24077  4621905      -  ...     False   \n",
      "24078  4623470      -  ...     False   \n",
      "24081  4628466      -  ...     False   \n",
      "24082  4628466      -  ...     False   \n",
      "\n",
      "                                              WT_aa_made comparison  \\\n",
      "0                                  MKRISTTITTTITITTGNGAG       True   \n",
      "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...       True   \n",
      "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...       True   \n",
      "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...       True   \n",
      "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...       True   \n",
      "...                                                  ...        ...   \n",
      "24076  MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...       True   \n",
      "24077  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...       True   \n",
      "24078  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...       True   \n",
      "24081  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...       True   \n",
      "24082  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...       True   \n",
      "\n",
      "                                             Mut_aa_made    SNP      seq_id_x  \\\n",
      "0                                  MKRISTTITTTITITTGHGAG   N18H      241_thrL   \n",
      "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  I427S     1615_thrA   \n",
      "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   D65E     8430_talB   \n",
      "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  M123L    12527_dnaK   \n",
      "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  Q605P    13974_dnaK   \n",
      "...                                                  ...    ...           ...   \n",
      "24076  MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...   K98Q  4612040_lplA   \n",
      "24077  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...  K136Q  4621500_yjjX   \n",
      "24078  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...  S110R   4623143_rob   \n",
      "24081  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  K178N  4627933_arcA   \n",
      "24082  MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...   D29A  4628381_arcA   \n",
      "\n",
      "             universal_id      seq_id_y  mut_name  esm_score  \n",
      "0           241_thrL_N18H      241_thrL      N18H  -4.279933  \n",
      "1         1615_thrA_I427S     1615_thrA     I427S -11.530375  \n",
      "2          8430_talB_D65E     8430_talB      D65E  -5.331029  \n",
      "3        12527_dnaK_M123L    12527_dnaK     M123L  -4.669276  \n",
      "4        13974_dnaK_Q605P    13974_dnaK     Q605P  -7.992450  \n",
      "...                   ...           ...       ...        ...  \n",
      "24076   4612040_lplA_K98Q  4612040_lplA      K98Q  -8.282178  \n",
      "24077  4621500_yjjX_K136Q  4621500_yjjX     K136Q  -5.321062  \n",
      "24078   4623143_rob_S110R   4623143_rob     S110R  -5.109406  \n",
      "24081  4627933_arcA_K178N  4627933_arcA     K178N  -4.694204  \n",
      "24082   4628381_arcA_D29A  4628381_arcA      D29A  -3.082379  \n",
      "\n",
      "[18318 rows x 26 columns]\n"
     ]
    }
   ],
   "source": [
    "# Count duplicate rows based on 'universal_id'\n",
    "duplicate_count = final_df.duplicated(subset=['universal_id']).sum()\n",
    "\n",
    "# Drop duplicate rows based on 'universal_id'\n",
    "final_df = final_df.drop_duplicates(subset=['universal_id'])\n",
    "\n",
    "print(\"Number of duplicate rows based on 'universal_id':\", duplicate_count)\n",
    "print(\"DataFrame after dropping duplicates:\")\n",
    "print(final_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of NaN values in 'esm-score' column: 259\n"
     ]
    }
   ],
   "source": [
    "# Count NaN values in the 'score' column\n",
    "nan_count = final_df['esm_score'].isna().sum()\n",
    "\n",
    "print(\"Number of NaN values in 'esm-score' column:\", nan_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Drop rows with NaN values in the 'score' column\n",
    "final_df = final_df.dropna(subset=['esm_score'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select the columns you want to keep\n",
    "Score_missense = final_df[['Position', 'Gene', 'Allele', 'Ref_allele', 'Alt_allele', 'Annotation',\n",
    "       'label', 'Start', 'End', 'Strand', 'DNA_Seq', 'Prt_Seq', 'Sequence_Mut', 'WT_aa_made','Mut_aa_made', 'SNP', 'seq_id_x', 'universal_id',\n",
    "       'seq_id_y', 'mut_name', 'esm_score']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>...</th>\n",
       "      <th>Prt_Seq</th>\n",
       "      <th>Sequence_Mut</th>\n",
       "      <th>WT_aa_made</th>\n",
       "      <th>Mut_aa_made</th>\n",
       "      <th>SNP</th>\n",
       "      <th>seq_id_x</th>\n",
       "      <th>universal_id</th>\n",
       "      <th>seq_id_y</th>\n",
       "      <th>mut_name</th>\n",
       "      <th>esm_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>MKRISTTITTTITITTGHGAG</td>\n",
       "      <td>N18H</td>\n",
       "      <td>241_thrL</td>\n",
       "      <td>241_thrL_N18H</td>\n",
       "      <td>241_thrL</td>\n",
       "      <td>N18H</td>\n",
       "      <td>-4.279933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>336</td>\n",
       "      <td>2798</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>I427S</td>\n",
       "      <td>1615_thrA</td>\n",
       "      <td>1615_thrA_I427S</td>\n",
       "      <td>1615_thrA</td>\n",
       "      <td>I427S</td>\n",
       "      <td>-11.530375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>8236</td>\n",
       "      <td>9189</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>D65E</td>\n",
       "      <td>8430_talB</td>\n",
       "      <td>8430_talB_D65E</td>\n",
       "      <td>8430_talB</td>\n",
       "      <td>D65E</td>\n",
       "      <td>-5.331029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>M123L</td>\n",
       "      <td>12527_dnaK</td>\n",
       "      <td>12527_dnaK_M123L</td>\n",
       "      <td>12527_dnaK</td>\n",
       "      <td>M123L</td>\n",
       "      <td>-4.669276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>Q605P</td>\n",
       "      <td>13974_dnaK</td>\n",
       "      <td>13974_dnaK_Q605P</td>\n",
       "      <td>13974_dnaK</td>\n",
       "      <td>Q605P</td>\n",
       "      <td>-7.992450</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24076</th>\n",
       "      <td>4612040</td>\n",
       "      <td>lplA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4611315</td>\n",
       "      <td>4612331</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...</td>\n",
       "      <td>ATGTCCACATTACGCCTGCTCATCTCTGACTCTTACGACCCGTGGT...</td>\n",
       "      <td>MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...</td>\n",
       "      <td>MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...</td>\n",
       "      <td>K98Q</td>\n",
       "      <td>4612040_lplA</td>\n",
       "      <td>4612040_lplA_K98Q</td>\n",
       "      <td>4612040_lplA</td>\n",
       "      <td>K98Q</td>\n",
       "      <td>-8.282178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24077</th>\n",
       "      <td>4621500</td>\n",
       "      <td>yjjX</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4621384</td>\n",
       "      <td>4621905</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...</td>\n",
       "      <td>K136Q</td>\n",
       "      <td>4621500_yjjX</td>\n",
       "      <td>4621500_yjjX_K136Q</td>\n",
       "      <td>4621500_yjjX</td>\n",
       "      <td>K136Q</td>\n",
       "      <td>-5.321062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24078</th>\n",
       "      <td>4623143</td>\n",
       "      <td>rob</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4622601</td>\n",
       "      <td>4623470</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...</td>\n",
       "      <td>S110R</td>\n",
       "      <td>4623143_rob</td>\n",
       "      <td>4623143_rob_S110R</td>\n",
       "      <td>4623143_rob</td>\n",
       "      <td>S110R</td>\n",
       "      <td>-5.109406</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24081</th>\n",
       "      <td>4627933</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>K178N</td>\n",
       "      <td>4627933_arcA</td>\n",
       "      <td>4627933_arcA_K178N</td>\n",
       "      <td>4627933_arcA</td>\n",
       "      <td>K178N</td>\n",
       "      <td>-4.694204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24082</th>\n",
       "      <td>4628381</td>\n",
       "      <td>arcA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>4627750</td>\n",
       "      <td>4628466</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...</td>\n",
       "      <td>MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...</td>\n",
       "      <td>D29A</td>\n",
       "      <td>4628381_arcA</td>\n",
       "      <td>4628381_arcA_D29A</td>\n",
       "      <td>4628381_arcA</td>\n",
       "      <td>D29A</td>\n",
       "      <td>-3.082379</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18059 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position  Gene Allele Ref_allele Alt_allele Annotation label    Start  \\\n",
       "0           241  thrL   A->C          A          C   missense    m1      190   \n",
       "1          1615  thrA   T->G          T          G   missense    m1      336   \n",
       "2          8430  talB   T->G          T          G   missense    m1     8236   \n",
       "3         12527  dnaK   A->C          A          C   missense    m1    12161   \n",
       "4         13974  dnaK   A->C          A          C   missense    m1    12161   \n",
       "...         ...   ...    ...        ...        ...        ...   ...      ...   \n",
       "24076   4612040  lplA   T->G          T          C   missense    p6  4611315   \n",
       "24077   4621500  yjjX   T->G          T          C   missense    p6  4621384   \n",
       "24078   4623143   rob   T->G          T          C   missense    p6  4622601   \n",
       "24081   4627933  arcA   T->G          T          C   missense    p6  4627750   \n",
       "24082   4628381  arcA   T->G          T          C   missense    p6  4627750   \n",
       "\n",
       "           End Strand  ...                                            Prt_Seq  \\\n",
       "0          255      +  ...                              MKRISTTITTTITITTGNGAG   \n",
       "1         2798      +  ...  MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
       "2         9189      +  ...  MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   \n",
       "3        14077      +  ...  MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "4        14077      +  ...  MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "...        ...    ...  ...                                                ...   \n",
       "24076  4612331      -  ...  MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...   \n",
       "24077  4621905      -  ...  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...   \n",
       "24078  4623470      -  ...  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...   \n",
       "24081  4628466      -  ...  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "24082  4628466      -  ...  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "\n",
       "                                            Sequence_Mut  \\\n",
       "0      ATGAAACGCATTAGCACCACCATTACCACCACCATCACCATTACCA...   \n",
       "1      ATGCGAGTGTTGAAGTTCGGCGGTACATCAGTGGCAAATGCAGAAC...   \n",
       "2      ATGACGGACAAATTGACCTCCCTTCGTCAGTACACCACCGTAGTGG...   \n",
       "3      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "4      ATGGGTAAAATAATTGGTATCGACCTGGGTACTACCAACTCTTGTG...   \n",
       "...                                                  ...   \n",
       "24076  ATGTCCACATTACGCCTGCTCATCTCTGACTCTTACGACCCGTGGT...   \n",
       "24077  ATGCACCAAGTTGTCTGTGCGACCACCAATCCCGCTAAAATTCAGG...   \n",
       "24078  ATGGATCAGGCCGGCATTATTCGCGACCTTTTAATCTGGCTGGAAG...   \n",
       "24081  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "24082  ATGCAGACCCCGCACATTCTTATCGTTGAAAACGAGTTGGTAACAC...   \n",
       "\n",
       "                                              WT_aa_made  \\\n",
       "0                                  MKRISTTITTTITITTGNGAG   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "...                                                  ...   \n",
       "24076  MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...   \n",
       "24077  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...   \n",
       "24078  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...   \n",
       "24081  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "24082  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...   \n",
       "\n",
       "                                             Mut_aa_made    SNP      seq_id_x  \\\n",
       "0                                  MKRISTTITTTITITTGHGAG   N18H      241_thrL   \n",
       "1      MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  I427S     1615_thrA   \n",
       "2      MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   D65E     8430_talB   \n",
       "3      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  M123L    12527_dnaK   \n",
       "4      MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  Q605P    13974_dnaK   \n",
       "...                                                  ...    ...           ...   \n",
       "24076  MSTLRLLISDSYDPWFNLAVEECIFRQMPATQRVLFLWRNADTVVI...   K98Q  4612040_lplA   \n",
       "24077  MHQVVCATTNPAKIQAILQAFHEIFGEGSCHIASVAVESGVPEQPF...  K136Q  4621500_yjjX   \n",
       "24078  MDQAGIIRDLLIWLEGHLDQPLSLDNVAAKAGYSKWHLQRMFKDVT...  S110R   4623143_rob   \n",
       "24081  MQTPHILIVENELVTRNTLKSIFEAEGYDVFEATDGAEMHQILSEY...  K178N  4627933_arcA   \n",
       "24082  MQTPHILIVENELVTRNTLKSIFEAEGYAVFEATDGAEMHQILSEY...   D29A  4628381_arcA   \n",
       "\n",
       "             universal_id      seq_id_y mut_name  esm_score  \n",
       "0           241_thrL_N18H      241_thrL     N18H  -4.279933  \n",
       "1         1615_thrA_I427S     1615_thrA    I427S -11.530375  \n",
       "2          8430_talB_D65E     8430_talB     D65E  -5.331029  \n",
       "3        12527_dnaK_M123L    12527_dnaK    M123L  -4.669276  \n",
       "4        13974_dnaK_Q605P    13974_dnaK    Q605P  -7.992450  \n",
       "...                   ...           ...      ...        ...  \n",
       "24076   4612040_lplA_K98Q  4612040_lplA     K98Q  -8.282178  \n",
       "24077  4621500_yjjX_K136Q  4621500_yjjX    K136Q  -5.321062  \n",
       "24078   4623143_rob_S110R   4623143_rob    S110R  -5.109406  \n",
       "24081  4627933_arcA_K178N  4627933_arcA    K178N  -4.694204  \n",
       "24082   4628381_arcA_D29A  4628381_arcA     D29A  -3.082379  \n",
       "\n",
       "[18059 rows x 21 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Score_missense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DataFrame saved to /home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/Final_esm1b_variant.xlsx\n"
     ]
    }
   ],
   "source": [
    "# Write in local \n",
    "# Specify the file path where you want to save the Excel file\n",
    "excel_file_path = '/home/bernadettem/bernadettenotebook/E.Coli_LTEE/Prt_final/esm1b_redo/Correspondance/Results/Final_esm1b_variant.xlsx'\n",
    "\n",
    "# Write the DataFrame to an Excel file\n",
    "Score_missense.to_excel(excel_file_path, index=False)\n",
    "\n",
    "print(f'DataFrame saved to {excel_file_path}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Few of them got droped out becuase of the stop codon introduced in mutant protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This is the sheet having all the variants and respective esm1b score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### The script for plot generation is in script 3.ipynb file "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "notebook",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
