{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# E. coli LTEE Oligo Promotion"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Read time annotated data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#### Here we used the kaos package to generate all the results \n",
    "#### Description - The annotated population files were taken from github LTEE-metagenomic-master\\data_files (After downloading their entire github). These files were converted to xlxs format and merged. Only the SNPs were filtered and flanks were generated (G:\\My Drive\\PhD Project\\Project #1\\Dry_Lab_analysis\\E. coli Long term evolutionary\\Data\\The Dynamics of Molecular Evolution Over 60k Generations\\code to generate flank.R). These flanks were saved as one excel file that file would be used for gain calculation. \n",
    "# Read the variant SNP with flanks \n",
    "\n",
    "import pandas as pd \n",
    "import numpy as np \n",
    "# Import Datasets\n",
    "import pandas as pd\n",
    "m1_population = pd.read_excel(\"../data/annotated_timecourse/m1_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m1_population[\"label\"] = \"m1\" \n",
    "m1_population = m1_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m2_population = pd.read_excel(\"../data/annotated_timecourse/m2_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m2_population[\"label\"] = \"m2\" \n",
    "m2_population = m2_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m3_population = pd.read_excel(\"../data/annotated_timecourse/m3_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m3_population[\"label\"] = \"m3\" \n",
    "m3_population = m3_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m4_population = pd.read_excel(\"../data/annotated_timecourse/m4_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m4_population[\"label\"] = \"m4\" \n",
    "m4_population = m4_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m5_population = pd.read_excel(\"../data/annotated_timecourse/m5_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m5_population[\"label\"] = \"m5\" \n",
    "m5_population = m5_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m6_population = pd.read_excel(\"../data/annotated_timecourse/m6_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m6_population[\"label\"] = \"m6\" \n",
    "m6_population = m6_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p1_population = pd.read_excel(\"../data/annotated_timecourse/p1_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p1_population[\"label\"] = \"p1\" \n",
    "p1_population = p1_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p2_population = pd.read_excel(\"../data/annotated_timecourse/p2_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p2_population[\"label\"] = \"p2\" \n",
    "p2_population = p2_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p3_population = pd.read_excel(\"../data/annotated_timecourse/p3_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p3_population[\"label\"] = \"p3\" \n",
    "p3_population = p3_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p4_population = pd.read_excel(\"../data/annotated_timecourse/p4_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p4_population[\"label\"] = \"p4\" \n",
    "p4_population = p4_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p5_population = pd.read_excel(\"../data/annotated_timecourse/p5_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p5_population[\"label\"] = \"p5\" \n",
    "p5_population = p5_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p6_population = pd.read_excel(\"../data/annotated_timecourse/p6_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p6_population[\"label\"] = \"p6\" \n",
    "p6_population = p6_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### concat all the population in single dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Test.statistic</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>ref_flanking_regions</th>\n",
       "      <th>alt_flanking_regions</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>2.887050</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>TACCACAGGTAACGGTGCGGG</td>\n",
       "      <td>TACCACAGGTCACGGTGCGGG</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>4.732560</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GGCTTTTTTTTCGACCAAAGG</td>\n",
       "      <td>GGCTTTTTTTGCGACCAAAGG</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>0.986369</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AATATCAACATTGTCGCCATT</td>\n",
       "      <td>AATATCAACAGTGTCGCCATT</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6046</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>C-&gt;A</td>\n",
       "      <td>missense</td>\n",
       "      <td>1.681940</td>\n",
       "      <td>0.077982</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>TTCCCTCGGGCATTCTCAAGA</td>\n",
       "      <td>TTCCCTCGGGAATTCTCAAGA</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6220</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>2.478710</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AACGCCAGAATCGCCTGGCGG</td>\n",
       "      <td>AACGCCAGAAGCGCCTGGCGG</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position   Gene Allele  Annotation  Test.statistic   P-value Ref_allele  \\\n",
       "0       241   thrL   A->C    missense        2.887050  0.000100          A   \n",
       "1       309   thrA   T->G   noncoding        4.732560  0.000300          T   \n",
       "2      1615   thrA   T->G    missense        0.986369  0.000100          T   \n",
       "3      6046   yaaA   C->A    missense        1.681940  0.077982          C   \n",
       "4      6220   yaaA   T->G    missense        2.478710  0.000100          T   \n",
       "\n",
       "  Alt_allele   ref_flanking_regions   alt_flanking_regions label  \n",
       "0          C  TACCACAGGTAACGGTGCGGG  TACCACAGGTCACGGTGCGGG    m1  \n",
       "1          G  GGCTTTTTTTTCGACCAAAGG  GGCTTTTTTTGCGACCAAAGG    m1  \n",
       "2          G  AATATCAACATTGTCGCCATT  AATATCAACAGTGTCGCCATT    m1  \n",
       "3          A  TTCCCTCGGGCATTCTCAAGA  TTCCCTCGGGAATTCTCAAGA    m1  \n",
       "4          G  AACGCCAGAATCGCCTGGCGG  AACGCCAGAAGCGCCTGGCGG    m1  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.concat([m1_population,m2_population,m3_population,m4_population,m5_population,m6_population,p1_population,p2_population,p3_population,p4_population,p5_population,p6_population])\n",
    "df.reset_index(inplace=True, drop=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### check duplicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Test.statistic</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>ref_flanking_regions</th>\n",
       "      <th>alt_flanking_regions</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Position, Gene, Allele, Annotation, Test.statistic, P-value, Ref_allele, Alt_allele, ref_flanking_regions, alt_flanking_regions, label]\n",
       "Index: []"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df.duplicated(subset=['Position', 'Gene', 'Allele', 'Annotation', 'Test.statistic', 'P-value',\n",
    "       'Ref_allele', 'Alt_allele', 'ref_flanking_regions',\n",
    "       'alt_flanking_regions', 'label'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Test.statistic</th>\n",
       "      <th>P-value</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>ref_flanking_regions</th>\n",
       "      <th>alt_flanking_regions</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>2.887050</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>TACCACAGGTAACGGTGCGGG</td>\n",
       "      <td>TACCACAGGTCACGGTGCGGG</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>4.732560</td>\n",
       "      <td>0.000300</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>GGCTTTTTTTTCGACCAAAGG</td>\n",
       "      <td>GGCTTTTTTTGCGACCAAAGG</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>0.986369</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AATATCAACATTGTCGCCATT</td>\n",
       "      <td>AATATCAACAGTGTCGCCATT</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6046</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>C-&gt;A</td>\n",
       "      <td>missense</td>\n",
       "      <td>1.681940</td>\n",
       "      <td>0.077982</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>TTCCCTCGGGCATTCTCAAGA</td>\n",
       "      <td>TTCCCTCGGGAATTCTCAAGA</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>6220</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>2.478710</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>AACGCCAGAATCGCCTGGCGG</td>\n",
       "      <td>AACGCCAGAAGCGCCTGGCGG</td>\n",
       "      <td>m1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36918</th>\n",
       "      <td>4628507</td>\n",
       "      <td>arcA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>1.517100</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>AACAGGAAGTACAAAAGTCCC</td>\n",
       "      <td>AACAGGAAGTGCAAAAGTCCC</td>\n",
       "      <td>p6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36919</th>\n",
       "      <td>4628701</td>\n",
       "      <td>yjjY</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>5.263680</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ACAGCAACATAACAGGCTAAG</td>\n",
       "      <td>ACAGCAACATCACAGGCTAAG</td>\n",
       "      <td>p6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36920</th>\n",
       "      <td>4628934</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>2.440750</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>ATATGTAGCATTTTTTAACAA</td>\n",
       "      <td>ATATGTAGCAGTTTTTAACAA</td>\n",
       "      <td>p6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36921</th>\n",
       "      <td>4629227</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>2.469960</td>\n",
       "      <td>0.024268</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>ACCTGGAGCCAGCCACCCGCT</td>\n",
       "      <td>ACCTGGAGCCCGCCACCCGCT</td>\n",
       "      <td>p6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36922</th>\n",
       "      <td>4629600</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>1.506350</td>\n",
       "      <td>0.000100</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>GAAAAGTGACACAACGGCAGA</td>\n",
       "      <td>GAAAAGTGACGCAACGGCAGA</td>\n",
       "      <td>p6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36923 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position         Gene Allele   Annotation  Test.statistic   P-value  \\\n",
       "0           241         thrL   A->C     missense        2.887050  0.000100   \n",
       "1           309         thrA   T->G    noncoding        4.732560  0.000300   \n",
       "2          1615         thrA   T->G     missense        0.986369  0.000100   \n",
       "3          6046         yaaA   C->A     missense        1.681940  0.077982   \n",
       "4          6220         yaaA   T->G     missense        2.478710  0.000100   \n",
       "...         ...          ...    ...          ...             ...       ...   \n",
       "36918   4628507         arcA   A->G    noncoding        1.517100  0.000500   \n",
       "36919   4628701         yjjY   A->C     missense        5.263680  0.000100   \n",
       "36920   4628934   intergenic   T->G    noncoding        2.440750  0.000100   \n",
       "36921   4629227         lasT   A->C   synonymous        2.469960  0.024268   \n",
       "36922   4629600         lasT   A->G     missense        1.506350  0.000100   \n",
       "\n",
       "      Ref_allele Alt_allele   ref_flanking_regions   alt_flanking_regions  \\\n",
       "0              A          C  TACCACAGGTAACGGTGCGGG  TACCACAGGTCACGGTGCGGG   \n",
       "1              T          G  GGCTTTTTTTTCGACCAAAGG  GGCTTTTTTTGCGACCAAAGG   \n",
       "2              T          G  AATATCAACATTGTCGCCATT  AATATCAACAGTGTCGCCATT   \n",
       "3              C          A  TTCCCTCGGGCATTCTCAAGA  TTCCCTCGGGAATTCTCAAGA   \n",
       "4              T          G  AACGCCAGAATCGCCTGGCGG  AACGCCAGAAGCGCCTGGCGG   \n",
       "...          ...        ...                    ...                    ...   \n",
       "36918          A          G  AACAGGAAGTACAAAAGTCCC  AACAGGAAGTGCAAAAGTCCC   \n",
       "36919          A          C  ACAGCAACATAACAGGCTAAG  ACAGCAACATCACAGGCTAAG   \n",
       "36920          T          G  ATATGTAGCATTTTTTAACAA  ATATGTAGCAGTTTTTAACAA   \n",
       "36921          A          C  ACCTGGAGCCAGCCACCCGCT  ACCTGGAGCCCGCCACCCGCT   \n",
       "36922          A          G  GAAAAGTGACACAACGGCAGA  GAAAAGTGACGCAACGGCAGA   \n",
       "\n",
       "      label  \n",
       "0        m1  \n",
       "1        m1  \n",
       "2        m1  \n",
       "3        m1  \n",
       "4        m1  \n",
       "...     ...  \n",
       "36918    p6  \n",
       "36919    p6  \n",
       "36920    p6  \n",
       "36921    p6  \n",
       "36922    p6  \n",
       "\n",
       "[36923 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "notebook",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
