{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### MetaData - Prepared manually by taking the Fitness -https://royalsocietypublishing.org/doi/10.1098/rspb.2015.2292 & the allele count data and sequencing data - https://github.com/benjaminhgood/LTEE-metagenomic."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### import libary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"../../..\")\n",
    "# sys.path.append(\"../../../data/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import mannwhitneyu, fisher_exact\n",
    "import utility, importlib\n",
    "from scipy.stats import spearmanr\n",
    "import pingouin as pg\n",
    "from scipy.stats import linregress\n",
    "import warnings\n",
    "import kaos\n",
    "warnings.filterwarnings('ignore', category=pd.errors.PerformanceWarning)\n",
    "pd.set_option('mode.chained_assignment', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import kaos\n",
    "kmer_length =10\n",
    "file_path_Ecoli = \"../../../data/GCF_000017985.1_ASM1798v1_genomic.fna\"\n",
    "seq_fasta = kaos.read_fasta(file_path_Ecoli)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Read gain sheet data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>accumulated_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>63</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>p6</td>\n",
       "      <td>-10.385914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>201</td>\n",
       "      <td>thrL</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>-1.750693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>-0.911836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>m1</td>\n",
       "      <td>4.009052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>322</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>p3</td>\n",
       "      <td>1.532477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position         Gene Allele Ref_allele Alt_allele  Annotation label  \\\n",
       "0        63   intergenic   A->C          A          C   noncoding    p6   \n",
       "1       201         thrL   T->G          T          G    missense    p6   \n",
       "2       241         thrL   A->C          A          C    missense    m1   \n",
       "3       309         thrA   T->G          T          G   noncoding    m1   \n",
       "4       322         thrA   A->G          A          G   noncoding    p3   \n",
       "\n",
       "   accumulated_gain  \n",
       "0        -10.385914  \n",
       "1         -1.750693  \n",
       "2         -0.911836  \n",
       "3          4.009052  \n",
       "4          1.532477  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Gain_calculated = pd.read_excel(\"../../../data/MetaData_ecoli_final.xlsx\", sheet_name= \"Gain score\")\n",
    "df_Gain_calculated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>63</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>201</td>\n",
       "      <td>thrL</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>322</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36916</th>\n",
       "      <td>4629441</td>\n",
       "      <td>lasT</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36917</th>\n",
       "      <td>4629541</td>\n",
       "      <td>lasT</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36918</th>\n",
       "      <td>4629573</td>\n",
       "      <td>lasT</td>\n",
       "      <td>T-&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36919</th>\n",
       "      <td>4629600</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36921</th>\n",
       "      <td>4629764</td>\n",
       "      <td>lasT</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>34391 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position         Gene Allele Ref_allele Alt_allele\n",
       "0            63   intergenic   A->C          A          C\n",
       "1           201         thrL   T->G          T          G\n",
       "2           241         thrL   A->C          A          C\n",
       "3           309         thrA   T->G          T          G\n",
       "4           322         thrA   A->G          A          G\n",
       "...         ...          ...    ...        ...        ...\n",
       "36916   4629441         lasT   C->T          C          T\n",
       "36917   4629541         lasT   C->T          C          T\n",
       "36918   4629573         lasT   T->C          T          C\n",
       "36919   4629600         lasT   A->G          A          G\n",
       "36921   4629764         lasT   T->G          T          G\n",
       "\n",
       "[34391 rows x 5 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Gain_calculated[['Position', 'Gene', 'Allele', 'Ref_allele', 'Alt_allele']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>accumulated_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Position, Gene, Allele, Ref_allele, Alt_allele, Annotation, label, accumulated_gain]\n",
       "Index: []"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Gain_calculated[df_Gain_calculated.duplicated(subset=['Position', 'Gene',\n",
    " 'Allele', 'Annotation', 'Ref_allele', 'Alt_allele','label'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Read Allele count and generation data from metasheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>AC:0</th>\n",
       "      <th>AC:1000</th>\n",
       "      <th>AC:1500</th>\n",
       "      <th>AC:2000</th>\n",
       "      <th>AC:2500</th>\n",
       "      <th>...</th>\n",
       "      <th>AC:57000</th>\n",
       "      <th>AC:57500</th>\n",
       "      <th>AC:58000</th>\n",
       "      <th>AC:58500</th>\n",
       "      <th>AC:59000</th>\n",
       "      <th>AC:59500</th>\n",
       "      <th>AC:60000</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1554</td>\n",
       "      <td>thrA</td>\n",
       "      <td>C-&gt;A</td>\n",
       "      <td>missense</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1556</td>\n",
       "      <td>thrA</td>\n",
       "      <td>C-&gt;A</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2450</td>\n",
       "      <td>thrA</td>\n",
       "      <td>G-&gt;T</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>41</td>\n",
       "      <td>109.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>4274</td>\n",
       "      <td>thrC</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>missense</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>6400</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 125 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM  Position   Gene Allele   Annotation  AC:0  AC:1000  AC:1500  AC:2000  \\\n",
       "0  chrI      1554   thrA   C->A     missense     0        1        0        2   \n",
       "1  chrI      1556   thrA   C->A   synonymous     0        0        0        2   \n",
       "2  chrI      2450   thrA   G->T   synonymous     0        0        0        0   \n",
       "3  chrI      4274   thrC   C->T     missense     0        0        0        0   \n",
       "4  chrI      6400   yaaA   T->G     missense     0        0        0        0   \n",
       "\n",
       "   AC:2500  ...  AC:57000  AC:57500  AC:58000  AC:58500  AC:59000  AC:59500  \\\n",
       "0      1.0  ...         0         0       1.0       0.0       1.0       0.0   \n",
       "1      0.0  ...         0         0       0.0       0.0       0.0       0.0   \n",
       "2      0.0  ...        81        41     109.0     129.0       4.0      67.0   \n",
       "3      0.0  ...         0         0       0.0       0.0       0.0       0.0   \n",
       "4      2.0  ...         0         0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   AC:60000  Ref_allele  Alt_allele  label  \n",
       "0       0.0           C           A     p1  \n",
       "1       0.0           C           A     p1  \n",
       "2       8.0           G           T     p1  \n",
       "3       0.0           C           T     p1  \n",
       "4       0.0           T           G     p1  \n",
       "\n",
       "[5 rows x 125 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Alt_allele_count_all_generation = pd.read_excel(\"../../../data/MetaData_ecoli_final.xlsx\", sheet_name= \"Mastersheet\")\n",
    "df_Alt_allele_count_all_generation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Check if any duplicated row"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>AC:0</th>\n",
       "      <th>AC:1000</th>\n",
       "      <th>AC:1500</th>\n",
       "      <th>AC:2000</th>\n",
       "      <th>AC:2500</th>\n",
       "      <th>...</th>\n",
       "      <th>AC:57000</th>\n",
       "      <th>AC:57500</th>\n",
       "      <th>AC:58000</th>\n",
       "      <th>AC:58500</th>\n",
       "      <th>AC:59000</th>\n",
       "      <th>AC:59500</th>\n",
       "      <th>AC:60000</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2028</th>\n",
       "      <td>chrI</td>\n",
       "      <td>450000</td>\n",
       "      <td>ylaB</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>p3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows Ã— 125 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM  Position   Gene Allele Annotation  AC:0  AC:1000  AC:1500  \\\n",
       "2028  chrI    450000   ylaB   A->G   missense     0        0        0   \n",
       "\n",
       "      AC:2000  AC:2500  ...  AC:57000  AC:57500  AC:58000  AC:58500  AC:59000  \\\n",
       "2028        0      0.0  ...         0         0       0.0       0.0       0.0   \n",
       "\n",
       "      AC:59500  AC:60000  Ref_allele  Alt_allele  label  \n",
       "2028       0.0       0.0           A           G     p3  \n",
       "\n",
       "[1 rows x 125 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Alt_allele_count_all_generation[df_Alt_allele_count_all_generation.duplicated(subset=['Position', 'Gene', 'Allele', 'Annotation', 'Ref_allele', 'Alt_allele','label'])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### delete duplicate row in df_Alt_allele_count_all_generation dataframe where duplicate indicate when ['CHROM','Position', 'Gene', 'Allele', 'Annotation', 'Ref_allele', 'Alt_allele','label'] are same in mutiple row\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "df_Alt_allele_count_all_generation.drop_duplicates(subset=['CHROM','Position', 'Gene', 'Allele', 'Annotation', 'Ref_allele', 'Alt_allele','label'], inplace=True)\n",
    "df_Alt_allele_count_all_generation.reset_index(inplace=True, drop=True)\n",
    "# df_Alt_allele_count_all_generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>AC:0</th>\n",
       "      <th>AC:1000</th>\n",
       "      <th>AC:1500</th>\n",
       "      <th>AC:2000</th>\n",
       "      <th>AC:2500</th>\n",
       "      <th>...</th>\n",
       "      <th>AC:57000</th>\n",
       "      <th>AC:57500</th>\n",
       "      <th>AC:58000</th>\n",
       "      <th>AC:58500</th>\n",
       "      <th>AC:59000</th>\n",
       "      <th>AC:59500</th>\n",
       "      <th>AC:60000</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1554</td>\n",
       "      <td>thrA</td>\n",
       "      <td>C-&gt;A</td>\n",
       "      <td>missense</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1556</td>\n",
       "      <td>thrA</td>\n",
       "      <td>C-&gt;A</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2450</td>\n",
       "      <td>thrA</td>\n",
       "      <td>G-&gt;T</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>81</td>\n",
       "      <td>41</td>\n",
       "      <td>109.0</td>\n",
       "      <td>129.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>4274</td>\n",
       "      <td>thrC</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>missense</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>6400</td>\n",
       "      <td>yaaA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>p1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 125 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  CHROM  Position   Gene Allele   Annotation  AC:0  AC:1000  AC:1500  AC:2000  \\\n",
       "0  chrI      1554   thrA   C->A     missense     0        1        0        2   \n",
       "1  chrI      1556   thrA   C->A   synonymous     0        0        0        2   \n",
       "2  chrI      2450   thrA   G->T   synonymous     0        0        0        0   \n",
       "3  chrI      4274   thrC   C->T     missense     0        0        0        0   \n",
       "4  chrI      6400   yaaA   T->G     missense     0        0        0        0   \n",
       "\n",
       "   AC:2500  ...  AC:57000  AC:57500  AC:58000  AC:58500  AC:59000  AC:59500  \\\n",
       "0      1.0  ...         0         0       1.0       0.0       1.0       0.0   \n",
       "1      0.0  ...         0         0       0.0       0.0       0.0       0.0   \n",
       "2      0.0  ...        81        41     109.0     129.0       4.0      67.0   \n",
       "3      0.0  ...         0         0       0.0       0.0       0.0       0.0   \n",
       "4      2.0  ...         0         0       0.0       0.0       0.0       0.0   \n",
       "\n",
       "   AC:60000  Ref_allele  Alt_allele  label  \n",
       "0       0.0           C           A     p1  \n",
       "1       0.0           C           A     p1  \n",
       "2       8.0           G           T     p1  \n",
       "3       0.0           C           T     p1  \n",
       "4       0.0           T           G     p1  \n",
       "\n",
       "[5 rows x 125 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_Alt_allele_count_all_generation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Read LLR Score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>...</th>\n",
       "      <th>has_space</th>\n",
       "      <th>WT_aa_made2</th>\n",
       "      <th>comparison2</th>\n",
       "      <th>Mut_aa_made</th>\n",
       "      <th>SNP</th>\n",
       "      <th>seq_id_x</th>\n",
       "      <th>universal_id</th>\n",
       "      <th>seq_id_y</th>\n",
       "      <th>mut_name</th>\n",
       "      <th>esm_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>190</td>\n",
       "      <td>255</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MKRISTTITTTITITTGNGAG</td>\n",
       "      <td>True</td>\n",
       "      <td>MKRISTTITTTITITTGHGAG</td>\n",
       "      <td>N18H</td>\n",
       "      <td>241_thrL</td>\n",
       "      <td>241_thrL_N18H</td>\n",
       "      <td>241_thrL</td>\n",
       "      <td>N18H</td>\n",
       "      <td>-4.279933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>336</td>\n",
       "      <td>2798</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>True</td>\n",
       "      <td>MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...</td>\n",
       "      <td>I427S</td>\n",
       "      <td>1615_thrA</td>\n",
       "      <td>1615_thrA_I427S</td>\n",
       "      <td>1615_thrA</td>\n",
       "      <td>I427S</td>\n",
       "      <td>-11.530375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>8236</td>\n",
       "      <td>9189</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>True</td>\n",
       "      <td>MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...</td>\n",
       "      <td>D65E</td>\n",
       "      <td>8430_talB</td>\n",
       "      <td>8430_talB_D65E</td>\n",
       "      <td>8430_talB</td>\n",
       "      <td>D65E</td>\n",
       "      <td>-5.331029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>M123L</td>\n",
       "      <td>12527_dnaK</td>\n",
       "      <td>12527_dnaK_M123L</td>\n",
       "      <td>12527_dnaK</td>\n",
       "      <td>M123L</td>\n",
       "      <td>-4.669276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>12161</td>\n",
       "      <td>14077</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>True</td>\n",
       "      <td>MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...</td>\n",
       "      <td>Q605P</td>\n",
       "      <td>13974_dnaK</td>\n",
       "      <td>13974_dnaK_Q605P</td>\n",
       "      <td>13974_dnaK</td>\n",
       "      <td>Q605P</td>\n",
       "      <td>-7.992450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position  Gene Allele Ref_allele Alt_allele Annotation label  Start    End  \\\n",
       "0       241  thrL   A->C          A          C   missense    m1    190    255   \n",
       "1      1615  thrA   T->G          T          G   missense    m1    336   2798   \n",
       "2      8430  talB   T->G          T          G   missense    m1   8236   9189   \n",
       "3     12527  dnaK   A->C          A          C   missense    m1  12161  14077   \n",
       "4     13974  dnaK   A->C          A          C   missense    m1  12161  14077   \n",
       "\n",
       "  Strand  ... has_space                                        WT_aa_made2  \\\n",
       "0      +  ...     False                              MKRISTTITTTITITTGNGAG   \n",
       "1      +  ...     False  MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...   \n",
       "2      +  ...     False  MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   \n",
       "3      +  ...     False  MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "4      +  ...     False  MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...   \n",
       "\n",
       "  comparison2                                        Mut_aa_made    SNP  \\\n",
       "0        True                              MKRISTTITTTITITTGHGAG   N18H   \n",
       "1        True  MRVLKFGGTSVANAERFLRVADILESNARQGQVATVLSAPAKITNH...  I427S   \n",
       "2        True  MTDKLTSLRQYTTVVADTGDIAAMKLYQPQDATTNPSLILNAAQIP...   D65E   \n",
       "3        True  MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  M123L   \n",
       "4        True  MGKIIGIDLGTTNSCVAIMDGTTPRVLENAEGDRTTPSIIAYTQDG...  Q605P   \n",
       "\n",
       "     seq_id_x      universal_id    seq_id_y  mut_name  esm_score  \n",
       "0    241_thrL     241_thrL_N18H    241_thrL      N18H  -4.279933  \n",
       "1   1615_thrA   1615_thrA_I427S   1615_thrA     I427S -11.530375  \n",
       "2   8430_talB    8430_talB_D65E   8430_talB      D65E  -5.331029  \n",
       "3  12527_dnaK  12527_dnaK_M123L  12527_dnaK     M123L  -4.669276  \n",
       "4  13974_dnaK  13974_dnaK_Q605P  13974_dnaK     Q605P  -7.992450  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_esm = pd.read_excel(\"../../../data/LLR.xlsx\")\n",
    "df_esm.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Drop Duplicates if any in df_esm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>label</th>\n",
       "      <th>esm_score</th>\n",
       "      <th>Allele</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>m1</td>\n",
       "      <td>-4.279933</td>\n",
       "      <td>A-&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1615</td>\n",
       "      <td>thrA</td>\n",
       "      <td>m1</td>\n",
       "      <td>-11.530375</td>\n",
       "      <td>T-&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8430</td>\n",
       "      <td>talB</td>\n",
       "      <td>m1</td>\n",
       "      <td>-5.331029</td>\n",
       "      <td>T-&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12527</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>m1</td>\n",
       "      <td>-4.669276</td>\n",
       "      <td>A-&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>13974</td>\n",
       "      <td>dnaK</td>\n",
       "      <td>m1</td>\n",
       "      <td>-7.992450</td>\n",
       "      <td>A-&gt;C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position  Gene label  esm_score Allele\n",
       "0       241  thrL    m1  -4.279933   A->C\n",
       "1      1615  thrA    m1 -11.530375   T->G\n",
       "2      8430  talB    m1  -5.331029   T->G\n",
       "3     12527  dnaK    m1  -4.669276   A->C\n",
       "4     13974  dnaK    m1  -7.992450   A->C"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_esm_no_duplicates = df_esm[[\"Position\",  \"Gene\", \"label\", \"esm_score\", \"Allele\"]].drop_duplicates(subset= [\"Position\",  \"Gene\", \"label\", \"Allele\"], keep='first')\n",
    "df_esm_no_duplicates.dropna(inplace=True)\n",
    "df_esm_no_duplicates.reset_index(inplace=True, drop=True)\n",
    "df_esm_no_duplicates.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Merge Gain score and allele count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>accumulated_gain</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>AC:0</th>\n",
       "      <th>...</th>\n",
       "      <th>AC:55500</th>\n",
       "      <th>AC:56000</th>\n",
       "      <th>AC:56500</th>\n",
       "      <th>AC:57000</th>\n",
       "      <th>AC:57500</th>\n",
       "      <th>AC:58000</th>\n",
       "      <th>AC:58500</th>\n",
       "      <th>AC:59000</th>\n",
       "      <th>AC:59500</th>\n",
       "      <th>AC:60000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>63</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>p6</td>\n",
       "      <td>-10.385914</td>\n",
       "      <td>chrI</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>49</td>\n",
       "      <td>78</td>\n",
       "      <td>45</td>\n",
       "      <td>22</td>\n",
       "      <td>36</td>\n",
       "      <td>39.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>201</td>\n",
       "      <td>thrL</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>-1.750693</td>\n",
       "      <td>chrI</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>100</td>\n",
       "      <td>140</td>\n",
       "      <td>93</td>\n",
       "      <td>28</td>\n",
       "      <td>79</td>\n",
       "      <td>62.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>-0.911836</td>\n",
       "      <td>chrI</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>m1</td>\n",
       "      <td>4.009052</td>\n",
       "      <td>chrI</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>322</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>p3</td>\n",
       "      <td>1.532477</td>\n",
       "      <td>chrI</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>23</td>\n",
       "      <td>24</td>\n",
       "      <td>13</td>\n",
       "      <td>16</td>\n",
       "      <td>26</td>\n",
       "      <td>8.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 126 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Position         Gene Allele Ref_allele Alt_allele  Annotation label  \\\n",
       "0        63   intergenic   A->C          A          C   noncoding    p6   \n",
       "1       201         thrL   T->G          T          G    missense    p6   \n",
       "2       241         thrL   A->C          A          C    missense    m1   \n",
       "3       309         thrA   T->G          T          G   noncoding    m1   \n",
       "4       322         thrA   A->G          A          G   noncoding    p3   \n",
       "\n",
       "   accumulated_gain CHROM  AC:0  ...  AC:55500  AC:56000  AC:56500  AC:57000  \\\n",
       "0        -10.385914  chrI     0  ...        49        78        45        22   \n",
       "1         -1.750693  chrI     0  ...       100       140        93        28   \n",
       "2         -0.911836  chrI     0  ...         6         1         2         0   \n",
       "3          4.009052  chrI     0  ...         0         0         0         0   \n",
       "4          1.532477  chrI     0  ...        23        24        13        16   \n",
       "\n",
       "   AC:57500  AC:58000  AC:58500  AC:59000  AC:59500  AC:60000  \n",
       "0        36      39.0      11.0       1.0       1.0       6.0  \n",
       "1        79      62.0      15.0       3.0       0.0      13.0  \n",
       "2         0       0.0       0.0       0.0       0.0       0.0  \n",
       "3         0       0.0       1.0       2.0       7.0       9.0  \n",
       "4        26       8.0      30.0      22.0      23.0       0.0  \n",
       "\n",
       "[5 rows x 126 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_complete = pd.merge(df_Gain_calculated,df_Alt_allele_count_all_generation, on=['Position','Gene','Allele','Annotation','Ref_allele','Alt_allele','label'], how = \"left\")\n",
    "df_complete.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Extract the column relavent for current analysis from allele count dataframe (all the generation wise allele count, 'CHROM','Position', 'Gene', 'Allele', 'Annotation', 'Ref_allele', 'Alt_allele','label')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC:0</th>\n",
       "      <th>AC:1000</th>\n",
       "      <th>AC:1500</th>\n",
       "      <th>AC:2000</th>\n",
       "      <th>AC:2500</th>\n",
       "      <th>AC:3000</th>\n",
       "      <th>AC:4000</th>\n",
       "      <th>AC:4500</th>\n",
       "      <th>AC:5000</th>\n",
       "      <th>AC:5500</th>\n",
       "      <th>...</th>\n",
       "      <th>AC:60000</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>label</th>\n",
       "      <th>accumulated_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>chrI</td>\n",
       "      <td>63</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>p6</td>\n",
       "      <td>-10.385914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>chrI</td>\n",
       "      <td>201</td>\n",
       "      <td>thrL</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>p6</td>\n",
       "      <td>-1.750693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>chrI</td>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>m1</td>\n",
       "      <td>-0.911836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>chrI</td>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>m1</td>\n",
       "      <td>4.009052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>chrI</td>\n",
       "      <td>322</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>p3</td>\n",
       "      <td>1.532477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 126 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   AC:0  AC:1000  AC:1500  AC:2000  AC:2500  AC:3000  AC:4000  AC:4500  \\\n",
       "0     0        0        0        0      0.0        0        0        0   \n",
       "1     0        0        0        0      0.0        0        0        0   \n",
       "2     0        0        0        0      0.0        0        0        0   \n",
       "3     0        0        0        0      0.0        0        0        0   \n",
       "4     0        0        0        0      0.0        0        0        0   \n",
       "\n",
       "   AC:5000  AC:5500  ...  AC:60000  CHROM  Position         Gene  Allele  \\\n",
       "0        0        0  ...       6.0   chrI        63   intergenic    A->C   \n",
       "1        0        0  ...      13.0   chrI       201         thrL    T->G   \n",
       "2        0        0  ...       0.0   chrI       241         thrL    A->C   \n",
       "3        0        0  ...       9.0   chrI       309         thrA    T->G   \n",
       "4        0        0  ...       0.0   chrI       322         thrA    A->G   \n",
       "\n",
       "   Annotation  Ref_allele  Alt_allele  label  accumulated_gain  \n",
       "0   noncoding           A           C     p6        -10.385914  \n",
       "1    missense           T           G     p6         -1.750693  \n",
       "2    missense           A           C     m1         -0.911836  \n",
       "3   noncoding           T           G     m1          4.009052  \n",
       "4   noncoding           A           G     p3          1.532477  \n",
       "\n",
       "[5 rows x 126 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_list = [item for item in df_complete if item.startswith(\"AC\")] ## Fiter the column starting with \"AC\" which indicate allele count\n",
    "filtered_list.extend(['CHROM','Position', 'Gene', 'Allele', 'Annotation', 'Ref_allele', 'Alt_allele','label', \"accumulated_gain\"]) ##Add other relavent information\n",
    "df_allele_generation = df_complete[filtered_list]\n",
    "df_allele_generation.head() # Get complete dataframe containing all generation wise allele count and other relavent column"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Define mutator and non-mutator population "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "metadata": {}
   },
   "outputs": [],
   "source": [
    "mutator_list = ['m1', 'm2', 'm3', 'm4', 'p3', 'p6']\n",
    "non_mutator_list = ['p1', 'p2', 'p4', 'p5', 'm5', 'm6']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Find the generation which have missing value "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Number of NaN values in each column:\n",
      "AC:2500      355\n",
      "AC:15000    4953\n",
      "AC:46500     331\n",
      "AC:58000    3753\n",
      "AC:58500     355\n",
      "AC:59000     355\n",
      "AC:59500     355\n",
      "AC:60000     355\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "nan_columns = df_allele_generation.columns[df_allele_generation.isna().any()]\n",
    "df_with_nan = df_allele_generation[nan_columns]\n",
    "nan_counts = df_allele_generation.isna().sum()\n",
    "nan_counts = nan_counts[nan_counts > 0]\n",
    "print(\"\\n Number of NaN values in each column:\")\n",
    "print(nan_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Drop the generation having nan values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AC:0</th>\n",
       "      <th>AC:1000</th>\n",
       "      <th>AC:1500</th>\n",
       "      <th>AC:2000</th>\n",
       "      <th>AC:3000</th>\n",
       "      <th>AC:4000</th>\n",
       "      <th>AC:4500</th>\n",
       "      <th>AC:5000</th>\n",
       "      <th>AC:5500</th>\n",
       "      <th>AC:6000</th>\n",
       "      <th>...</th>\n",
       "      <th>AC:57500</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>label</th>\n",
       "      <th>accumulated_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>36</td>\n",
       "      <td>chrI</td>\n",
       "      <td>63</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>p6</td>\n",
       "      <td>-10.385914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>79</td>\n",
       "      <td>chrI</td>\n",
       "      <td>201</td>\n",
       "      <td>thrL</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>p6</td>\n",
       "      <td>-1.750693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>chrI</td>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>m1</td>\n",
       "      <td>-0.911836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>chrI</td>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>m1</td>\n",
       "      <td>4.009052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>26</td>\n",
       "      <td>chrI</td>\n",
       "      <td>322</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>p3</td>\n",
       "      <td>1.532477</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 118 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   AC:0  AC:1000  AC:1500  AC:2000  AC:3000  AC:4000  AC:4500  AC:5000  \\\n",
       "0     0        0        0        0        0        0        0        0   \n",
       "1     0        0        0        0        0        0        0        0   \n",
       "2     0        0        0        0        0        0        0        0   \n",
       "3     0        0        0        0        0        0        0        0   \n",
       "4     0        0        0        0        0        0        0        0   \n",
       "\n",
       "   AC:5500  AC:6000  ...  AC:57500  CHROM  Position         Gene  Allele  \\\n",
       "0        0        0  ...        36   chrI        63   intergenic    A->C   \n",
       "1        0        0  ...        79   chrI       201         thrL    T->G   \n",
       "2        0        0  ...         0   chrI       241         thrL    A->C   \n",
       "3        0        0  ...         0   chrI       309         thrA    T->G   \n",
       "4        0        0  ...        26   chrI       322         thrA    A->G   \n",
       "\n",
       "   Annotation  Ref_allele  Alt_allele  label  accumulated_gain  \n",
       "0   noncoding           A           C     p6        -10.385914  \n",
       "1    missense           T           G     p6         -1.750693  \n",
       "2    missense           A           C     m1         -0.911836  \n",
       "3   noncoding           T           G     m1          4.009052  \n",
       "4   noncoding           A           G     p3          1.532477  \n",
       "\n",
       "[5 rows x 118 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_allele_generation.drop(nan_counts.index.tolist(), axis = 1, inplace=True)\n",
    "df_allele_generation.reset_index(inplace=True, drop = True)\n",
    "df_allele_generation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Create dataset having all mutation info in proper form for easy calculation\n",
    "##### Make a dataframe having column ['label', 'CHROM', 'Position', 'Gene', 'Allele', 'Annotation', 'generation', 'allele count', 'generation_number']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "metadata": {}
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>generation</th>\n",
       "      <th>AG</th>\n",
       "      <th>allele count</th>\n",
       "      <th>generation_number</th>\n",
       "      <th>mutator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m6</td>\n",
       "      <td>chrI</td>\n",
       "      <td>1457</td>\n",
       "      <td>thrA</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-8.160701</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>non mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>2098</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>0.007780</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>2942</td>\n",
       "      <td>thrB</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-4.953477</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m4</td>\n",
       "      <td>chrI</td>\n",
       "      <td>5437</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-10.279634</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m5</td>\n",
       "      <td>chrI</td>\n",
       "      <td>7129</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>2.182288</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>non mutator</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  label CHROM  Position   Gene Allele Annotation generation         AG  \\\n",
       "0    m6  chrI      1457   thrA   G->A   missense       AC:0  -8.160701   \n",
       "1    m2  chrI      2098   thrA   A->G   missense       AC:0   0.007780   \n",
       "2    m2  chrI      2942   thrB   A->G   missense       AC:0  -4.953477   \n",
       "3    m4  chrI      5437   yaaX   A->G   missense       AC:0 -10.279634   \n",
       "4    m5  chrI      7129   yaaJ   A->C   missense       AC:0   2.182288   \n",
       "\n",
       "   allele count  generation_number      mutator  \n",
       "0             1                  0  non mutator  \n",
       "1             1                  0      mutator  \n",
       "2             1                  0      mutator  \n",
       "3             1                  0      mutator  \n",
       "4             2                  0  non mutator  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_all_copy = df_allele_generation.copy()\n",
    "all_mutation = pd.DataFrame() # The all_mutation dataframe contain generation and population wise accumulated gain\n",
    "\n",
    "for i in data_all_copy.columns[:-9]:\n",
    "    new_df_tmp = pd.DataFrame()\n",
    "    new_df = data_all_copy[data_all_copy.loc[:, i]!=0]\n",
    "    new_df_tmp = new_df[[\"label\", 'CHROM', 'Position', 'Gene', 'Allele', 'Annotation']]\n",
    "    new_df_tmp.loc[:, \"generation\"] = i\n",
    "    new_df_tmp.loc[:, \"AG\"] = new_df.loc[:, \"accumulated_gain\"]\n",
    "    new_df_tmp.loc[:, \"allele count\"] = new_df.loc[:, i]\n",
    "    all_mutation = pd.concat([all_mutation, new_df_tmp], axis = 0)\n",
    "    \n",
    "    \n",
    "all_mutation.reset_index(inplace=True, drop=True)\n",
    "all_mutation['generation_number'] = all_mutation['generation'].str.extract(r'(\\d+)').astype(int)\n",
    "for i in range(all_mutation.shape[0]):\n",
    "    \n",
    "    if all_mutation.loc[i, \"label\"] in mutator_list:\n",
    "        all_mutation.loc[i, \"mutator\"] = \"mutator\"\n",
    "    if all_mutation.loc[i, \"label\"] in non_mutator_list:\n",
    "        all_mutation.loc[i, \"mutator\"] = \"non mutator\"\n",
    "\n",
    "all_mutation.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Remove white spaces from columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_esm_no_duplicates[\"Gene\"] = df_esm_no_duplicates[\"Gene\"].str.strip()\n",
    "all_mutation[\"Gene\"] = all_mutation[\"Gene\"].str.strip()\n",
    "# df_esm_no_duplicates[\"Allele\"] = df_esm_no_duplicates[\"Allele\"].str.strip()\n",
    "all_mutation[\"Allele\"] = all_mutation[\"Allele\"].str.strip()\n",
    "all_mutation[\"Allele\"] = all_mutation[\"Allele\"].str.strip()\n",
    "df_Gain_calculated[\"Allele\"] = df_Gain_calculated[\"Allele\"].str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_mutation[\"Gene\"] = all_mutation[\"Gene\"].str.strip()\n",
    "df_Gain_calculated[\"Gene\"] = df_Gain_calculated[\"Gene\"].str.strip()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>generation</th>\n",
       "      <th>AG</th>\n",
       "      <th>allele count</th>\n",
       "      <th>generation_number</th>\n",
       "      <th>mutator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m6</td>\n",
       "      <td>chrI</td>\n",
       "      <td>1457</td>\n",
       "      <td>thrA</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-8.160701</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>non mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>2098</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>0.007780</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>2942</td>\n",
       "      <td>thrB</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-4.953477</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m4</td>\n",
       "      <td>chrI</td>\n",
       "      <td>5437</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-10.279634</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m5</td>\n",
       "      <td>chrI</td>\n",
       "      <td>7129</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>2.182288</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>non mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048642</th>\n",
       "      <td>m3</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629225</td>\n",
       "      <td>lasT</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>-5.614587</td>\n",
       "      <td>72</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048643</th>\n",
       "      <td>p6</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629227</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>-5.942512</td>\n",
       "      <td>1</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048644</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629386</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>7.409069</td>\n",
       "      <td>1</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048645</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629441</td>\n",
       "      <td>lasT</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>-8.417367</td>\n",
       "      <td>4</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048646</th>\n",
       "      <td>p3</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629600</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>2.403010</td>\n",
       "      <td>35</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1048647 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        label CHROM  Position  Gene Allele   Annotation generation         AG  \\\n",
       "0          m6  chrI      1457  thrA   G->A     missense       AC:0  -8.160701   \n",
       "1          m2  chrI      2098  thrA   A->G     missense       AC:0   0.007780   \n",
       "2          m2  chrI      2942  thrB   A->G     missense       AC:0  -4.953477   \n",
       "3          m4  chrI      5437  yaaX   A->G     missense       AC:0 -10.279634   \n",
       "4          m5  chrI      7129  yaaJ   A->C     missense       AC:0   2.182288   \n",
       "...       ...   ...       ...   ...    ...          ...        ...        ...   \n",
       "1048642    m3  chrI   4629225  lasT   C->T     missense   AC:57500  -5.614587   \n",
       "1048643    p6  chrI   4629227  lasT   A->C   synonymous   AC:57500  -5.942512   \n",
       "1048644    m2  chrI   4629386  lasT   A->G   synonymous   AC:57500   7.409069   \n",
       "1048645    m2  chrI   4629441  lasT   C->T     missense   AC:57500  -8.417367   \n",
       "1048646    p3  chrI   4629600  lasT   A->G     missense   AC:57500   2.403010   \n",
       "\n",
       "         allele count  generation_number      mutator  \n",
       "0                   1                  0  non mutator  \n",
       "1                   1                  0      mutator  \n",
       "2                   1                  0      mutator  \n",
       "3                   1                  0      mutator  \n",
       "4                   2                  0  non mutator  \n",
       "...               ...                ...          ...  \n",
       "1048642            72              57500      mutator  \n",
       "1048643             1              57500      mutator  \n",
       "1048644             1              57500      mutator  \n",
       "1048645             4              57500      mutator  \n",
       "1048646            35              57500      mutator  \n",
       "\n",
       "[1048647 rows x 11 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# all_mutation.rename(columns = {\"accumulated_gain\":\"kGain 10 mer\"}, inplace=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Extract all mutation of 57500 generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>Position</th>\n",
       "      <th>label</th>\n",
       "      <th>Allele</th>\n",
       "      <th>generation_number</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>accumulated_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intergenic</td>\n",
       "      <td>63</td>\n",
       "      <td>p6</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>57500</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>-10.385914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>thrL</td>\n",
       "      <td>201</td>\n",
       "      <td>p6</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>57500</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>-1.750693</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>thrA</td>\n",
       "      <td>322</td>\n",
       "      <td>p3</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>57500</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>1.532477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>thrA</td>\n",
       "      <td>535</td>\n",
       "      <td>m2</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>57500</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense</td>\n",
       "      <td>-5.822150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>thrA</td>\n",
       "      <td>1170</td>\n",
       "      <td>p3</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>57500</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense</td>\n",
       "      <td>2.256065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15255</th>\n",
       "      <td>lasT</td>\n",
       "      <td>4629225</td>\n",
       "      <td>m3</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>57500</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense</td>\n",
       "      <td>-5.614587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15256</th>\n",
       "      <td>lasT</td>\n",
       "      <td>4629227</td>\n",
       "      <td>p6</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>57500</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>-5.942512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15257</th>\n",
       "      <td>lasT</td>\n",
       "      <td>4629386</td>\n",
       "      <td>m2</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>57500</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>7.409069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15258</th>\n",
       "      <td>lasT</td>\n",
       "      <td>4629441</td>\n",
       "      <td>m2</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>57500</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense</td>\n",
       "      <td>-8.417367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15259</th>\n",
       "      <td>lasT</td>\n",
       "      <td>4629600</td>\n",
       "      <td>p3</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>57500</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>2.403010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15260 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Gene  Position label Allele  generation_number Ref_allele  \\\n",
       "0      intergenic        63    p6   A->C              57500          A   \n",
       "1            thrL       201    p6   T->G              57500          T   \n",
       "2            thrA       322    p3   A->G              57500          A   \n",
       "3            thrA       535    m2   C->T              57500          C   \n",
       "4            thrA      1170    p3   C->T              57500          C   \n",
       "...           ...       ...   ...    ...                ...        ...   \n",
       "15255        lasT   4629225    m3   C->T              57500          C   \n",
       "15256        lasT   4629227    p6   A->C              57500          A   \n",
       "15257        lasT   4629386    m2   A->G              57500          A   \n",
       "15258        lasT   4629441    m2   C->T              57500          C   \n",
       "15259        lasT   4629600    p3   A->G              57500          A   \n",
       "\n",
       "      Alt_allele   Annotation  accumulated_gain  \n",
       "0              C    noncoding        -10.385914  \n",
       "1              G     missense         -1.750693  \n",
       "2              G    noncoding          1.532477  \n",
       "3              T     missense         -5.822150  \n",
       "4              T     missense          2.256065  \n",
       "...          ...          ...               ...  \n",
       "15255          T     missense         -5.614587  \n",
       "15256          C   synonymous         -5.942512  \n",
       "15257          G   synonymous          7.409069  \n",
       "15258          T     missense         -8.417367  \n",
       "15259          G     missense          2.403010  \n",
       "\n",
       "[15260 rows x 9 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gen_57500 = all_mutation[['label', 'CHROM', 'Position', 'Gene', 'Allele', \"generation_number\"]]\n",
    "df_gen_57500 = df_gen_57500[df_gen_57500[\"generation_number\"]==57500]\n",
    "df_gen_57500 = pd.merge(df_gen_57500[[ 'Gene', 'Position', \"label\", \"Allele\", \"generation_number\"]], df_Gain_calculated, on = [ 'Gene', 'Position', \"label\", \"Allele\"])\n",
    "df_gen_57500"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def return_mutated_fasta(data, fasta_seq, generation_number):\n",
    "\n",
    "    data = data.copy()\n",
    "\n",
    "    tmp = data[['label', 'CHROM', 'Position', 'Gene', 'Allele', \"generation_number\"]].drop_duplicates()\n",
    "    tmp_df_new = tmp[tmp[\"generation_number\"]==generation_number]\n",
    "    tmp_df_new = pd.merge(tmp_df_new[[ 'Gene', 'Position', \"label\", \"Allele\", \"generation_number\"]], df_Gain_calculated, on = [ 'Gene', 'Position', \"label\", \"Allele\"])\n",
    "    # df_gen_57500\n",
    "\n",
    "    dct_fasta_sequence = dict()\n",
    "\n",
    "    def replace_multiple_chars(s, positions, new_chars):\n",
    "        s_list = list(s)  # Convert to mutable list\n",
    "        \n",
    "        for pos, char in zip(positions, new_chars):\n",
    "            if 0 <= pos < len(s_list):  # Ensure position is within bounds\n",
    "                s_list[pos] = char\n",
    "        \n",
    "        return ''.join(s_list)  # Convert back to string\n",
    "\n",
    "\n",
    "    for i in set(tmp_df_new[\"label\"]):\n",
    "\n",
    "        tmp_df = tmp_df_new[tmp_df_new[\"label\"]==i]\n",
    "\n",
    "        tmp_df.reset_index(inplace=True, drop=True)\n",
    "\n",
    "        dct_fasta_sequence[i] = fasta_seq\n",
    "\n",
    "        for j in range(tmp_df.shape[0]):\n",
    "\n",
    "            dct_fasta_sequence[i] = replace_multiple_chars(dct_fasta_sequence[i], positions = list(tmp_df[\"Position\"].astype(int)-1), new_chars = list(tmp_df[\"Alt_allele\"]))\n",
    "\n",
    "        # break\n",
    "\n",
    "    return dct_fasta_sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Store mutated fasta population wise in a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "dct_fasta_57500 = return_mutated_fasta(data = all_mutation, fasta_seq = seq_fasta, generation_number = 57500)\n",
    "dct_fasta_0 = return_mutated_fasta(data = all_mutation, fasta_seq = seq_fasta, generation_number = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>label</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>generation</th>\n",
       "      <th>AG</th>\n",
       "      <th>allele count</th>\n",
       "      <th>generation_number</th>\n",
       "      <th>mutator</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>m6</td>\n",
       "      <td>chrI</td>\n",
       "      <td>1457</td>\n",
       "      <td>thrA</td>\n",
       "      <td>G-&gt;A</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-8.160701</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>non mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>2098</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>0.007780</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>2942</td>\n",
       "      <td>thrB</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-4.953477</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m4</td>\n",
       "      <td>chrI</td>\n",
       "      <td>5437</td>\n",
       "      <td>yaaX</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>-10.279634</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>m5</td>\n",
       "      <td>chrI</td>\n",
       "      <td>7129</td>\n",
       "      <td>yaaJ</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:0</td>\n",
       "      <td>2.182288</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>non mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048642</th>\n",
       "      <td>m3</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629225</td>\n",
       "      <td>lasT</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>-5.614587</td>\n",
       "      <td>72</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048643</th>\n",
       "      <td>p6</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629227</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>-5.942512</td>\n",
       "      <td>1</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048644</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629386</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>7.409069</td>\n",
       "      <td>1</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048645</th>\n",
       "      <td>m2</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629441</td>\n",
       "      <td>lasT</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>-8.417367</td>\n",
       "      <td>4</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1048646</th>\n",
       "      <td>p3</td>\n",
       "      <td>chrI</td>\n",
       "      <td>4629600</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>missense</td>\n",
       "      <td>AC:57500</td>\n",
       "      <td>2.403010</td>\n",
       "      <td>35</td>\n",
       "      <td>57500</td>\n",
       "      <td>mutator</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1048647 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        label CHROM  Position  Gene Allele   Annotation generation         AG  \\\n",
       "0          m6  chrI      1457  thrA   G->A     missense       AC:0  -8.160701   \n",
       "1          m2  chrI      2098  thrA   A->G     missense       AC:0   0.007780   \n",
       "2          m2  chrI      2942  thrB   A->G     missense       AC:0  -4.953477   \n",
       "3          m4  chrI      5437  yaaX   A->G     missense       AC:0 -10.279634   \n",
       "4          m5  chrI      7129  yaaJ   A->C     missense       AC:0   2.182288   \n",
       "...       ...   ...       ...   ...    ...          ...        ...        ...   \n",
       "1048642    m3  chrI   4629225  lasT   C->T     missense   AC:57500  -5.614587   \n",
       "1048643    p6  chrI   4629227  lasT   A->C   synonymous   AC:57500  -5.942512   \n",
       "1048644    m2  chrI   4629386  lasT   A->G   synonymous   AC:57500   7.409069   \n",
       "1048645    m2  chrI   4629441  lasT   C->T     missense   AC:57500  -8.417367   \n",
       "1048646    p3  chrI   4629600  lasT   A->G     missense   AC:57500   2.403010   \n",
       "\n",
       "         allele count  generation_number      mutator  \n",
       "0                   1                  0  non mutator  \n",
       "1                   1                  0      mutator  \n",
       "2                   1                  0      mutator  \n",
       "3                   1                  0      mutator  \n",
       "4                   2                  0  non mutator  \n",
       "...               ...                ...          ...  \n",
       "1048642            72              57500      mutator  \n",
       "1048643             1              57500      mutator  \n",
       "1048644             1              57500      mutator  \n",
       "1048645             4              57500      mutator  \n",
       "1048646            35              57500      mutator  \n",
       "\n",
       "[1048647 rows x 11 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_mutation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Verify few position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('C', 'T')"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos = 77883\n",
    "seq_fasta[pos-1], dct_fasta_57500[\"m5\"][pos-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1155840/1595551989.py:101: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  df_complete = pd.concat([df_complete, df_sliding])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Position</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Ref_allele</th>\n",
       "      <th>Alt_allele</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>label</th>\n",
       "      <th>accumulated_gain</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>63</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>p6</td>\n",
       "      <td>-1.937942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>201</td>\n",
       "      <td>thrL</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>1.157555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>241</td>\n",
       "      <td>thrL</td>\n",
       "      <td>A-&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>-0.911836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>309</td>\n",
       "      <td>thrA</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>m1</td>\n",
       "      <td>3.989634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>322</td>\n",
       "      <td>thrA</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>noncoding</td>\n",
       "      <td>p3</td>\n",
       "      <td>8.106801</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36917</th>\n",
       "      <td>4629541</td>\n",
       "      <td>lasT</td>\n",
       "      <td>C-&gt;T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>missense</td>\n",
       "      <td>m2</td>\n",
       "      <td>-5.104473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36918</th>\n",
       "      <td>4629573</td>\n",
       "      <td>lasT</td>\n",
       "      <td>T-&gt;C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>synonymous</td>\n",
       "      <td>m4</td>\n",
       "      <td>-7.698813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36919</th>\n",
       "      <td>4629600</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>p3</td>\n",
       "      <td>5.607030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36920</th>\n",
       "      <td>4629600</td>\n",
       "      <td>lasT</td>\n",
       "      <td>A-&gt;G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>p6</td>\n",
       "      <td>2.753987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36921</th>\n",
       "      <td>4629764</td>\n",
       "      <td>lasT</td>\n",
       "      <td>T-&gt;G</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>missense</td>\n",
       "      <td>m1</td>\n",
       "      <td>-9.703003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36922 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Position        Gene Allele Ref_allele Alt_allele   Annotation label  \\\n",
       "0            63  intergenic   A->C          A          C    noncoding    p6   \n",
       "1           201        thrL   T->G          T          G     missense    p6   \n",
       "2           241        thrL   A->C          A          C     missense    m1   \n",
       "3           309        thrA   T->G          T          G    noncoding    m1   \n",
       "4           322        thrA   A->G          A          G    noncoding    p3   \n",
       "...         ...         ...    ...        ...        ...          ...   ...   \n",
       "36917   4629541        lasT   C->T          C          T     missense    m2   \n",
       "36918   4629573        lasT   T->C          T          C   synonymous    m4   \n",
       "36919   4629600        lasT   A->G          A          G     missense    p3   \n",
       "36920   4629600        lasT   A->G          A          G     missense    p6   \n",
       "36921   4629764        lasT   T->G          T          G     missense    m1   \n",
       "\n",
       "       accumulated_gain  \n",
       "0             -1.937942  \n",
       "1              1.157555  \n",
       "2             -0.911836  \n",
       "3              3.989634  \n",
       "4              8.106801  \n",
       "...                 ...  \n",
       "36917         -5.104473  \n",
       "36918         -7.698813  \n",
       "36919          5.607030  \n",
       "36920          2.753987  \n",
       "36921         -9.703003  \n",
       "\n",
       "[36922 rows x 8 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "m1_population = pd.read_excel(\"../../../data/m1_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m1_population[\"label\"] = \"m1\" \n",
    "m1_population = m1_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m2_population = pd.read_excel(\"../../../data/m2_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m2_population[\"label\"] = \"m2\" \n",
    "m2_population = m2_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m3_population = pd.read_excel(\"../../../data/m3_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m3_population[\"label\"] = \"m3\" \n",
    "m3_population = m3_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m4_population = pd.read_excel(\"../../../data/m4_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m4_population[\"label\"] = \"m4\" \n",
    "m4_population = m4_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m5_population = pd.read_excel(\"../../../data/m5_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m5_population[\"label\"] = \"m5\" \n",
    "m5_population = m5_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "m6_population = pd.read_excel(\"../../../data/m6_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "m6_population[\"label\"] = \"m6\" \n",
    "m6_population = m6_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p1_population = pd.read_excel(\"../../../data/p1_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p1_population[\"label\"] = \"p1\" \n",
    "p1_population = p1_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p2_population = pd.read_excel(\"../../../data/p2_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p2_population[\"label\"] = \"p2\" \n",
    "p2_population = p2_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p3_population = pd.read_excel(\"../../../data/p3_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p3_population[\"label\"] = \"p3\" \n",
    "p3_population = p3_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p4_population = pd.read_excel(\"../../../data/p4_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p4_population[\"label\"] = \"p4\" \n",
    "p4_population = p4_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p5_population = pd.read_excel(\"../../../data/p5_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p5_population[\"label\"] = \"p5\" \n",
    "p5_population = p5_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "p6_population = pd.read_excel(\"../../../data/p6_annotated_timecourse.xlsx\",\"Sheet2\") \n",
    "p6_population[\"label\"] = \"p6\" \n",
    "p6_population = p6_population[['Position','Gene','Allele','Annotation','Test.statistic','P-value','Ref_allele','Alt_allele','ref_flanking_regions','alt_flanking_regions','label']]\n",
    "\n",
    "df = pd.concat([m1_population,m2_population,m3_population,m4_population,m5_population,m6_population,p1_population,p2_population,p3_population,p4_population,p5_population,p6_population])\n",
    "df.reset_index(inplace=True, drop=True)\n",
    "\n",
    "fcgr_dict = dict()\n",
    "\n",
    "for i in dct_fasta_57500.keys():\n",
    "\n",
    "    fcgr_dict[i] = kaos.chaos_frequency_dictionary(fasta_string= dct_fasta_57500[i], kmer_length=kmer_length, pseudo_count=True, chaos_game_kmer_array=kaos.chaos_game_representation_key(kmer_length = kmer_length))\n",
    "\n",
    "df[\"Gene\"] = df[\"Gene\"].str.strip()\n",
    "df[\"Allele\"] = df[\"Allele\"].str.strip()\n",
    "\n",
    "df_sliding = pd.DataFrame(columns = ['Position', 'Gene', 'Allele', 'Annotation', 'Test.statistic', 'P-value',\n",
    "       'Ref_allele', 'Alt_allele', 'ref_flanking_regions',\n",
    "       'alt_flanking_regions', 'label'])\n",
    "df_complete = pd.DataFrame(columns = ['Position', 'Gene', 'Allele', 'Annotation', 'Test.statistic', 'P-value',\n",
    "       'Ref_allele', 'Alt_allele', 'ref_flanking_regions',\n",
    "       'alt_flanking_regions', 'label'])\n",
    "\n",
    "for i in range(df.shape[0]):\n",
    "\n",
    "    pop = df.loc[i, \"label\"]\n",
    "    \n",
    "    ref_seq = df.loc[i, \"ref_flanking_regions\"]\n",
    "    alt_seq = df.loc[i, \"alt_flanking_regions\"]\n",
    "    \n",
    "    center_pos = int(len(ref_seq)/2)\n",
    "    \n",
    "    gain = 0\n",
    "    \n",
    "    for j in range(kmer_length):\n",
    "        \n",
    "        df_sliding.loc[j, \"altered sliding window\"] = alt_seq[center_pos - kmer_length+j+1:center_pos+1+j]\n",
    "        df_sliding.loc[j, \"reference sliding window\"] = ref_seq[center_pos - kmer_length+j+1:center_pos+j+1]\n",
    "\n",
    "        df_sliding.loc[j, \"gain\"] = np.log(fcgr_dict[pop][df_sliding.loc[j, \"altered sliding window\"]]/ fcgr_dict[pop][df_sliding.loc[j, \"reference sliding window\"]])\n",
    "    \n",
    "        gain = gain + df_sliding.loc[j, \"gain\"]\n",
    "    \n",
    "    df_sliding['Position'] = df.loc[i, 'Position']\n",
    "    df_sliding['Gene'] = df.loc[i, 'Gene']\n",
    "    df_sliding['Allele'] = df.loc[i, 'Allele']\n",
    "    df_sliding['Annotation'] = df.loc[i, 'Annotation']\n",
    "    df_sliding['Test.statistic'] = df.loc[i, 'Test.statistic']\n",
    "    df_sliding['P-value'] = df.loc[i, 'P-value']\n",
    "    df_sliding['Ref_allele'] = df.loc[i, 'Ref_allele']\n",
    "    df_sliding['Alt_allele'] = df.loc[i, 'Alt_allele']\n",
    "    df_sliding['ref_flanking_regions'] = df.loc[i, 'ref_flanking_regions']\n",
    "    df_sliding['alt_flanking_regions'] = df.loc[i, 'alt_flanking_regions']\n",
    "    df_sliding['label'] = df.loc[i, 'label']\n",
    "    df_sliding[\"accumulated_gain\"] = gain\n",
    "\n",
    "    df_complete = pd.concat([df_complete, df_sliding])\n",
    "\n",
    "\n",
    "df_complete_groupby = df_complete.groupby([\"Position\", \"Gene\", \"Allele\", \"Ref_allele\", \"Alt_allele\", \"Annotation\", \"label\"])[\"accumulated_gain\"].mean().reset_index()\n",
    "df_complete_groupby\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df_complete_groupby.rename({\"accumulated_gain\": \"kGain 57.5k\"}, inplace=True)\n",
    "# df_complete_groupby.to_csv(\"accumulated_gain with 57.5k as reference.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### pickle population wise mutated fasta wrt 57.5k generation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import pickle\n",
    "\n",
    "\n",
    "\n",
    "# # Open a file in write-binary mode and save the data\n",
    "# with open(\"dct_fasta_57500.pkl\", \"wb\") as file:\n",
    "#     pickle.dump(dct_fasta_57500, file)\n",
    "\n",
    "# print(\"Data has been pickled successfully!\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "amr",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
